

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>selene_sdk.samplers.online_sampler &mdash; Selene 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-selene-with-pip">Installing selene with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene_sdk.samplers.online_sampler</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene_sdk.samplers.online_sampler</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module provides the `OnlineSampler` class and supporting methods.</span>
<span class="sd">Objects of the class `OnlineSampler`, are samplers which load examples</span>
<span class="sd">&quot;on the fly&quot; rather than storing them all persistently in memory.</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">abc</span> <span class="kn">import</span> <span class="n">ABCMeta</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">.sampler</span> <span class="kn">import</span> <span class="n">Sampler</span>
<span class="kn">from</span> <span class="nn">..targets</span> <span class="kn">import</span> <span class="n">GenomicFeatures</span>


<div class="viewcode-block" id="OnlineSampler"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler">[docs]</a><span class="k">class</span> <span class="nc">OnlineSampler</span><span class="p">(</span><span class="n">Sampler</span><span class="p">,</span> <span class="n">metaclass</span><span class="o">=</span><span class="n">ABCMeta</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A sampler in which training/validation/test data is constructed</span>
<span class="sd">    from random sampling of the dataset for each batch passed to the</span>
<span class="sd">    model. This form of sampling may alleviate the problem of loading an</span>
<span class="sd">    extremely large dataset into memory when developing a new model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    reference_sequence : selene_sdk.sequences.Sequence</span>
<span class="sd">        A reference sequence from which to create examples.</span>
<span class="sd">    target_path : str</span>
<span class="sd">        Path to tabix-indexed, compressed BED file (`*.bed.gz`) of genomic</span>
<span class="sd">        coordinates mapped to the genomic features we want to predict.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        List of distinct features that we aim to predict.</span>
<span class="sd">    seed : int, optional</span>
<span class="sd">        Default is 436. Sets the random seed for sampling.</span>
<span class="sd">    validation_holdout : list(str) or float, optional</span>
<span class="sd">        Default is `[&#39;chr6&#39;, &#39;chr7&#39;]`. Holdout can be regional or</span>
<span class="sd">        proportional. If regional, expects a list (e.g. `[&#39;X&#39;, &#39;Y&#39;]`).</span>
<span class="sd">        Regions must match those specified in the first column of the</span>
<span class="sd">        tabix-indexed BED file. If proportional, specify a percentage</span>
<span class="sd">        between (0.0, 1.0). Typically 0.10 or 0.20.</span>
<span class="sd">    test_holdout : list(str) or float, optional</span>
<span class="sd">        Default is `[&#39;chr8&#39;, &#39;chr9&#39;]`. See documentation for</span>
<span class="sd">        `validation_holdout` for additional information.</span>
<span class="sd">    sequence_length : int, optional</span>
<span class="sd">        Default is 1000. Model is trained on sequences of `sequence_length`</span>
<span class="sd">        where genomic features are annotated to the center regions of</span>
<span class="sd">        these sequences.</span>
<span class="sd">    center_bin_to_predict : int, optional</span>
<span class="sd">        Default is 200. Query the tabix-indexed file for a region of</span>
<span class="sd">        length `center_bin_to_predict`.</span>
<span class="sd">    feature_thresholds : float [0.0, 1.0], optional</span>
<span class="sd">        Default is 0.5. The `feature_threshold` to pass to the</span>
<span class="sd">        `GenomicFeatures` object.</span>
<span class="sd">    mode : {&#39;train&#39;, &#39;validate&#39;, &#39;test&#39;}, optional</span>
<span class="sd">        Default is `&#39;train&#39;`. The mode to run the sampler in.</span>
<span class="sd">    save_datasets : list(str), optional</span>
<span class="sd">        Default is `[]` the empty list. The list of modes for which we should</span>
<span class="sd">        save the sampled data to file (e.g. `[&quot;test&quot;, &quot;validate&quot;]`).</span>
<span class="sd">    output_dir : str or None, optional</span>
<span class="sd">        Default is None. The path to the directory where we should</span>
<span class="sd">        save sampled examples for a mode. If `save_datasets` is</span>
<span class="sd">        a non-empty list, `output_dir` must be specified. If</span>
<span class="sd">        the path in `output_dir` does not exist it will be created</span>
<span class="sd">        automatically.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    reference_sequence : selene_sdk.sequences.Sequence</span>
<span class="sd">        The reference sequence that examples are created from.</span>
<span class="sd">    target : selene_sdk.targets.Target</span>
<span class="sd">        The `selene_sdk.targets.Target` object holding the features that we</span>
<span class="sd">        would like to predict.</span>
<span class="sd">    validation_holdout : list(str) or float</span>
<span class="sd">        The samples to hold out for validating model performance. These</span>
<span class="sd">        can be &quot;regional&quot; or &quot;proportional&quot;. If regional, this is a list</span>
<span class="sd">        of region names (e.g. `[&#39;chrX&#39;, &#39;chrY&#39;]`). These regions must</span>
<span class="sd">        match those specified in the first column of the tabix-indexed</span>
<span class="sd">        BED file. If proportional, this is the fraction of total samples</span>
<span class="sd">        that will be held out.</span>
<span class="sd">    test_holdout : list(str) or float</span>
<span class="sd">        The samples to hold out for testing model performance. See the</span>
<span class="sd">        documentation for `validation_holdout` for more details.</span>
<span class="sd">    sequence_length : int</span>
<span class="sd">        The length of the sequences to  train the model on.</span>
<span class="sd">    modes : list(str)</span>
<span class="sd">        The list of modes that the sampler can be run in.</span>
<span class="sd">    mode : str</span>
<span class="sd">        The current mode that the sampler is running in. Must be one of</span>
<span class="sd">        the modes listed in `modes`.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">            If `mode` is not a valid mode.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the parities of `sequence_length` and `center_bin_to_predict`</span>
<span class="sd">        are not the same.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If `sequence_length` is smaller than `center_bin_to_predict` is.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the types of `validation_holdout` and `test_holdout` are not</span>
<span class="sd">        the same.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">STRAND_SIDES</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">)</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Defines the strands that features can be sampled from.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">reference_sequence</span><span class="p">,</span>
                 <span class="n">target_path</span><span class="p">,</span>
                 <span class="n">features</span><span class="p">,</span>
                 <span class="n">seed</span><span class="o">=</span><span class="mi">436</span><span class="p">,</span>
                 <span class="n">validation_holdout</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr6&#39;</span><span class="p">,</span> <span class="s1">&#39;chr7&#39;</span><span class="p">],</span>
                 <span class="n">test_holdout</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr8&#39;</span><span class="p">,</span> <span class="s1">&#39;chr9&#39;</span><span class="p">],</span>
                 <span class="n">sequence_length</span><span class="o">=</span><span class="mi">1001</span><span class="p">,</span>
                 <span class="n">center_bin_to_predict</span><span class="o">=</span><span class="mi">201</span><span class="p">,</span>
                 <span class="n">feature_thresholds</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
                 <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;train&quot;</span><span class="p">,</span>
                 <span class="n">save_datasets</span><span class="o">=</span><span class="p">[],</span>
                 <span class="n">output_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a new `OnlineSampler` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">OnlineSampler</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">features</span><span class="p">,</span>
            <span class="n">save_datasets</span><span class="o">=</span><span class="n">save_datasets</span><span class="p">,</span>
            <span class="n">output_dir</span><span class="o">=</span><span class="n">output_dir</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">seed</span> <span class="o">=</span> <span class="n">seed</span>
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seed</span><span class="p">)</span>
        <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seed</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">center_bin_to_predict</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">sequence_length</span> <span class="o">+</span> <span class="n">center_bin_to_predict</span><span class="p">)</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Sequence length of </span><span class="si">{0}</span><span class="s2"> with a center bin length of </span><span class="si">{1}</span><span class="s2"> &quot;</span>
                    <span class="s2">&quot;is invalid. These 2 inputs should both be odd or both be &quot;</span>
                    <span class="s2">&quot;even.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">center_bin_to_predict</span><span class="p">))</span>

        <span class="c1"># specifying a test holdout partition is optional</span>
        <span class="k">if</span> <span class="n">test_holdout</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;test&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,))</span> <span class="ow">and</span> \
                    <span class="nb">isinstance</span><span class="p">(</span><span class="n">test_holdout</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">validation_holdout</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">test_holdout</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">test_holdout</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_holdout_type</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="ow">and</span> \
                    <span class="nb">isinstance</span><span class="p">(</span><span class="n">test_holdout</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="n">validation_holdout</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">test_holdout</span> <span class="o">=</span> <span class="n">test_holdout</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_holdout_type</span> <span class="o">=</span> <span class="s2">&quot;proportion&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Validation holdout and test holdout must have the &quot;</span>
                    <span class="s2">&quot;same type (list or float) but validation was &quot;</span>
                    <span class="s2">&quot;type </span><span class="si">{0}</span><span class="s2"> and test was type </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="nb">type</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">),</span> <span class="nb">type</span><span class="p">(</span><span class="n">test_holdout</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">test_holdout</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">validation_holdout</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_holdout_type</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="n">validation_holdout</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_holdout_type</span> <span class="o">=</span> <span class="s2">&quot;proportion&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Validation holdout must be of type list (chromosomal &quot;</span>
                    <span class="s2">&quot;holdout) or float (proportion holdout) but was type &quot;</span>
                    <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">)))</span>

        <span class="k">if</span> <span class="n">mode</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Mode must be one of </span><span class="si">{0}</span><span class="s2">. Input was &#39;</span><span class="si">{1}</span><span class="s2">&#39;.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="p">,</span> <span class="n">mode</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="n">mode</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">=</span> <span class="n">sequence_length</span>
        <span class="n">window_radius</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_start_window_radius</span> <span class="o">=</span> <span class="n">window_radius</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_end_window_radius</span> <span class="o">=</span> <span class="n">window_radius</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_end_window_radius</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">center_bin_to_predict</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">bin_radius</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">center_bin_to_predict</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">=</span> <span class="n">bin_radius</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">=</span> <span class="n">bin_radius</span>
            <span class="k">if</span> <span class="n">center_bin_to_predict</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">center_bin_to_predict</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">or</span> \
                    <span class="nb">len</span><span class="p">(</span><span class="n">center_bin_to_predict</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;`center_bin_to_predict` needs to be either an int or a list of &quot;</span>
                    <span class="s2">&quot;two ints, but was type &#39;</span><span class="si">{0}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="nb">type</span><span class="p">(</span><span class="n">center_bin_to_predict</span><span class="p">)))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">bin_start</span><span class="p">,</span> <span class="n">bin_end</span> <span class="o">=</span> <span class="n">center_bin_to_predict</span>
                <span class="k">if</span> <span class="n">bin_start</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">bin_end</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                    <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;center_bin_to_predict [</span><span class="si">{0}</span><span class="s2">, </span><span class="si">{1}</span><span class="s2">]&quot;</span>
                        <span class="s2">&quot;is out-of-bound for sequence length </span><span class="si">{3}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">bin_start</span><span class="p">,</span> <span class="n">bin_end</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">))</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_window_radius</span> <span class="o">-</span> <span class="n">bin_start</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_end_window_radius</span> <span class="o">-</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="n">bin_end</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span> <span class="o">=</span> <span class="n">reference_sequence</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_features</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_features</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">GenomicFeatures</span><span class="p">(</span>
            <span class="n">target_path</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_features</span><span class="p">,</span>
            <span class="n">feature_thresholds</span><span class="o">=</span><span class="n">feature_thresholds</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_filehandles</span> <span class="o">=</span> <span class="p">{}</span>

<div class="viewcode-block" id="OnlineSampler.get_feature_from_index"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler.get_feature_from_index">[docs]</a>    <span class="k">def</span> <span class="nf">get_feature_from_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the feature corresponding to an index in the feature</span>
<span class="sd">        vector.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        index : int</span>
<span class="sd">            The index of the feature to retrieve the name for.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            The name of the feature occurring at the specified index.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">index_feature_dict</span><span class="p">[</span><span class="n">index</span><span class="p">]</span></div>

<div class="viewcode-block" id="OnlineSampler.get_sequence_from_encoding"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler.get_sequence_from_encoding">[docs]</a>    <span class="k">def</span> <span class="nf">get_sequence_from_encoding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">encoding</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Gets the string sequence from the one-hot encoding</span>
<span class="sd">        of the sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        encoding : numpy.ndarray</span>
<span class="sd">            An :math:`L \\times N` array (where :math:`L` is the length</span>
<span class="sd">            of the sequence and :math:`N` is the size of the sequence</span>
<span class="sd">            type&#39;s alphabet) containing the one-hot encoding of the</span>
<span class="sd">            sequence.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            The sequence of :math:`L` characters decoded from the input.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">encoding_to_sequence</span><span class="p">(</span><span class="n">encoding</span><span class="p">)</span></div>

<div class="viewcode-block" id="OnlineSampler.save_dataset_to_file"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler.save_dataset_to_file">[docs]</a>    <span class="k">def</span> <span class="nf">save_dataset_to_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">close_filehandle</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Save samples for each partition (i.e. train/validate/test) to</span>
<span class="sd">        disk.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mode : str</span>
<span class="sd">            Must be one of the modes specified in `save_datasets` during</span>
<span class="sd">            sampler initialization.</span>
<span class="sd">        close_filehandle : bool, optional</span>
<span class="sd">            Default is False. `close_filehandle=True` assumes that all</span>
<span class="sd">            data corresponding to the input `mode` has been saved to</span>
<span class="sd">            file and `save_dataset_to_file` will not be called with</span>
<span class="sd">            `mode` again.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_save_datasets</span><span class="p">:</span>
            <span class="k">return</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_save_datasets</span><span class="p">[</span><span class="n">mode</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_save_filehandles</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_save_filehandles</span><span class="p">[</span><span class="n">mode</span><span class="p">]</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_output_dir</span><span class="p">,</span>
                             <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_data.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mode</span><span class="p">)),</span>
                <span class="s1">&#39;w+&#39;</span><span class="p">)</span>
        <span class="n">file_handle</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_save_filehandles</span><span class="p">[</span><span class="n">mode</span><span class="p">]</span>
        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="n">samples</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">line</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">])</span>
            <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">close_filehandle</span><span class="p">:</span>
            <span class="n">file_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="OnlineSampler.get_data_and_targets"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler.get_data_and_targets">[docs]</a>    <span class="k">def</span> <span class="nf">get_data_and_targets</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method fetches a subset of the data from the sampler,</span>
<span class="sd">        divided into batches. This method also allows the user to</span>
<span class="sd">        specify what operating mode to run the sampler in when fetching</span>
<span class="sd">        the data.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        batch_size : int</span>
<span class="sd">            The size of the batches to divide the data into.</span>
<span class="sd">        n_samples : int or None, optional</span>
<span class="sd">            Default is None. The total number of samples to retrieve.</span>
<span class="sd">            If `n_samples` is None and the mode is `validate`, will</span>
<span class="sd">            set `n_samples` to 32000; if the mode is `test`, will set</span>
<span class="sd">            `n_samples` to 640000 if it is None. If the mode is `train`</span>
<span class="sd">            you must have specified a value for `n_samples`.</span>
<span class="sd">        mode : str, optional</span>
<span class="sd">            Default is None. The mode to run the sampler in when</span>
<span class="sd">            fetching the samples. See</span>
<span class="sd">            `selene_sdk.samplers.IntervalsSampler.modes` for more</span>
<span class="sd">            information. If None, will use the current mode `self.mode`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sequences_and_targets, targets_matrix : \</span>
<span class="sd">        tuple(list(tuple(numpy.ndarray, numpy.ndarray)), numpy.ndarray)</span>
<span class="sd">            Tuple containing the list of sequence-target pairs, as well</span>
<span class="sd">            as a single matrix with all targets in the same order.</span>
<span class="sd">            Note that `sequences_and_targets`&#39;s sequence elements are of</span>
<span class="sd">            the shape :math:`B \\times L \\times N` and its target</span>
<span class="sd">            elements are of the shape :math:`B \\times F`, where</span>
<span class="sd">            :math:`B` is `batch_size`, :math:`L` is the sequence length,</span>
<span class="sd">            :math:`N` is the size of the sequence type&#39;s alphabet, and</span>
<span class="sd">            :math:`F` is the number of features. Further,</span>
<span class="sd">            `target_matrix` is of the shape :math:`S \\times F`, where</span>
<span class="sd">            :math:`S =` `n_samples`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">set_mode</span><span class="p">(</span><span class="n">mode</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mode</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span>
        <span class="n">sequences_and_targets</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">n_samples</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;validate&quot;</span><span class="p">:</span>
            <span class="n">n_samples</span> <span class="o">=</span> <span class="mi">32000</span>
        <span class="k">elif</span> <span class="n">n_samples</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;test&quot;</span><span class="p">:</span>
            <span class="n">n_samples</span> <span class="o">=</span> <span class="mi">640000</span>

        <span class="n">n_batches</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_samples</span> <span class="o">/</span> <span class="n">batch_size</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_batches</span><span class="p">):</span>
            <span class="n">inputs</span><span class="p">,</span> <span class="n">targets</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">batch_size</span><span class="p">)</span>
            <span class="n">sequences_and_targets</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">inputs</span><span class="p">,</span> <span class="n">targets</span><span class="p">))</span>
        <span class="n">targets_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">t</span> <span class="k">for</span> <span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sequences_and_targets</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_save_datasets</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">save_dataset_to_file</span><span class="p">(</span><span class="n">mode</span><span class="p">,</span> <span class="n">close_filehandle</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sequences_and_targets</span><span class="p">,</span> <span class="n">targets_mat</span></div>

<div class="viewcode-block" id="OnlineSampler.get_dataset_in_batches"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler.get_dataset_in_batches">[docs]</a>    <span class="k">def</span> <span class="nf">get_dataset_in_batches</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method returns a subset of the data for a specified run</span>
<span class="sd">        mode, divided into mini-batches.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mode : {&#39;test&#39;, &#39;validate&#39;}</span>
<span class="sd">            The mode to run the sampler in when fetching the samples.</span>
<span class="sd">            See `selene_sdk.samplers.IntervalsSampler.modes` for more</span>
<span class="sd">            information.</span>
<span class="sd">        batch_size : int</span>
<span class="sd">            The size of the batches to divide the data into.</span>
<span class="sd">        n_samples : int or None, optional</span>
<span class="sd">            Default is `None`. The total number of samples to retrieve.</span>
<span class="sd">            If `None`, it will retrieve 32000 samples if `mode` is validate</span>
<span class="sd">            or 640000 samples if `mode` is test or train.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sequences_and_targets, targets_matrix : \</span>
<span class="sd">        tuple(list(tuple(numpy.ndarray, numpy.ndarray)), numpy.ndarray)</span>
<span class="sd">            Tuple containing the list of sequence-target pairs, as well</span>
<span class="sd">            as a single matrix with all targets in the same order.</span>
<span class="sd">            The list is length :math:`S`, where :math:`S =` `n_samples`.</span>
<span class="sd">            Note that `sequences_and_targets`&#39;s sequence elements are of</span>
<span class="sd">            the shape :math:`B \\times L \\times N` and its target</span>
<span class="sd">            elements are of the shape :math:`B \\times F`, where</span>
<span class="sd">            :math:`B` is `batch_size`, :math:`L` is the sequence length,</span>
<span class="sd">            :math:`N` is the size of the sequence type&#39;s alphabet, and</span>
<span class="sd">            :math:`F` is the number of features. Further,</span>
<span class="sd">            `target_matrix` is of the shape :math:`S \\times F`</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_data_and_targets</span><span class="p">(</span>
            <span class="n">batch_size</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="n">n_samples</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">)</span></div>

<div class="viewcode-block" id="OnlineSampler.get_validation_set"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler.get_validation_set">[docs]</a>    <span class="k">def</span> <span class="nf">get_validation_set</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method returns a subset of validation data from the</span>
<span class="sd">        sampler, divided into batches.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        batch_size : int</span>
<span class="sd">            The size of the batches to divide the data into.</span>
<span class="sd">        n_samples : int or None, optional</span>
<span class="sd">            Default is `None`. The total number of validation examples</span>
<span class="sd">            to retrieve. If `None`, 32000 examples are retrieved.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sequences_and_targets, targets_matrix : \</span>
<span class="sd">        tuple(list(tuple(numpy.ndarray, numpy.ndarray)), numpy.ndarray)</span>
<span class="sd">            Tuple containing the list of sequence-target pairs, as well</span>
<span class="sd">            as a single matrix with all targets in the same order.</span>
<span class="sd">            Note that `sequences_and_targets`&#39;s sequence elements are of</span>
<span class="sd">            the shape :math:`B \\times L \\times N` and its target</span>
<span class="sd">            elements are of the shape :math:`B \\times F`, where</span>
<span class="sd">            :math:`B` is `batch_size`, :math:`L` is the sequence length,</span>
<span class="sd">            :math:`N` is the size of the sequence type&#39;s alphabet, and</span>
<span class="sd">            :math:`F` is the number of features. Further,</span>
<span class="sd">            `target_matrix` is of the shape :math:`S \\times F`, where</span>
<span class="sd">            :math:`S =` `n_samples`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_dataset_in_batches</span><span class="p">(</span>
            <span class="s2">&quot;validate&quot;</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="n">n_samples</span><span class="p">)</span></div>

<div class="viewcode-block" id="OnlineSampler.get_test_set"><a class="viewcode-back" href="../../../samplers.html#selene_sdk.samplers.OnlineSampler.get_test_set">[docs]</a>    <span class="k">def</span> <span class="nf">get_test_set</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method returns a subset of testing data from the</span>
<span class="sd">        sampler, divided into batches.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        batch_size : int</span>
<span class="sd">            The size of the batches to divide the data into.</span>
<span class="sd">        n_samples : int or None, optional</span>
<span class="sd">            Default is `None`. The total number of validation examples</span>
<span class="sd">            to retrieve. If `None`, 640000 examples are retrieved.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sequences_and_targets, targets_matrix : \</span>
<span class="sd">        tuple(list(tuple(numpy.ndarray, numpy.ndarray)), numpy.ndarray)</span>
<span class="sd">            Tuple containing the list of sequence-target pairs, as well</span>
<span class="sd">            as a single matrix with all targets in the same order.</span>
<span class="sd">            Note that `sequences_and_targets`&#39;s sequence elements are of</span>
<span class="sd">            the shape :math:`B \\times L \\times N` and its target</span>
<span class="sd">            elements are of the shape :math:`B \\times F`, where</span>
<span class="sd">            :math:`B` is `batch_size`, :math:`L` is the sequence length,</span>
<span class="sd">            :math:`N` is the size of the sequence type&#39;s alphabet, and</span>
<span class="sd">            :math:`F` is the number of features. Further,</span>
<span class="sd">            `target_matrix` is of the shape :math:`S \\times F`, where</span>
<span class="sd">            :math:`S =` `n_samples`.</span>


<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If no test partition of the data was specified during</span>
<span class="sd">            sampler initialization.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;test&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No test partition of the data was specified &quot;</span>
                             <span class="s2">&quot;during initialization. Cannot use method &quot;</span>
                             <span class="s2">&quot;`get_test_set`.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_dataset_in_batches</span><span class="p">(</span><span class="s2">&quot;test&quot;</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">n_samples</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2018, Selene Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>