

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene_sdk.samplers &mdash; Selene master documentation</title>
  

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="selene_sdk.samplers.file_samplers" href="samplers.file_samplers.html" />
    <link rel="prev" title="selene_sdk.predict" href="predict.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Selene
          

          
            
            <img src="_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.4.4)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/overview.html#help">Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-with-anaconda">Installing with Anaconda</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/installation.html#installing-selene-with-pip">Installing selene with pip:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#additional-dependency-for-running-the-cli">Additional dependency for running the CLI</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#variant-effect-prediction">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id6">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#in-silico-mutagenesis"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id13">Some notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id14">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>selene_sdk.samplers</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/samplers.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene_sdk.samplers">
<span id="selene-sdk-samplers"></span><h1>selene_sdk.samplers<a class="headerlink" href="#module-selene_sdk.samplers" title="Permalink to this headline">¶</a></h1>
<p>This module provides classes and methods for sampling labeled data
examples.</p>
<div class="section" id="sampler">
<h2>Sampler<a class="headerlink" href="#sampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.Sampler">
<em class="property">class </em><code class="sig-prename descclassname">selene_sdk.samplers.</code><code class="sig-name descname">Sampler</code><span class="sig-paren">(</span><em class="sig-param">features</em>, <em class="sig-param">save_datasets=[]</em>, <em class="sig-param">output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The base class for sampler currently enforces that all samplers
have modes for drawing training and validation samples to train a
model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>features</strong> (<em>list(str)</em>) – The list of features (classes) the model predicts.</p></li>
<li><p><strong>save_datasets</strong> (<em>list(str), optional</em>) – Default is <cite>[]</cite> the empty list. The list of modes for which we should
save sampled data to file (1 or more of [‘train’, ‘validate’, ‘test’]).</p></li>
<li><p><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. Path to the output directory. Used if we save
any of the data sampled. If <cite>save_datasets</cite> is non-empty,
<cite>output_dir</cite> must be a valid path. If the directory does not
yet exist, it will be created for you.</p></li>
</ul>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>~Sampler.modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>)</em>) – A list of the names of the modes that the object may operate in.</p></li>
<li><p><strong>~Sampler.mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – The current mode that the object is operating in.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="selene_sdk.samplers.Sampler.BASE_MODES">
<code class="sig-name descname">BASE_MODES</code><em class="property"> = ('train', 'validate')</em><a class="headerlink" href="#selene_sdk.samplers.Sampler.BASE_MODES" title="Permalink to this definition">¶</a></dt>
<dd><p>The types of modes that the <cite>Sampler</cite> object can run in.</p>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.get_data_and_targets">
<em class="property">abstract </em><code class="sig-name descname">get_data_and_targets</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples</em>, <em class="sig-param">mode=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.get_data_and_targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.get_data_and_targets" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the data from the sampler,
divided into batches. This method also allows the user to
specify what operating mode to run the sampler in when fetching
the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int</em>) – The total number of samples to retrieve.</p></li>
<li><p><strong>mode</strong> (<em>str, optional</em>) – Default is None. The operating mode that the object should run in.
If None, will use the current mode <cite>self.mode</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.get_feature_from_index">
<em class="property">abstract </em><code class="sig-name descname">get_feature_from_index</code><span class="sig-paren">(</span><em class="sig-param">index</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.get_feature_from_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.get_feature_from_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the feature corresponding to an index in the feature
vector.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<em>int</em>) – The index of the feature to retrieve the name for.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The name of the feature occurring at the specified index.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.get_test_set">
<em class="property">abstract </em><code class="sig-name descname">get_test_set</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.get_test_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.get_test_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of testing data from the
sampler, divided into batches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of validation examples
to retrieve. If <cite>None</cite>, 640000 examples are retrieved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If no test partition of the data was specified during
    sampler initialization.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.get_validation_set">
<em class="property">abstract </em><code class="sig-name descname">get_validation_set</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.get_validation_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.get_validation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of validation data from the
sampler, divided into batches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int, optional</em>) – Default is None. The total number of validation examples to
retrieve. Handling for <cite>n_samples=None</cite> should be done by
all classes that subclass <cite>selene_sdk.samplers.Sampler</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.sample">
<em class="property">abstract </em><code class="sig-name descname">sample</code><span class="sig-paren">(</span><em class="sig-param">batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches a mini-batch of the data from the sampler.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The size of the batch to retrieve.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.save_dataset_to_file">
<em class="property">abstract </em><code class="sig-name descname">save_dataset_to_file</code><span class="sig-paren">(</span><em class="sig-param">mode</em>, <em class="sig-param">close_filehandle=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.save_dataset_to_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.save_dataset_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Save samples for each partition (i.e. train/validate/test) to
disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> (<em>str</em>) – Must be one of the modes specified in <cite>save_datasets</cite> during
sampler initialization.</p></li>
<li><p><strong>close_filehandle</strong> (<em>bool, optional</em>) – Default is False. <cite>close_filehandle=True</cite> assumes that all
data corresponding to the input <cite>mode</cite> has been saved to
file and <cite>save_dataset_to_file</cite> will not be called with
<cite>mode</cite> again.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.set_mode">
<code class="sig-name descname">set_mode</code><span class="sig-paren">(</span><em class="sig-param">mode</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.set_mode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.set_mode" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the sampling mode.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mode</strong> (<em>str</em>) – The name of the mode to use. It must be one of
<cite>Sampler.BASE_MODES</cite>.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If <cite>mode</cite> is not a valid mode.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="onlinesampler">
<h2>OnlineSampler<a class="headerlink" href="#onlinesampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.OnlineSampler">
<em class="property">class </em><code class="sig-prename descclassname">selene_sdk.samplers.</code><code class="sig-name descname">OnlineSampler</code><span class="sig-paren">(</span><em class="sig-param">reference_sequence, target_path, features, seed=436, validation_holdout=['chr6', 'chr7'], test_holdout=['chr8', 'chr9'], sequence_length=1001, center_bin_to_predict=201, feature_thresholds=0.5, mode='train', save_datasets=[], output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.sampler.Sampler</span></code></p>
<p>A sampler in which training/validation/test data is constructed
from random sampling of the dataset for each batch passed to the
model. This form of sampling may alleviate the problem of loading an
extremely large dataset into memory when developing a new model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference_sequence</strong> (<em>selene_sdk.sequences.Sequence</em>) – A reference sequence from which to create examples.</p></li>
<li><p><strong>target_path</strong> (<em>str</em>) – Path to tabix-indexed, compressed BED file (<cite>*.bed.gz</cite>) of genomic
coordinates mapped to the genomic features we want to predict.</p></li>
<li><p><strong>features</strong> (<em>list(str)</em>) – List of distinct features that we aim to predict.</p></li>
<li><p><strong>seed</strong> (<em>int, optional</em>) – Default is 436. Sets the random seed for sampling.</p></li>
<li><p><strong>validation_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr6’, ‘chr7’]</cite>. Holdout can be regional or
proportional. If regional, expects a list (e.g. <cite>[‘X’, ‘Y’]</cite>).
Regions must match those specified in the first column of the
tabix-indexed BED file. If proportional, specify a percentage
between (0.0, 1.0). Typically 0.10 or 0.20.</p></li>
<li><p><strong>test_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr8’, ‘chr9’]</cite>. See documentation for
<cite>validation_holdout</cite> for additional information.</p></li>
<li><p><strong>sequence_length</strong> (<em>int, optional</em>) – Default is 1000. Model is trained on sequences of <cite>sequence_length</cite>
where genomic features are annotated to the center regions of
these sequences.</p></li>
<li><p><strong>center_bin_to_predict</strong> (<em>int, optional</em>) – Default is 200. Query the tabix-indexed file for a region of
length <cite>center_bin_to_predict</cite>.</p></li>
<li><p><strong>feature_thresholds</strong> (<em>float [0.0, 1.0], optional</em>) – Default is 0.5. The <cite>feature_threshold</cite> to pass to the
<cite>GenomicFeatures</cite> object.</p></li>
<li><p><strong>mode</strong> (<em>{‘train’, ‘validate’, ‘test’}, optional</em>) – Default is <cite>‘train’</cite>. The mode to run the sampler in.</p></li>
<li><p><strong>save_datasets</strong> (<em>list(str), optional</em>) – Default is <cite>[]</cite> the empty list. The list of modes for which we should
save the sampled data to file (e.g. <cite>[“test”, “validate”]</cite>).</p></li>
<li><p><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. The path to the directory where we should
save sampled examples for a mode. If <cite>save_datasets</cite> is
a non-empty list, <cite>output_dir</cite> must be specified. If
the path in <cite>output_dir</cite> does not exist it will be created
automatically.</p></li>
</ul>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>~OnlineSampler.reference_sequence</strong> (<a class="reference internal" href="sequences.html#selene_sdk.sequences.Sequence" title="selene_sdk.sequences.Sequence"><em>selene_sdk.sequences.Sequence</em></a>) – The reference sequence that examples are created from.</p></li>
<li><p><strong>~OnlineSampler.target</strong> (<a class="reference internal" href="targets.html#selene_sdk.targets.Target" title="selene_sdk.targets.Target"><em>selene_sdk.targets.Target</em></a>) – The <cite>selene_sdk.targets.Target</cite> object holding the features that we
would like to predict.</p></li>
<li><p><strong>~OnlineSampler.validation_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – The samples to hold out for validating model performance. These
can be “regional” or “proportional”. If regional, this is a list
of region names (e.g. <cite>[‘chrX’, ‘chrY’]</cite>). These regions must
match those specified in the first column of the tabix-indexed
BED file. If proportional, this is the fraction of total samples
that will be held out.</p></li>
<li><p><strong>~OnlineSampler.test_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – The samples to hold out for testing model performance. See the
documentation for <cite>validation_holdout</cite> for more details.</p></li>
<li><p><strong>~OnlineSampler.sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The length of the sequences to  train the model on.</p></li>
<li><p><strong>~OnlineSampler.bin_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – From the center of the sequence, the radius in which to detect
a feature annotation in order to include it as a sample’s label.</p></li>
<li><p><strong>~OnlineSampler.surrounding_sequence_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The length of sequence falling outside of the feature detection
bin (i.e. <cite>bin_radius</cite>) center, but still within the
<cite>sequence_length</cite>.</p></li>
<li><p><strong>~OnlineSampler.modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>)</em>) – The list of modes that the sampler can be run in.</p></li>
<li><p><strong>~OnlineSampler.mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The current mode that the sampler is running in. Must be one of
the modes listed in <cite>modes</cite>.</p></li>
</ul>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If <cite>mode</cite> is not a valid mode.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If the parities of <cite>sequence_length</cite> and <cite>center_bin_to_predict</cite>
    are not the same.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If <cite>sequence_length</cite> is smaller than <cite>center_bin_to_predict</cite> is.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If the types of <cite>validation_holdout</cite> and <cite>test_holdout</cite> are not
    the same.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="selene_sdk.samplers.OnlineSampler.STRAND_SIDES">
<code class="sig-name descname">STRAND_SIDES</code><em class="property"> = ('+', '-')</em><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.STRAND_SIDES" title="Permalink to this definition">¶</a></dt>
<dd><p>Defines the strands that features can be sampled from.</p>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_data_and_targets">
<code class="sig-name descname">get_data_and_targets</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em>, <em class="sig-param">mode=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_data_and_targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_data_and_targets" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the data from the sampler,
divided into batches. This method also allows the user to
specify what operating mode to run the sampler in when fetching
the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is None. The total number of samples to retrieve.
If <cite>n_samples</cite> is None and the mode is <cite>validate</cite>, will
set <cite>n_samples</cite> to 32000; if the mode is <cite>test</cite>, will set
<cite>n_samples</cite> to 640000 if it is None. If the mode is <cite>train</cite>
you must have specified a value for <cite>n_samples</cite>.</p></li>
<li><p><strong>mode</strong> (<em>str, optional</em>) – Default is None. The mode to run the sampler in when
fetching the samples. See
<cite>selene_sdk.samplers.IntervalsSampler.modes</cite> for more
information. If None, will use the current mode <cite>self.mode</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_dataset_in_batches">
<code class="sig-name descname">get_dataset_in_batches</code><span class="sig-paren">(</span><em class="sig-param">mode</em>, <em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_dataset_in_batches"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_dataset_in_batches" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of the data for a specified run
mode, divided into mini-batches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> (<em>{‘test’, ‘validate’}</em>) – The mode to run the sampler in when fetching the samples.
See <cite>selene_sdk.samplers.IntervalsSampler.modes</cite> for more
information.</p></li>
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of samples to retrieve.
If <cite>None</cite>, it will retrieve 32000 samples if <cite>mode</cite> is validate
or 640000 samples if <cite>mode</cite> is test or train.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
The list is length <span class="math notranslate nohighlight">\(S\)</span>, where <span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_feature_from_index">
<code class="sig-name descname">get_feature_from_index</code><span class="sig-paren">(</span><em class="sig-param">index</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_feature_from_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_feature_from_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the feature corresponding to an index in the feature
vector.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<em>int</em>) – The index of the feature to retrieve the name for.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The name of the feature occurring at the specified index.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_sequence_from_encoding">
<code class="sig-name descname">get_sequence_from_encoding</code><span class="sig-paren">(</span><em class="sig-param">encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_sequence_from_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_sequence_from_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the string sequence from the one-hot encoding
of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>encoding</strong> (<em>numpy.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span> is the length
of the sequence and <span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence
type’s alphabet) containing the one-hot encoding of the
sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sequence of <span class="math notranslate nohighlight">\(L\)</span> characters decoded from the input.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_test_set">
<code class="sig-name descname">get_test_set</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_test_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_test_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of testing data from the
sampler, divided into batches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of validation examples
to retrieve. If <cite>None</cite>, 640000 examples are retrieved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If no test partition of the data was specified during
    sampler initialization.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_validation_set">
<code class="sig-name descname">get_validation_set</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_validation_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_validation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of validation data from the
sampler, divided into batches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of validation examples
to retrieve. If <cite>None</cite>, 32000 examples are retrieved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.save_dataset_to_file">
<code class="sig-name descname">save_dataset_to_file</code><span class="sig-paren">(</span><em class="sig-param">mode</em>, <em class="sig-param">close_filehandle=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.save_dataset_to_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.save_dataset_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Save samples for each partition (i.e. train/validate/test) to
disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> (<em>str</em>) – Must be one of the modes specified in <cite>save_datasets</cite> during
sampler initialization.</p></li>
<li><p><strong>close_filehandle</strong> (<em>bool, optional</em>) – Default is False. <cite>close_filehandle=True</cite> assumes that all
data corresponding to the input <cite>mode</cite> has been saved to
file and <cite>save_dataset_to_file</cite> will not be called with
<cite>mode</cite> again.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="intervalssampler">
<h2>IntervalsSampler<a class="headerlink" href="#intervalssampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.IntervalsSampler">
<em class="property">class </em><code class="sig-prename descclassname">selene_sdk.samplers.</code><code class="sig-name descname">IntervalsSampler</code><span class="sig-paren">(</span><em class="sig-param">reference_sequence, target_path, features, intervals_path, sample_negative=False, seed=436, validation_holdout=['chr6', 'chr7'], test_holdout=['chr8', 'chr9'], sequence_length=1000, center_bin_to_predict=200, feature_thresholds=0.5, mode='train', save_datasets=['test'], output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/intervals_sampler.html#IntervalsSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.IntervalsSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.online_sampler.OnlineSampler</span></code></p>
<p>Draws samples from pre-specified windows in the reference sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference_sequence</strong> (<em>selene_sdk.sequences.Sequence</em>) – A reference sequence from which to create examples.</p></li>
<li><p><strong>target_path</strong> (<em>str</em>) – Path to tabix-indexed, compressed BED file (<cite>*.bed.gz</cite>) of genomic
coordinates mapped to the genomic features we want to predict.</p></li>
<li><p><strong>features</strong> (<em>list(str)</em>) – List of distinct features that we aim to predict.</p></li>
<li><p><strong>intervals_path</strong> (<em>str</em>) – The path to the file that contains the intervals to sample from.
In this file, each interval should occur on a separate line.</p></li>
<li><p><strong>sample_negative</strong> (<em>bool, optional</em>) – Default is <cite>False</cite>. This tells the sampler whether negative
examples (i.e. with no positive labels) should be drawn when
generating samples. If <cite>True</cite>, both negative and positive
samples will be drawn. If <cite>False</cite>, only samples with at least
one positive label will be drawn.</p></li>
<li><p><strong>seed</strong> (<em>int, optional</em>) – Default is 436. Sets the random seed for sampling.</p></li>
<li><p><strong>validation_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr6’, ‘chr7’]</cite>. Holdout can be regional or
proportional. If regional, expects a list (e.g. <cite>[‘X’, ‘Y’]</cite>).
Regions must match those specified in the first column of the
tabix-indexed BED file. If proportional, specify a percentage
between (0.0, 1.0). Typically 0.10 or 0.20.</p></li>
<li><p><strong>test_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr8’, ‘chr9’]</cite>. See documentation for
<cite>validation_holdout</cite> for additional information.</p></li>
<li><p><strong>sequence_length</strong> (<em>int, optional</em>) – Default is 1000. Model is trained on sequences of <cite>sequence_length</cite>
where genomic features are annotated to the center regions of
these sequences.</p></li>
<li><p><strong>center_bin_to_predict</strong> (<em>int, optional</em>) – Default is 200. Query the tabix-indexed file for a region of
length <cite>center_bin_to_predict</cite>.</p></li>
<li><p><strong>feature_thresholds</strong> (<em>float [0.0, 1.0] or None, optional</em>) –  Default is 0.5. The <cite>feature_threshold</cite> to pass to the
<cite>GenomicFeatures</cite> object.</p></li>
<li><p><strong>mode</strong> (<em>{‘train’, ‘validate’, ‘test’}</em>) – Default is <cite>‘train’</cite>. The mode to run the sampler in.</p></li>
<li><p><strong>save_datasets</strong> (<em>list of str</em>) – Default is <cite>[“test”]</cite>. The list of modes for which we should
save the sampled data to file.</p></li>
<li><p><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. The path to the directory where we should
save sampled examples for a mode. If <cite>save_datasets</cite> is
a non-empty list, <cite>output_dir</cite> must be specified. If
the path in <cite>output_dir</cite> does not exist it will be created
automatically.</p></li>
</ul>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>~IntervalsSampler.reference_sequence</strong> (<a class="reference internal" href="sequences.html#selene_sdk.sequences.Sequence" title="selene_sdk.sequences.Sequence"><em>selene_sdk.sequences.Sequence</em></a>) – The reference sequence that examples are created from.</p></li>
<li><p><strong>~IntervalsSampler.target</strong> (<a class="reference internal" href="targets.html#selene_sdk.targets.Target" title="selene_sdk.targets.Target"><em>selene_sdk.targets.Target</em></a>) – The <cite>selene_sdk.targets.Target</cite> object holding the features that we
would like to predict.</p></li>
<li><p><strong>~IntervalsSampler.sample_from_intervals</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>)</em><em>)</em>) – A list of coordinates that specify the intervals we can draw
samples from.</p></li>
<li><p><strong>~IntervalsSampler.interval_lengths</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>)</em>) – A list of the lengths of the intervals that we can draw samples
from. The probability that we will draw a sample from an
interval is a function of that interval’s length and the length
of all other intervals.</p></li>
<li><p><strong>~IntervalsSampler.sample_negative</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Whether negative examples (i.e. with no positive label) should
be drawn when generating samples. If <cite>True</cite>, both negative and
positive samples will be drawn. If <cite>False</cite>, only samples with at
least one positive label will be drawn.</p></li>
<li><p><strong>~IntervalsSampler.validation_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – The samples to hold out for validating model performance. These
can be “regional” or “proportional”. If regional, this is a list
of region names (e.g. <cite>[‘chrX’, ‘chrY’]</cite>). These Regions must
match those specified in the first column of the tabix-indexed
BED file. If proportional, this is the fraction of total samples
that will be held out.</p></li>
<li><p><strong>~IntervalsSampler.test_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – The samples to hold out for testing model performance. See the
documentation for <cite>validation_holdout</cite> for more details.</p></li>
<li><p><strong>~IntervalsSampler.sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The length of the sequences to  train the model on.</p></li>
<li><p><strong>~IntervalsSampler.bin_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – From the center of the sequence, the radius in which to detect
a feature annotation in order to include it as a sample’s label.</p></li>
<li><p><strong>~IntervalsSampler.surrounding_sequence_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The length of sequence falling outside of the feature detection
bin (i.e. <cite>bin_radius</cite>) center, but still within the
<cite>sequence_length</cite>.</p></li>
<li><p><strong>~IntervalsSampler.modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>)</em>) – The list of modes that the sampler can be run in.</p></li>
<li><p><strong>~IntervalsSampler.mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The current mode that the sampler is running in. Must be one of
the modes listed in <cite>modes</cite>.</p></li>
</ul>
</dd>
</dl>
<dl class="method">
<dt id="selene_sdk.samplers.IntervalsSampler.sample">
<code class="sig-name descname">sample</code><span class="sig-paren">(</span><em class="sig-param">batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/intervals_sampler.html#IntervalsSampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.IntervalsSampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Randomly draws a mini-batch of examples and their corresponding
labels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The number of examples to include in the
mini-batch.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences, targets</strong> – A tuple containing the numeric representation of the
sequence examples and their corresponding labels. The
shape of <cite>sequences</cite> will be
<span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span> is
<cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length, and
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.
The shape of <cite>targets</cite> will be <span class="math notranslate nohighlight">\(B \times F\)</span>,
where <span class="math notranslate nohighlight">\(F\)</span> is the number of features.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="randompositionssampler">
<h2>RandomPositionsSampler<a class="headerlink" href="#randompositionssampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.RandomPositionsSampler">
<em class="property">class </em><code class="sig-prename descclassname">selene_sdk.samplers.</code><code class="sig-name descname">RandomPositionsSampler</code><span class="sig-paren">(</span><em class="sig-param">reference_sequence, target_path, features, seed=436, validation_holdout=['chr6', 'chr7'], test_holdout=['chr8', 'chr9'], sequence_length=1000, center_bin_to_predict=200, feature_thresholds=0.5, mode='train', save_datasets=[], output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/random_positions_sampler.html#RandomPositionsSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.RandomPositionsSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.online_sampler.OnlineSampler</span></code></p>
<p>This sampler randomly selects a position in the genome and queries for
a sequence centered at that position for input to the model.</p>
<p>TODO: generalize to selene_sdk.sequences.Sequence?</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference_sequence</strong> (<em>selene_sdk.sequences.Genome</em>) – A reference sequence from which to create examples.</p></li>
<li><p><strong>target_path</strong> (<em>str</em>) – Path to tabix-indexed, compressed BED file (<cite>*.bed.gz</cite>) of genomic
coordinates mapped to the genomic features we want to predict.</p></li>
<li><p><strong>features</strong> (<em>list(str)</em>) – List of distinct features that we aim to predict.</p></li>
<li><p><strong>seed</strong> (<em>int, optional</em>) – Default is 436. Sets the random seed for sampling.</p></li>
<li><p><strong>validation_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr6’, ‘chr7’]</cite>. Holdout can be regional or
proportional. If regional, expects a list (e.g. <cite>[‘chrX’, ‘chrY’]</cite>).
Regions must match those specified in the first column of the
tabix-indexed BED file. If proportional, specify a percentage
between (0.0, 1.0). Typically 0.10 or 0.20.</p></li>
<li><p><strong>test_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr8’, ‘chr9’]</cite>. See documentation for
<cite>validation_holdout</cite> for additional information.</p></li>
<li><p><strong>sequence_length</strong> (<em>int, optional</em>) – Default is 1000. Model is trained on sequences of <cite>sequence_length</cite>
where genomic features are annotated to the center regions of
these sequences.</p></li>
<li><p><strong>center_bin_to_predict</strong> (<em>int, optional</em>) – Default is 200. Query the tabix-indexed file for a region of
length <cite>center_bin_to_predict</cite>.</p></li>
<li><p><strong>feature_thresholds</strong> (<em>float [0.0, 1.0], optional</em>) – Default is 0.5. The <cite>feature_threshold</cite> to pass to the
<cite>GenomicFeatures</cite> object.</p></li>
<li><p><strong>mode</strong> (<em>{‘train’, ‘validate’, ‘test’}</em>) – Default is <cite>‘train’</cite>. The mode to run the sampler in.</p></li>
<li><p><strong>save_datasets</strong> (<em>list(str), optional</em>) – Default is <cite>[‘test’]</cite>. The list of modes for which we should
save the sampled data to file.</p></li>
<li><p><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. The path to the directory where we should
save sampled examples for a mode. If <cite>save_datasets</cite> is
a non-empty list, <cite>output_dir</cite> must be specified. If
the path in <cite>output_dir</cite> does not exist it will be created
automatically.</p></li>
</ul>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>~RandomPositionsSampler.reference_sequence</strong> (<a class="reference internal" href="sequences.html#selene_sdk.sequences.Genome" title="selene_sdk.sequences.Genome"><em>selene_sdk.sequences.Genome</em></a>) – The reference sequence that examples are created from.</p></li>
<li><p><strong>~RandomPositionsSampler.target</strong> (<a class="reference internal" href="targets.html#selene_sdk.targets.Target" title="selene_sdk.targets.Target"><em>selene_sdk.targets.Target</em></a>) – The <cite>selene_sdk.targets.Target</cite> object holding the features that we
would like to predict.</p></li>
<li><p><strong>~RandomPositionsSampler.validation_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – The samples to hold out for validating model performance. These
can be “regional” or “proportional”. If regional, this is a list
of region names (e.g. <cite>[‘chrX’, ‘chrY’]</cite>). These regions must
match those specified in the first column of the tabix-indexed
BED file. If proportional, this is the fraction of total samples
that will be held out.</p></li>
<li><p><strong>~RandomPositionsSampler.test_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – The samples to hold out for testing model performance. See the
documentation for <cite>validation_holdout</cite> for more details.</p></li>
<li><p><strong>~RandomPositionsSampler.sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The length of the sequences to  train the model on.</p></li>
<li><p><strong>~RandomPositionsSampler.bin_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – From the center of the sequence, the radius in which to detect
a feature annotation in order to include it as a sample’s label.</p></li>
<li><p><strong>~RandomPositionsSampler.surrounding_sequence_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The length of sequence falling outside of the feature detection
bin (i.e. <cite>bin_radius</cite>) center, but still within the
<cite>sequence_length</cite>.</p></li>
<li><p><strong>~RandomPositionsSampler.modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>)</em>) – The list of modes that the sampler can be run in.</p></li>
<li><p><strong>~RandomPositionsSampler.mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The current mode that the sampler is running in. Must be one of
the modes listed in <cite>modes</cite>.</p></li>
</ul>
</dd>
</dl>
<dl class="method">
<dt id="selene_sdk.samplers.RandomPositionsSampler.sample">
<code class="sig-name descname">sample</code><span class="sig-paren">(</span><em class="sig-param">batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/random_positions_sampler.html#RandomPositionsSampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.RandomPositionsSampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Randomly draws a mini-batch of examples and their corresponding
labels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The number of examples to include in the
mini-batch.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences, targets</strong> – A tuple containing the numeric representation of the
sequence examples and their corresponding labels. The
shape of <cite>sequences</cite> will be
<span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span> is
<cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length, and
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.
The shape of <cite>targets</cite> will be <span class="math notranslate nohighlight">\(B \times F\)</span>,
where <span class="math notranslate nohighlight">\(F\)</span> is the number of features.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="multifilesampler">
<h2>MultiFileSampler<a class="headerlink" href="#multifilesampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.MultiFileSampler">
<em class="property">class </em><code class="sig-prename descclassname">selene_sdk.samplers.</code><code class="sig-name descname">MultiFileSampler</code><span class="sig-paren">(</span><em class="sig-param">train_sampler</em>, <em class="sig-param">validate_sampler</em>, <em class="sig-param">features</em>, <em class="sig-param">test_sampler=None</em>, <em class="sig-param">mode='train'</em>, <em class="sig-param">save_datasets=[]</em>, <em class="sig-param">output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.sampler.Sampler</span></code></p>
<p>This sampler contains individual file samplers for each mode.
The file samplers parse .bed/.mat files that correspond to
training, validation, and testing and MultiFileSampler calls on
the correct file sampler to draw samples for a given mode.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>~MultiFileSampler.train_sampler</strong> (<em>selene_sdk.samplers.file_samplers.FileSampler</em>) – Load your training data as a <cite>FileSampler</cite> before passing it
into the <cite>MultiFileSampler</cite> constructor.</p></li>
<li><p><strong>~MultiFileSampler.validate_sampler</strong> (<em>selene_sdk.samplers.file_samplers.FileSampler</em>) – The validation dataset file sampler.</p></li>
<li><p><strong>~MultiFileSampler.features</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>)</em>) – The list of features the model should predict</p></li>
<li><p><strong>~MultiFileSampler.test_sampler</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a><em> or </em><em>selene_sdk.samplers.file_samplers.FileSampler</em><em>, </em><em>optional</em>) – Default is None. The test file sampler is optional.</p></li>
<li><p><strong>~MultiFileSampler.mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – Default is “train”. Must be one of <cite>{train, validate, test}</cite>. The
starting mode in which to run the sampler.</p></li>
<li><p><strong>~MultiFileSampler.save_datasets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>)</em><em>, </em><em>optional</em>) – Default is None. Currently, we are only including these parameters
so that <cite>MultiFileSampler</cite> is consistent with <cite>Sampler</cite>. The save
dataset functionality for MultiFileSampler has not been defined
yet.</p></li>
<li><p><strong>~MultiFileSampler.output_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a><em>, </em><em>optional</em>) – Default is None. Used if the sampler has any data or logging
statements to save to file. Currently not useful for
<cite>MultiFileSampler</cite>.</p></li>
<li><p><strong>~MultiFileSampler.modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>)</em>) – A list of the names of the modes that the object may operate in.</p></li>
<li><p><strong>~MultiFileSampler.mode</strong> – Default is <cite>None</cite>. The current mode that the object is operating in.</p></li>
</ul>
</dd>
</dl>
<dl class="method">
<dt id="selene_sdk.samplers.MultiFileSampler.get_data_and_targets">
<code class="sig-name descname">get_data_and_targets</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples</em>, <em class="sig-param">mode=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler.get_data_and_targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler.get_data_and_targets" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the data from the sampler,
divided into batches. This method also allows the user to
specify what operating mode to run the sampler in when fetching
the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int</em>) – The total number of samples to retrieve.</p></li>
<li><p><strong>mode</strong> (<em>str, optional</em>) – Default is None. The operating mode that the sampler
should run in. If None, will use the current
<cite>self.mode</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MultiFileSampler.get_feature_from_index">
<code class="sig-name descname">get_feature_from_index</code><span class="sig-paren">(</span><em class="sig-param">index</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler.get_feature_from_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler.get_feature_from_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the feature corresponding to an index in the feature
vector.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<em>int</em>) – The index of the feature to retrieve the name for.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The name of the feature occurring at the specified index.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MultiFileSampler.get_test_set">
<code class="sig-name descname">get_test_set</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler.get_test_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler.get_test_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of testing data from the
sampler, divided into batches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of validation examples
to retrieve. If <cite>None</cite>, 640000 examples are retrieved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.17)">numpy.ndarray</a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If no test partition of the data was specified during
    sampler initialization.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MultiFileSampler.get_validation_set">
<code class="sig-name descname">get_validation_set</code><span class="sig-paren">(</span><em class="sig-param">batch_size</em>, <em class="sig-param">n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler.get_validation_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler.get_validation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of validation data from the
sampler, divided into batches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</p></li>
<li><p><strong>n_samples</strong> (<em>int, optional</em>) – Default is None. The total number of validation examples to
retrieve. Handling for <cite>n_samples=None</cite> should be done by
all classes that subclass <cite>selene_sdk.samplers.Sampler</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MultiFileSampler.sample">
<code class="sig-name descname">sample</code><span class="sig-paren">(</span><em class="sig-param">batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches a mini-batch of the data from the sampler.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The size of the batch to retrieve.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MultiFileSampler.save_dataset_to_file">
<code class="sig-name descname">save_dataset_to_file</code><span class="sig-paren">(</span><em class="sig-param">mode</em>, <em class="sig-param">close_filehandle=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler.save_dataset_to_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler.save_dataset_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>We implement this function in this class only because the
TrainModel class calls this method. In the future, we will
likely remove this method or implement a different way
of “saving the data” for file samplers. For example, we
may only output the row numbers sampled so that users may
reproduce exactly what order the data was sampled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> (<em>str</em>) – Must be one of the modes specified in <cite>save_datasets</cite> during
sampler initialization.</p></li>
<li><p><strong>close_filehandle</strong> (<em>bool, optional</em>) – Default is False. <cite>close_filehandle=True</cite> assumes that all
data corresponding to the input <cite>mode</cite> has been saved to
file and <cite>save_dataset_to_file</cite> will not be called with
<cite>mode</cite> again.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MultiFileSampler.set_mode">
<code class="sig-name descname">set_mode</code><span class="sig-paren">(</span><em class="sig-param">mode</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/multi_file_sampler.html#MultiFileSampler.set_mode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MultiFileSampler.set_mode" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the sampling mode.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mode</strong> (<em>str</em>) – The name of the mode to use. It must be one of
<cite>Sampler.BASE_MODES</cite> (“train”, “validate”) or “test” if
the test data is supplied.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If <cite>mode</cite> is not a valid mode.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="samplers.file_samplers.html" class="btn btn-neutral float-right" title="selene_sdk.samplers.file_samplers" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="predict.html" class="btn btn-neutral float-left" title="selene_sdk.predict" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>