

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>selene_sdk.sequences.genome &mdash; Selene 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-selene-with-pip">Installing selene with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene_sdk.sequences.genome</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene_sdk.sequences.genome</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module provides the `Genome` class. This class wraps the indexed</span>
<span class="sd">FASTA file for an organism&#39;s genomic sequence. It supports retrieving</span>
<span class="sd">parts of the sequence and converting these parts into their one-hot</span>
<span class="sd">encodings.</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">pkg_resources</span>
<span class="kn">import</span> <span class="nn">pyfaidx</span>
<span class="kn">import</span> <span class="nn">tabix</span>

<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">wraps</span>
<span class="kn">from</span> <span class="nn">.sequence</span> <span class="kn">import</span> <span class="n">Sequence</span>
<span class="kn">from</span> <span class="nn">.sequence</span> <span class="kn">import</span> <span class="n">sequence_to_encoding</span>
<span class="kn">from</span> <span class="nn">.sequence</span> <span class="kn">import</span> <span class="n">encoding_to_sequence</span>

<span class="k">def</span> <span class="nf">_not_blacklist_region</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">blacklist_tabix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if the input coordinates are not overlapping with blacklist regions.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    chrom : str</span>
<span class="sd">        The name of the chromosomes, e.g. &quot;chr1&quot;.</span>
<span class="sd">    start : int</span>
<span class="sd">        The 0-based start coordinate of the sequence.</span>
<span class="sd">    end : int</span>
<span class="sd">        One past the last coordinate of the sequence.</span>
<span class="sd">    blacklist_tabix : tabix.open or None, optional</span>
<span class="sd">        Default is `None`. Tabix file handle if a file of blacklist regions</span>
<span class="sd">        is available.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool</span>
<span class="sd">        False if the coordinates are overlaping with blacklist regions</span>
<span class="sd">        (if specified). Otherwise, return True.</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">blacklist_tabix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">rows</span> <span class="o">=</span> <span class="n">blacklist_tabix</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
        <span class="k">except</span> <span class="n">tabix</span><span class="o">.</span><span class="n">TabixError</span><span class="p">:</span>
            <span class="k">pass</span>
    <span class="k">return</span> <span class="kc">True</span>


<span class="k">def</span> <span class="nf">_check_coords</span><span class="p">(</span><span class="n">len_chrs</span><span class="p">,</span>
                  <span class="n">chrom</span><span class="p">,</span>
                  <span class="n">start</span><span class="p">,</span>
                  <span class="n">end</span><span class="p">,</span>
                  <span class="n">pad</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                  <span class="n">blacklist_tabix</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if the input coordinates are valid.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    len_chrs : dict</span>
<span class="sd">        A dictionary mapping chromosome names to lengths.</span>
<span class="sd">    chrom : str</span>
<span class="sd">        The name of the chromosomes, e.g. &quot;chr1&quot;.</span>
<span class="sd">    start : int</span>
<span class="sd">        The 0-based start coordinate of the sequence.</span>
<span class="sd">    end : int</span>
<span class="sd">        One past the last coordinate of the sequence.</span>
<span class="sd">    pad : bool, optional</span>
<span class="sd">        Default is `False`. Allow coordinates that are partially</span>
<span class="sd">        out of bounds.</span>
<span class="sd">    blacklist_tabix : tabix.open or None, optional</span>
<span class="sd">        Default is `None`. Tabix file handle if a file of blacklist regions</span>
<span class="sd">        is available.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool</span>
<span class="sd">        True if the coordinates are valid (`start` and `end` are within</span>
<span class="sd">        chromosome boundaries and not overlaping with blacklist regions</span>
<span class="sd">        (if specified). Otherwise, return False.</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">len_chrs</span> <span class="ow">and</span> \
         <span class="n">start</span> <span class="o">&lt;</span> <span class="n">len_chrs</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="ow">and</span> \
         <span class="n">start</span> <span class="o">&lt;</span> <span class="n">end</span> <span class="ow">and</span> \
         <span class="n">end</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> \
         <span class="p">(</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">pad</span> <span class="k">else</span> <span class="kc">True</span><span class="p">)</span> <span class="ow">and</span> \
         <span class="p">(</span><span class="n">end</span> <span class="o">&lt;=</span> <span class="n">len_chrs</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">pad</span> <span class="k">else</span> <span class="kc">True</span><span class="p">)</span> <span class="ow">and</span> \
         <span class="n">_not_blacklist_region</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">blacklist_tabix</span><span class="p">)</span>



<span class="k">def</span> <span class="nf">_get_sequence_from_coords</span><span class="p">(</span><span class="n">len_chrs</span><span class="p">,</span>
                              <span class="n">genome_sequence</span><span class="p">,</span>
                              <span class="n">chrom</span><span class="p">,</span>
                              <span class="n">start</span><span class="p">,</span>
                              <span class="n">end</span><span class="p">,</span>
                              <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span>
                              <span class="n">pad</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                              <span class="n">blacklist_tabix</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Gets the genomic sequence at the input coordinates.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    len_chrs : dict</span>
<span class="sd">        A dictionary mapping chromosome names to lengths.</span>
<span class="sd">    genome_sequence : function</span>
<span class="sd">        A closure that extracts a sequence from a genome.</span>
<span class="sd">    chrom : str</span>
<span class="sd">        The name of the chromosomes, e.g. &quot;chr1&quot;.</span>
<span class="sd">    start : int</span>
<span class="sd">        The 0-based start coordinate of the sequence.</span>
<span class="sd">    end : int</span>
<span class="sd">        One past the last coordinate of the sequence.</span>
<span class="sd">    strand : {&#39;+&#39;, &#39;-&#39;, &#39;.&#39;}, optional</span>
<span class="sd">        Default is &#39;+&#39;. The strand the sequence is located on. &#39;.&#39; is treated</span>
<span class="sd">        as &#39;+&#39;.</span>
<span class="sd">    pad : bool, optional</span>
<span class="sd">        Default is `False`. If the coordinates are out of bounds, make an</span>
<span class="sd">        in-bounds query and then pad the sequence to return the desired</span>
<span class="sd">        sequence length.</span>
<span class="sd">    blacklist_tabix : tabix.open or None, optional</span>
<span class="sd">        Default is `None`. Tabix file handle if a file of blacklist regions</span>
<span class="sd">        is available.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        The genomic sequence occurring at the input coordinates.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the input char to `strand` is not one of the specified</span>
<span class="sd">        choices.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">_check_coords</span><span class="p">(</span><span class="n">len_chrs</span><span class="p">,</span>
        <span class="n">chrom</span><span class="p">,</span>
        <span class="n">start</span><span class="p">,</span>
        <span class="n">end</span><span class="p">,</span>
        <span class="n">pad</span><span class="o">=</span><span class="n">pad</span><span class="p">,</span>
        <span class="n">blacklist_tabix</span><span class="o">=</span><span class="n">blacklist_tabix</span><span class="p">):</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>

    <span class="k">if</span> <span class="n">strand</span> <span class="o">!=</span> <span class="s1">&#39;+&#39;</span> <span class="ow">and</span> <span class="n">strand</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span> <span class="ow">and</span> <span class="n">strand</span> <span class="o">!=</span> <span class="s1">&#39;.&#39;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Strand must be one of &#39;+&#39;, &#39;-&#39;, or &#39;.&#39;. Input was </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">strand</span><span class="p">))</span>

    <span class="n">end_pad</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">start_pad</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">len_chrs</span><span class="p">[</span><span class="n">chrom</span><span class="p">]:</span>
        <span class="n">end_pad</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="n">len_chrs</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">len_chrs</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">start_pad</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">start</span>
        <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">Genome</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">start_pad</span> <span class="o">+</span>
            <span class="n">genome_sequence</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="o">+</span>
            <span class="n">Genome</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">end_pad</span><span class="p">)</span>


<div class="viewcode-block" id="Genome"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome">[docs]</a><span class="k">class</span> <span class="nc">Genome</span><span class="p">(</span><span class="n">Sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;This class provides access to an organism&#39;s genomic sequence.</span>

<span class="sd">    This class supports retrieving parts of the sequence and converting</span>
<span class="sd">    these parts into their one-hot encodings. It is essentially a</span>
<span class="sd">    wrapper class around the `pyfaidx.Fasta` class.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input_path : str</span>
<span class="sd">        Path to an indexed FASTA file, that is, a `*.fasta` file with</span>
<span class="sd">        a corresponding `*.fai` file in the same directory. This file</span>
<span class="sd">        should contain the target organism&#39;s genome sequence.</span>
<span class="sd">    blacklist_regions : str or None, optional</span>
<span class="sd">        Default is None. Path to a tabix-indexed list of regions from</span>
<span class="sd">        which we should not output sequences. This is used to ensure that</span>
<span class="sd">        we are not sampling from areas where we will never collect</span>
<span class="sd">        measurements. You can pass as input &quot;hg19&quot; or &quot;hg38&quot; to use the</span>
<span class="sd">        blacklist regions released by ENCODE. You can also pass in your own</span>
<span class="sd">        tabix-indexed .gz file.</span>
<span class="sd">    bases_order : list(str) or None, optional</span>
<span class="sd">        Default is None (use the default base ordering of</span>
<span class="sd">        `[&#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;]`). Specify a different ordering of</span>
<span class="sd">        DNA bases for one-hot encoding.</span>
<span class="sd">    init_unpicklable : bool, optional</span>
<span class="sd">        Default is False. Delays initialization until a relevant method</span>
<span class="sd">        is called. This enables the object to be pickled after instantiation.</span>
<span class="sd">        `init_unpicklable` must be `False` when multi-processing is needed e.g.</span>
<span class="sd">        DataLoader. Set `init_unpicklable` to True if you are using this class</span>
<span class="sd">        directly through Selene&#39;s API and want to access class attributes</span>
<span class="sd">        without having to call on a specific method in Genome.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    genome : pyfaidx.Fasta</span>
<span class="sd">        The FASTA file containing the genome sequence.</span>
<span class="sd">    chrs : list(str)</span>
<span class="sd">        The list of chromosome names.</span>
<span class="sd">    len_chrs : dict</span>
<span class="sd">        A dictionary mapping the names of each chromosome in the file to</span>
<span class="sd">        the length of said chromosome.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">BASES_ARR</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is an array with the alphabet (i.e. all possible symbols</span>
<span class="sd">    that may occur in a sequence). We expect that</span>
<span class="sd">    `INDEX_TO_BASE[i]==BASES_ARR[i]` is `True` for all valid `i`.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">BASE_TO_INDEX</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
        <span class="s1">&#39;a&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">&#39;t&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A dictionary mapping members of the alphabet (i.e. all</span>
<span class="sd">    possible symbols that can occur in a sequence) to integers.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">INDEX_TO_BASE</span> <span class="o">=</span> <span class="p">{</span>
        <span class="mi">0</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span>
    <span class="p">}</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A dictionary mapping integers to members of the alphabet (i.e.</span>
<span class="sd">    all possible symbols that can occur in a sequence). We expect</span>
<span class="sd">    that `INDEX_TO_BASE[i]==BASES_ARR[i]` is `True` for all</span>
<span class="sd">    valid `i`.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">COMPLEMENTARY_BASE_DICT</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">:</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span>
        <span class="s1">&#39;a&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;t&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;n&#39;</span><span class="p">:</span> <span class="s1">&#39;N&#39;</span>
    <span class="p">}</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A dictionary mapping each base to its complementary base.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">UNK_BASE</span> <span class="o">=</span> <span class="s2">&quot;N&quot;</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is a base used to represent unknown positions. This is not</span>
<span class="sd">    the same as a character from outside the sequence&#39;s alphabet. A</span>
<span class="sd">    character from outside the alphabet is an error. A position with</span>
<span class="sd">    an unknown base signifies that the position is one of the bases</span>
<span class="sd">    from the alphabet, but we are uncertain which.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">,</span> <span class="n">blacklist_regions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bases_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">init_unpicklable</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructs a `Genome` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">=</span> <span class="n">input_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">blacklist_regions</span> <span class="o">=</span> <span class="n">blacklist_regions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_initialized</span> <span class="o">=</span><span class="kc">False</span>

        <span class="k">if</span> <span class="n">bases_order</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">bases</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bases_order</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">BASES_ARR</span> <span class="o">=</span> <span class="n">bases</span>
            <span class="n">lc_bases</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="o">.</span><span class="n">lower</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bases</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">BASE_TO_INDEX</span> <span class="o">=</span> <span class="p">{</span>
                <span class="o">**</span><span class="p">{</span><span class="n">b</span><span class="p">:</span> <span class="n">ix</span> <span class="k">for</span> <span class="p">(</span><span class="n">ix</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bases</span><span class="p">)},</span>
                <span class="o">**</span><span class="p">{</span><span class="n">b</span><span class="p">:</span> <span class="n">ix</span> <span class="k">for</span> <span class="p">(</span><span class="n">ix</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lc_bases</span><span class="p">)}}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">INDEX_TO_BASE</span> <span class="o">=</span> <span class="p">{</span><span class="n">ix</span><span class="p">:</span> <span class="n">b</span> <span class="k">for</span> <span class="p">(</span><span class="n">ix</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bases</span><span class="p">)}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_bases_order</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">init_unpicklable</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_unpicklable_init</span><span class="p">()</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">update_bases_order</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">bases</span><span class="p">):</span>
        <span class="bp">cls</span><span class="o">.</span><span class="n">BASES_ARR</span> <span class="o">=</span> <span class="n">bases</span>
        <span class="n">lc_bases</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="o">.</span><span class="n">lower</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bases</span><span class="p">]</span>
        <span class="bp">cls</span><span class="o">.</span><span class="n">BASE_TO_INDEX</span> <span class="o">=</span> <span class="p">{</span>
            <span class="o">**</span><span class="p">{</span><span class="n">b</span><span class="p">:</span> <span class="n">ix</span> <span class="k">for</span> <span class="p">(</span><span class="n">ix</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bases</span><span class="p">)},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">b</span><span class="p">:</span> <span class="n">ix</span> <span class="k">for</span> <span class="p">(</span><span class="n">ix</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lc_bases</span><span class="p">)}}</span>
        <span class="bp">cls</span><span class="o">.</span><span class="n">INDEX_TO_BASE</span> <span class="o">=</span> <span class="p">{</span><span class="n">ix</span><span class="p">:</span> <span class="n">b</span> <span class="k">for</span> <span class="p">(</span><span class="n">ix</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bases</span><span class="p">)}</span>

    <span class="k">def</span> <span class="nf">_unpicklable_init</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">chrs</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">len_chrs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_len_chrs</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_blacklist_tabix</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">blacklist_regions</span> <span class="o">==</span> <span class="s2">&quot;hg19&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_blacklist_tabix</span> <span class="o">=</span> <span class="n">tabix</span><span class="o">.</span><span class="n">open</span><span class="p">(</span>
                    <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span>
                        <span class="s2">&quot;selene_sdk&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;sequences/data/hg19_blacklist_ENCFF001TDO.bed.gz&quot;</span><span class="p">))</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">blacklist_regions</span> <span class="o">==</span> <span class="s2">&quot;hg38&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_blacklist_tabix</span> <span class="o">=</span> <span class="n">tabix</span><span class="o">.</span><span class="n">open</span><span class="p">(</span>
                    <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span>
                        <span class="s2">&quot;selene_sdk&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;sequences/data/hg38.blacklist.bed.gz&quot;</span><span class="p">))</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">blacklist_regions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># user-specified file</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_blacklist_tabix</span> <span class="o">=</span> <span class="n">tabix</span><span class="o">.</span><span class="n">open</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">blacklist_regions</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_initialized</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">def</span> <span class="nf">init</span><span class="p">(</span><span class="n">func</span><span class="p">):</span>
        <span class="c1"># delay initialization to allow  multiprocessing</span>
        <span class="nd">@wraps</span><span class="p">(</span><span class="n">func</span><span class="p">)</span>
        <span class="k">def</span> <span class="nf">dfunc</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_unpicklable_init</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">dfunc</span>

<div class="viewcode-block" id="Genome.get_chrs"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.get_chrs">[docs]</a>    <span class="nd">@init</span>
    <span class="k">def</span> <span class="nf">get_chrs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets the list of chromosome names.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list(str)</span>
<span class="sd">            A list of the chromosome names.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrs</span></div>

<div class="viewcode-block" id="Genome.get_chr_lens"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.get_chr_lens">[docs]</a>    <span class="nd">@init</span>
    <span class="k">def</span> <span class="nf">get_chr_lens</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets the name and length of each chromosome sequence in the file.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list(tuple(str, int))</span>
<span class="sd">            A list of tuples of the chromosome names and lengths.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[(</span><span class="n">k</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">len_chrs</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chrs</span><span class="p">()]</span></div>

    <span class="k">def</span> <span class="nf">_get_len_chrs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">len_chrs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrs</span><span class="p">:</span>
            <span class="n">len_chrs</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">len_chrs</span>

    <span class="k">def</span> <span class="nf">_genome_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="ow">or</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;.&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">reverse</span><span class="o">.</span><span class="n">complement</span><span class="o">.</span><span class="n">seq</span>

<div class="viewcode-block" id="Genome.coords_in_bounds"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.coords_in_bounds">[docs]</a>    <span class="nd">@init</span>
    <span class="k">def</span> <span class="nf">coords_in_bounds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check if the region we want to query is within the bounds of the</span>
<span class="sd">        queried chromosome and non-overlapping with blacklist regions</span>
<span class="sd">        (if given).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            The name of the chromosomes, e.g. &quot;chr1&quot;.</span>
<span class="sd">        start : int</span>
<span class="sd">            The 0-based start coordinate of the sequence.</span>
<span class="sd">        end : int</span>
<span class="sd">            One past the 0-based last position in the sequence.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bool</span>
<span class="sd">            Whether we can retrieve a sequence from the bounds specified</span>
<span class="sd">            in the input.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">_check_coords</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">len_chrs</span><span class="p">,</span>
                             <span class="n">chrom</span><span class="p">,</span>
                             <span class="n">start</span><span class="p">,</span>
                             <span class="n">end</span><span class="p">,</span>
                             <span class="n">blacklist_tabix</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_blacklist_tabix</span><span class="p">)</span></div>

<div class="viewcode-block" id="Genome.get_sequence_from_coords"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.get_sequence_from_coords">[docs]</a>    <span class="nd">@init</span>
    <span class="k">def</span> <span class="nf">get_sequence_from_coords</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                 <span class="n">chrom</span><span class="p">,</span>
                                 <span class="n">start</span><span class="p">,</span>
                                 <span class="n">end</span><span class="p">,</span>
                                 <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span>
                                 <span class="n">pad</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Gets the queried chromosome&#39;s sequence at the input coordinates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            The name of the chromosomes, e.g. &quot;chr1&quot;.</span>
<span class="sd">        start : int</span>
<span class="sd">            The 0-based start coordinate of the sequence.</span>
<span class="sd">        end : int</span>
<span class="sd">            One past the 0-based last position in the sequence.</span>
<span class="sd">        strand : {&#39;+&#39;, &#39;-&#39;, &#39;.&#39;}, optional</span>
<span class="sd">            Default is &#39;+&#39;. The strand the sequence is located on. &#39;.&#39; is</span>
<span class="sd">            treated as &#39;.&#39;.</span>
<span class="sd">        pad : bool, optional</span>
<span class="sd">            Default is `False`. Pad the output sequence with &#39;N&#39; if `start`</span>
<span class="sd">            and/or `end` are out of bounds to return a sequence of length</span>
<span class="sd">            `end - start`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            The genomic sequence of length :math:`L` where</span>
<span class="sd">            :math:`L = end - start`. If `pad` is `False` and one/both of</span>
<span class="sd">            `start` and `end` are out of bounds, will return an empty string.</span>
<span class="sd">            Also returns an empty string if `chrom` cannot be found in the</span>
<span class="sd">            input FASTA file.</span>
<span class="sd">            Otherwise, will return the sequence with padding at the start/end</span>
<span class="sd">            if appropriate.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the input char to `strand` is not one of the specified</span>
<span class="sd">            choices.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">_get_sequence_from_coords</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">len_chrs</span><span class="p">,</span>
                                         <span class="bp">self</span><span class="o">.</span><span class="n">_genome_sequence</span><span class="p">,</span>
                                         <span class="n">chrom</span><span class="p">,</span>
                                         <span class="n">start</span><span class="p">,</span>
                                         <span class="n">end</span><span class="p">,</span>
                                         <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span>
                                         <span class="n">pad</span><span class="o">=</span><span class="n">pad</span><span class="p">,</span>
                                         <span class="n">blacklist_tabix</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_blacklist_tabix</span><span class="p">)</span></div>

<div class="viewcode-block" id="Genome.get_encoding_from_coords"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.get_encoding_from_coords">[docs]</a>    <span class="nd">@init</span>
    <span class="k">def</span> <span class="nf">get_encoding_from_coords</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                 <span class="n">chrom</span><span class="p">,</span>
                                 <span class="n">start</span><span class="p">,</span>
                                 <span class="n">end</span><span class="p">,</span>
                                 <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span>
                                 <span class="n">pad</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets the one-hot encoding of the genomic sequence at the</span>
<span class="sd">        queried coordinates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            The name of the chromosome or region, e.g. &quot;chr1&quot;.</span>
<span class="sd">        start : int</span>
<span class="sd">            The 0-based start coordinate of the first position in the</span>
<span class="sd">            sequence.</span>
<span class="sd">        end : int</span>
<span class="sd">            One past the 0-based last position in the sequence.</span>
<span class="sd">        strand : {&#39;+&#39;, &#39;-&#39;, &#39;.&#39;}, optional</span>
<span class="sd">            Default is &#39;+&#39;. The strand the sequence is located on. &#39;.&#39; is</span>
<span class="sd">            treated as &#39;+&#39;.</span>
<span class="sd">        pad : bool, optional</span>
<span class="sd">            Default is `False`. Pad the output sequence with &#39;N&#39; if `start`</span>
<span class="sd">            and/or `end` are out of bounds to return a sequence of length</span>
<span class="sd">            `end - start`.</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        numpy.ndarray, dtype=numpy.float32</span>
<span class="sd">            The :math:`L \\times 4` encoding of the sequence, where</span>
<span class="sd">            :math:`L = end - start`, unless `chrom` cannot be found</span>
<span class="sd">            in the input FASTA, `start` or `end` are out of bounds,</span>
<span class="sd">            or (if a blacklist exists) the region overlaps with a blacklist</span>
<span class="sd">            region. In these cases, it will return an empty encoding--that is,</span>
<span class="sd">            `L` = 0 for the NumPy array returned.</span>


<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the input char to `strand` is not one of the specified</span>
<span class="sd">            choices.</span>
<span class="sd">            (Raised in the call to `self.get_sequence_from_coords`)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_from_coords</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="n">pad</span><span class="p">)</span>
        <span class="n">encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">encoding</span></div>

<div class="viewcode-block" id="Genome.get_encoding_from_coords_check_unk"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.get_encoding_from_coords_check_unk">[docs]</a>    <span class="nd">@init</span>
    <span class="k">def</span> <span class="nf">get_encoding_from_coords_check_unk</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                 <span class="n">chrom</span><span class="p">,</span>
                                 <span class="n">start</span><span class="p">,</span>
                                 <span class="n">end</span><span class="p">,</span>
                                 <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span>
                                 <span class="n">pad</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets the one-hot encoding of the genomic sequence at the</span>
<span class="sd">        queried coordinates and check whether the sequence contains</span>
<span class="sd">        unknown base(s).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            The name of the chromosome or region, e.g. &quot;chr1&quot;.</span>
<span class="sd">        start : int</span>
<span class="sd">            The 0-based start coordinate of the first position in the</span>
<span class="sd">            sequence.</span>
<span class="sd">        end : int</span>
<span class="sd">            One past the 0-based last position in the sequence.</span>
<span class="sd">        strand : {&#39;+&#39;, &#39;-&#39;, &#39;.&#39;}, optional</span>
<span class="sd">            Default is &#39;+&#39;. The strand the sequence is located on. &#39;.&#39; is</span>
<span class="sd">            treated as &#39;+&#39;.</span>
<span class="sd">        pad : bool, optional</span>
<span class="sd">            Default is `False`. Pad the output sequence with &#39;N&#39; if `start`</span>
<span class="sd">            and/or `end` are out of bounds to return a sequence of length</span>
<span class="sd">            `end - start`.</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        tuple(numpy.ndarray, bool)</span>

<span class="sd">            * `tuple[0]` is the :math:`L \\times 4` encoding of the sequence</span>
<span class="sd">            containing data of `numpy.float32` type, where</span>
<span class="sd">            :math:`L = end - start`, unless `chrom` cannot be found</span>
<span class="sd">            in the input FASTA, `start` or `end` are out of bounds,</span>
<span class="sd">            or (if a blacklist exists) the region overlaps with a blacklist</span>
<span class="sd">            region. In these cases, it will return an empty encoding--that is,</span>
<span class="sd">            `L` = 0 for the NumPy array returned.</span>
<span class="sd">            * `tuple[1]` is the boolean value that indicates whether the</span>
<span class="sd">            sequence contains any unknown base(s) specified in self.UNK_BASE</span>


<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the input char to `strand` is not one of the specified</span>
<span class="sd">            choices.</span>
<span class="sd">            (Raised in the call to `self.get_sequence_from_coords`)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_from_coords</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="n">pad</span><span class="p">)</span>
        <span class="n">encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">encoding</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="ow">in</span> <span class="n">sequence</span></div>


<div class="viewcode-block" id="Genome.sequence_to_encoding"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.sequence_to_encoding">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">sequence_to_encoding</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Converts an input sequence to its one-hot encoding.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            A nucleotide sequence of length :math:`L`</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        numpy.ndarray, dtype=numpy.float32</span>
<span class="sd">            The :math:`L \\times 4` one-hot encoding of the sequence.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">sequence_to_encoding</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">BASE_TO_INDEX</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)</span></div>

<div class="viewcode-block" id="Genome.encoding_to_sequence"><a class="viewcode-back" href="../../../sequences.html#selene_sdk.sequences.Genome.encoding_to_sequence">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">encoding_to_sequence</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">encoding</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Converts an input one-hot encoding to its DNA sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        encoding : numpy.ndarray, dtype=numpy.float32</span>
<span class="sd">            An :math:`L \\times 4` one-hot encoding of the sequence,</span>
<span class="sd">            where :math:`L` is the length of the output sequence.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            The sequence of :math:`L` nucleotides decoded from the</span>
<span class="sd">            input array.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">encoding_to_sequence</span><span class="p">(</span><span class="n">encoding</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">UNK_BASE</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2018, Selene Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>