

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene_sdk.predict.model_predict &mdash; Selene 0.5.0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-selene-with-pip">Installing selene with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene_sdk.predict.model_predict</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene_sdk.predict.model_predict</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module provides the `AnalyzeSequences` class and supporting</span>
<span class="sd">methods.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">time</span> <span class="kn">import</span> <span class="n">time</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pyfaidx</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">torch.nn</span> <span class="k">as</span> <span class="nn">nn</span>

<span class="kn">from</span> <span class="nn">._common</span> <span class="kn">import</span> <span class="n">_pad_sequence</span>
<span class="kn">from</span> <span class="nn">._common</span> <span class="kn">import</span> <span class="n">_truncate_sequence</span>
<span class="kn">from</span> <span class="nn">._common</span> <span class="kn">import</span> <span class="n">get_reverse_complement</span>
<span class="kn">from</span> <span class="nn">._common</span> <span class="kn">import</span> <span class="n">get_reverse_complement_encoding</span>
<span class="kn">from</span> <span class="nn">._common</span> <span class="kn">import</span> <span class="n">predict</span>
<span class="kn">from</span> <span class="nn">._in_silico_mutagenesis</span> <span class="kn">import</span> <span class="n">_ism_sample_id</span>
<span class="kn">from</span> <span class="nn">._in_silico_mutagenesis</span> <span class="kn">import</span> <span class="n">in_silico_mutagenesis_sequences</span>
<span class="kn">from</span> <span class="nn">._in_silico_mutagenesis</span> <span class="kn">import</span> <span class="n">mutate_sequence</span>
<span class="kn">from</span> <span class="nn">._variant_effect_prediction</span> <span class="kn">import</span> <span class="n">_handle_long_ref</span>
<span class="kn">from</span> <span class="nn">._variant_effect_prediction</span> <span class="kn">import</span> <span class="n">_handle_standard_ref</span>
<span class="kn">from</span> <span class="nn">._variant_effect_prediction</span> <span class="kn">import</span> <span class="n">_handle_ref_alt_predictions</span>
<span class="kn">from</span> <span class="nn">._variant_effect_prediction</span> <span class="kn">import</span> <span class="n">_process_alt</span>
<span class="kn">from</span> <span class="nn">._variant_effect_prediction</span> <span class="kn">import</span> <span class="n">read_vcf_file</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="kn">import</span> <span class="n">AbsDiffScoreHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="kn">import</span> <span class="n">DiffScoreHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="kn">import</span> <span class="n">LogitScoreHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="kn">import</span> <span class="n">WritePredictionsHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="kn">import</span> <span class="n">WriteRefAltHandler</span>
<span class="kn">from</span> <span class="nn">..sequences</span> <span class="kn">import</span> <span class="n">Genome</span>
<span class="kn">from</span> <span class="nn">..utils</span> <span class="kn">import</span> <span class="n">_is_lua_trained_model</span>
<span class="kn">from</span> <span class="nn">..utils</span> <span class="kn">import</span> <span class="n">load_model_from_state_dict</span>


<span class="c1"># TODO: MAKE THESE GENERIC:</span>
<span class="n">ISM_COLS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;alt&quot;</span><span class="p">]</span>
<span class="n">VARIANTEFFECT_COLS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;alt&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;ref_match&quot;</span><span class="p">,</span> <span class="s2">&quot;contains_unk&quot;</span><span class="p">]</span>

<div class="viewcode-block" id="AnalyzeSequences"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences">[docs]</a><span class="k">class</span> <span class="nc">AnalyzeSequences</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Score sequences and their variants using the predictions made</span>
<span class="sd">    by a trained model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    model : torch.nn.Module</span>
<span class="sd">        A sequence-based model architecture.</span>
<span class="sd">    trained_model_path : str or list(str)</span>
<span class="sd">        The path(s) to the weights file for a trained sequence-based model.</span>
<span class="sd">        For a single path, the model architecture must match `model`. For</span>
<span class="sd">        a list of paths, assumes that the `model` passed in is of type</span>
<span class="sd">        `selene_sdk.utils.MultiModelWrapper`, which takes in a list of</span>
<span class="sd">        models. The paths must be ordered the same way the models</span>
<span class="sd">        are ordered in that list. `list(str)` input is an API-only function--</span>
<span class="sd">        Selene&#39;s config file CLI does not support the `MultiModelWrapper`</span>
<span class="sd">        functionality at this time.</span>
<span class="sd">    sequence_length : int</span>
<span class="sd">        The length of sequences that the model is expecting.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        The names of the features that the model is predicting.</span>
<span class="sd">    batch_size : int, optional</span>
<span class="sd">        Default is 64. The size of the mini-batches to use.</span>
<span class="sd">    use_cuda : bool, optional</span>
<span class="sd">        Default is `False`. Specifies whether CUDA-enabled GPUs are available</span>
<span class="sd">        for torch to use.</span>
<span class="sd">    data_parallel : bool, optional</span>
<span class="sd">        Default is `False`. Specify whether multiple GPUs are available for</span>
<span class="sd">        torch to use during training.</span>
<span class="sd">    reference_sequence : class, optional</span>
<span class="sd">        Default is `selene_sdk.sequences.Genome`. The type of sequence on</span>
<span class="sd">        which this analysis will be performed. Please note that if you need</span>
<span class="sd">        to use variant effect prediction, you cannot only pass in the</span>
<span class="sd">        class--you must pass in the constructed `selene_sdk.sequences.Sequence`</span>
<span class="sd">        object with a particular sequence version (e.g. `Genome(&quot;hg19.fa&quot;)`).</span>
<span class="sd">        This version does NOT have to be the same sequence version that the</span>
<span class="sd">        model was trained on. That is, if the sequences in your variants file</span>
<span class="sd">        are hg19 but your model was trained on hg38 sequences, you should pass</span>
<span class="sd">        in hg19.</span>
<span class="sd">    write_mem_limit : int, optional</span>
<span class="sd">        Default is 5000. Specify, in MB, the amount of memory you want to</span>
<span class="sd">        allocate to storing model predictions/scores. When running one of</span>
<span class="sd">        _in silico_ mutagenesis, variant effect prediction, or prediction,</span>
<span class="sd">        prediction/score handlers will accumulate data in memory and only</span>
<span class="sd">        write this data to files periodically. By default, Selene will write</span>
<span class="sd">        to files when the total amount of data (across all handlers) takes up</span>
<span class="sd">        5000MB of space. Please keep in mind that Selene will not monitor the</span>
<span class="sd">        memory needed to actually carry out the operations (e.g. variant effect</span>
<span class="sd">        prediction) or load the model, so `write_mem_limit` should always be</span>
<span class="sd">        less than the total amount of CPU memory you have available on your</span>
<span class="sd">        machine. For example, for variant effect prediction, we load all</span>
<span class="sd">        the variants in 1 file into memory before getting the predictions, so</span>
<span class="sd">        your machine must have enough memory to accommodate that. Another</span>
<span class="sd">        possible consideration is your model size and whether you are</span>
<span class="sd">        using it on the CPU or a CUDA-enabled GPU (i.e. setting</span>
<span class="sd">        `use_cuda` to True).</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    model : torch.nn.Module</span>
<span class="sd">        A sequence-based model that has already been trained.</span>
<span class="sd">    sequence_length : int</span>
<span class="sd">        The length of sequences that the model is expecting.</span>
<span class="sd">    batch_size : int</span>
<span class="sd">        The size of the mini-batches to use.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        The names of the features that the model is predicting.</span>
<span class="sd">    use_cuda : bool</span>
<span class="sd">        Specifies whether to use a CUDA-enabled GPU or not.</span>
<span class="sd">    data_parallel : bool</span>
<span class="sd">        Whether to use multiple GPUs or not.</span>
<span class="sd">    reference_sequence : class</span>
<span class="sd">        The type of sequence on which this analysis will be performed.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">model</span><span class="p">,</span>
                 <span class="n">trained_model_path</span><span class="p">,</span>
                 <span class="n">sequence_length</span><span class="p">,</span>
                 <span class="n">features</span><span class="p">,</span>
                 <span class="n">batch_size</span><span class="o">=</span><span class="mi">64</span><span class="p">,</span>
                 <span class="n">use_cuda</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">data_parallel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">reference_sequence</span><span class="o">=</span><span class="n">Genome</span><span class="p">,</span>
                 <span class="n">write_mem_limit</span><span class="o">=</span><span class="mi">1500</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructs a new `AnalyzeSequences` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">model</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">trained_model_path</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">trained_model</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
                <span class="n">trained_model_path</span><span class="p">,</span>
                <span class="n">map_location</span><span class="o">=</span><span class="k">lambda</span> <span class="n">storage</span><span class="p">,</span> <span class="n">location</span><span class="p">:</span> <span class="n">storage</span><span class="p">)</span>

            <span class="n">load_model_from_state_dict</span><span class="p">(</span>
                <span class="n">trained_model</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">trained_model_path</span><span class="p">,</span> <span class="s1">&#39;__len__&#39;</span><span class="p">):</span>
            <span class="n">state_dicts</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">mp</span> <span class="ow">in</span> <span class="n">trained_model_path</span><span class="p">:</span>
                <span class="n">state_dict</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
                    <span class="n">mp</span><span class="p">,</span> <span class="n">map_location</span><span class="o">=</span><span class="k">lambda</span> <span class="n">storage</span><span class="p">,</span> <span class="n">location</span><span class="p">:</span> <span class="n">storage</span><span class="p">)</span>
                <span class="n">state_dicts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">state_dict</span><span class="p">)</span>

            <span class="k">for</span> <span class="p">(</span><span class="n">sd</span><span class="p">,</span> <span class="n">sub_model</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">state_dicts</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">sub_models</span><span class="p">):</span>
                <span class="n">load_model_from_state_dict</span><span class="p">(</span><span class="n">sd</span><span class="p">,</span> <span class="n">sub_model</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s1">&#39;`trained_model_path` should be a str or list of strs &#39;</span>
                <span class="s1">&#39;specifying the full paths to model weights files, but was &#39;</span>
                <span class="s1">&#39;type </span><span class="si">{0}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">trained_model_path</span><span class="p">)))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data_parallel</span> <span class="o">=</span> <span class="n">data_parallel</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_parallel</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">DataParallel</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span> <span class="o">=</span> <span class="n">use_cuda</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">=</span> <span class="n">sequence_length</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">=</span> <span class="n">sequence_length</span> <span class="o">//</span> <span class="mi">2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span>
        <span class="k">if</span> <span class="n">sequence_length</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">=</span> <span class="n">batch_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">features</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span> <span class="o">=</span> <span class="n">reference_sequence</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">_initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">_unpicklable_init</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span> <span class="o">==</span> <span class="n">Genome</span> <span class="ow">and</span> \
                <span class="n">_is_lua_trained_model</span><span class="p">(</span><span class="n">model</span><span class="p">):</span>
            <span class="n">Genome</span><span class="o">.</span><span class="n">update_bases_order</span><span class="p">([</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># even if not using Genome, I guess we can update?</span>
            <span class="n">Genome</span><span class="o">.</span><span class="n">update_bases_order</span><span class="p">([</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_mem_limit</span> <span class="o">=</span> <span class="n">write_mem_limit</span>

    <span class="k">def</span> <span class="nf">_initialize_reporters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                              <span class="n">save_data</span><span class="p">,</span>
                              <span class="n">output_path_prefix</span><span class="p">,</span>
                              <span class="n">output_format</span><span class="p">,</span>
                              <span class="n">colnames_for_ids</span><span class="p">,</span>
                              <span class="n">output_size</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                              <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;ism&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize the handlers to which Selene reports model predictions</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_path_prefix : str</span>
<span class="sd">            Path to which the reporters will output data files. Selene will</span>
<span class="sd">            add a prefix to the resulting filename, where the prefix is based</span>
<span class="sd">            on the name of the user-specified input file. This allows a user</span>
<span class="sd">            to distinguish between output files from different inputs when</span>
<span class="sd">            a user specifies the same output directory for multiple inputs.</span>
<span class="sd">        output_format : {&#39;tsv&#39;, &#39;hdf5&#39;}</span>
<span class="sd">            The desired output format. Currently Selene supports TSV and HDF5</span>
<span class="sd">            formats.</span>
<span class="sd">        colnames_for_ids : list(str)</span>
<span class="sd">            Specify the names of columns that will be used to identify the</span>
<span class="sd">            sequence for which Selene has made predictions (e.g. (chrom,</span>
<span class="sd">            pos, id, ref, alt) will be the column names for variant effect</span>
<span class="sd">            prediction outputs).</span>
<span class="sd">        output_size : int, optional</span>
<span class="sd">            The total number of rows in the output. Must be specified when</span>
<span class="sd">            the output_format is hdf5.</span>
<span class="sd">        mode : {&#39;prediction&#39;, &#39;ism&#39;, &#39;varianteffect&#39;}</span>
<span class="sd">            If saving model predictions, the handler Selene chooses for the</span>
<span class="sd">            task is dependent on the mode. For example, the reporter for</span>
<span class="sd">            variant effect prediction writes paired ref and alt predictions</span>
<span class="sd">            to different files.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list(selene_sdk.predict.predict_handlers.PredictionsHandler)</span>
<span class="sd">            List of reporters to update as Selene receives model predictions.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">save_data</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">save_data</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;diffs&quot;</span><span class="p">,</span> <span class="s2">&quot;abs_diffs&quot;</span><span class="p">,</span> <span class="s2">&quot;logits&quot;</span><span class="p">,</span> <span class="s2">&quot;predictions&quot;</span><span class="p">])</span>
        <span class="n">save_data</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">save_data</span><span class="p">))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">save_data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;save_data&#39; parameter must be a list that &quot;</span>
                             <span class="s2">&quot;contains one of [&#39;diffs&#39;, &#39;abs_diffs&#39;, &quot;</span>
                             <span class="s2">&quot;&#39;logits&#39;, &#39;predictions&#39;].&quot;</span><span class="p">)</span>
        <span class="n">reporters</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">constructor_args</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">,</span>
                            <span class="n">colnames_for_ids</span><span class="p">,</span>
                            <span class="n">output_path_prefix</span><span class="p">,</span>
                            <span class="n">output_format</span><span class="p">,</span>
                            <span class="n">output_size</span><span class="p">,</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_write_mem_limit</span> <span class="o">//</span> <span class="nb">len</span><span class="p">(</span><span class="n">save_data</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">save_data</span><span class="p">):</span>
            <span class="n">write_labels</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">write_labels</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">if</span> <span class="s2">&quot;diffs&quot;</span> <span class="o">==</span> <span class="n">s</span><span class="p">:</span>
                <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">DiffScoreHandler</span><span class="p">(</span>
                    <span class="o">*</span><span class="n">constructor_args</span><span class="p">,</span> <span class="n">write_labels</span><span class="o">=</span><span class="n">write_labels</span><span class="p">))</span>
            <span class="k">elif</span> <span class="s2">&quot;abs_diffs&quot;</span> <span class="o">==</span> <span class="n">s</span><span class="p">:</span>
                <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">AbsDiffScoreHandler</span><span class="p">(</span>
                    <span class="o">*</span><span class="n">constructor_args</span><span class="p">,</span> <span class="n">write_labels</span><span class="o">=</span><span class="n">write_labels</span><span class="p">))</span>
            <span class="k">elif</span> <span class="s2">&quot;logits&quot;</span> <span class="o">==</span> <span class="n">s</span><span class="p">:</span>
                <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">LogitScoreHandler</span><span class="p">(</span>
                    <span class="o">*</span><span class="n">constructor_args</span><span class="p">,</span> <span class="n">write_labels</span><span class="o">=</span><span class="n">write_labels</span><span class="p">))</span>
            <span class="k">elif</span> <span class="s2">&quot;predictions&quot;</span> <span class="o">==</span> <span class="n">s</span> <span class="ow">and</span> <span class="n">mode</span> <span class="o">!=</span> <span class="s2">&quot;varianteffect&quot;</span><span class="p">:</span>
                <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">WritePredictionsHandler</span><span class="p">(</span>
                    <span class="o">*</span><span class="n">constructor_args</span><span class="p">,</span> <span class="n">write_labels</span><span class="o">=</span><span class="n">write_labels</span><span class="p">))</span>
            <span class="k">elif</span> <span class="s2">&quot;predictions&quot;</span> <span class="o">==</span> <span class="n">s</span> <span class="ow">and</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;varianteffect&quot;</span><span class="p">:</span>
                <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">WriteRefAltHandler</span><span class="p">(</span>
                    <span class="o">*</span><span class="n">constructor_args</span><span class="p">,</span> <span class="n">write_labels</span><span class="o">=</span><span class="n">write_labels</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">reporters</span>

    <span class="k">def</span> <span class="nf">_get_sequences_from_bed_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                     <span class="n">input_path</span><span class="p">,</span>
                                     <span class="n">strand_index</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                     <span class="n">output_NAs_to_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                     <span class="n">reference_sequence</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the adjusted sequence coordinates and labels corresponding</span>
<span class="sd">        to each row of coordinates in an input BED file. The coordinates</span>
<span class="sd">        specified in each row are only used to find the center position</span>
<span class="sd">        for the resulting sequence--all regions returned will have the</span>
<span class="sd">        length expected by the model.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : str</span>
<span class="sd">            Input filepath to BED file.</span>
<span class="sd">        strand_index : int or None, optional</span>
<span class="sd">            Default is None. If sequences must be strand-specific,</span>
<span class="sd">            the input BED file may include a column specifying the</span>
<span class="sd">            strand ({&#39;+&#39;, &#39;-&#39;, &#39;.&#39;}).</span>
<span class="sd">        output_NAs_to_file : str or None, optional</span>
<span class="sd">            Default is None. Only used if `reference_sequence` is also not None.</span>
<span class="sd">            Specify a filepath to which invalid variants are written.</span>
<span class="sd">            Invalid = sequences that cannot be fetched, either because</span>
<span class="sd">            the exact chromosome cannot be found in the `reference_sequence` FASTA</span>
<span class="sd">            file or because the sequence retrieved is out of bounds or overlapping</span>
<span class="sd">            with any of the blacklist regions.</span>
<span class="sd">        reference_sequence : selene_sdk.sequences.Genome or None, optional</span>
<span class="sd">            Default is None. The reference genome.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list(tup), list(tup)</span>
<span class="sd">            The sequence query information (chrom, start, end, strand)</span>
<span class="sd">            and the labels (the index, genome coordinates, and sequence</span>
<span class="sd">            specified in the BED file).</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">na_rows</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">check_chr</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_chrs</span><span class="p">():</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">chrom</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">):</span>
                <span class="n">check_chr</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">break</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">read_handle</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">read_handle</span><span class="p">):</span>
                <span class="n">cols</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">3</span><span class="p">:</span>
                    <span class="n">na_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">strand_index</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">strand_index</span><span class="p">:</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="n">strand_index</span><span class="p">]</span>
                <span class="k">if</span> <span class="s1">&#39;chr&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">chrom</span> <span class="ow">and</span> <span class="n">check_chr</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="s2">&quot;chr</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="nb">str</span><span class="o">.</span><span class="n">isdigit</span><span class="p">(</span><span class="n">start</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">str</span><span class="o">.</span><span class="n">isdigit</span><span class="p">(</span><span class="n">end</span><span class="p">)</span> \
                        <span class="ow">or</span> <span class="n">chrom</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span>
                    <span class="n">na_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span>
                <span class="n">mid_pos</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="p">((</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
                <span class="n">seq_start</span> <span class="o">=</span> <span class="n">mid_pos</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span>
                <span class="n">seq_end</span> <span class="o">=</span> <span class="n">mid_pos</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span>
                <span class="k">if</span> <span class="n">reference_sequence</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">coords_in_bounds</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seq_start</span><span class="p">,</span> <span class="n">seq_end</span><span class="p">):</span>
                        <span class="n">na_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                        <span class="k">continue</span>
                <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seq_start</span><span class="p">,</span> <span class="n">seq_end</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
                <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">i</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">reference_sequence</span> <span class="ow">and</span> <span class="n">output_NAs_to_file</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_NAs_to_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">na_row</span> <span class="ow">in</span> <span class="n">na_rows</span><span class="p">:</span>
                    <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">na_row</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">labels</span>

<div class="viewcode-block" id="AnalyzeSequences.get_predictions_for_bed_file"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.get_predictions_for_bed_file">[docs]</a>    <span class="k">def</span> <span class="nf">get_predictions_for_bed_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                     <span class="n">input_path</span><span class="p">,</span>
                                     <span class="n">output_dir</span><span class="p">,</span>
                                     <span class="n">output_format</span><span class="o">=</span><span class="s2">&quot;tsv&quot;</span><span class="p">,</span>
                                     <span class="n">strand_index</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get model predictions for sequences specified as genome coordinates</span>
<span class="sd">        in a BED file. Coordinates do not need to be the same length as the</span>
<span class="sd">        model expected sequence input--predictions will be centered at the</span>
<span class="sd">        midpoint of the specified start and end coordinates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : str</span>
<span class="sd">            Input path to the BED file.</span>
<span class="sd">        output_dir : str</span>
<span class="sd">            Output directory to write the model predictions.</span>
<span class="sd">        output_format : {&#39;tsv&#39;, &#39;hdf5&#39;}, optional</span>
<span class="sd">            Default is &#39;tsv&#39;. Choose whether to save TSV or HDF5 output files.</span>
<span class="sd">            TSV is easier to access (i.e. open with text editor/Excel) and</span>
<span class="sd">            quickly peruse, whereas HDF5 files must be accessed through</span>
<span class="sd">            specific packages/viewers that support this format (e.g. h5py</span>
<span class="sd">            Python package). Choose</span>

<span class="sd">                * &#39;tsv&#39; if your list of sequences is relatively small</span>
<span class="sd">                  (:math:`10^4` or less in order of magnitude) and/or your</span>
<span class="sd">                  model has a small number of features (&lt;1000).</span>
<span class="sd">                * &#39;hdf5&#39; for anything larger and/or if you would like to</span>
<span class="sd">                  access the predictions/scores as a matrix that you can</span>
<span class="sd">                  easily filter, apply computations, or use in a subsequent</span>
<span class="sd">                  classifier/model. In this case, you may access the matrix</span>
<span class="sd">                  using `mat[&quot;data&quot;]` after opening the HDF5 file using</span>
<span class="sd">                  `mat = h5py.File(&quot;&lt;output.h5&gt;&quot;, &#39;r&#39;)`. The matrix columns</span>
<span class="sd">                  are the features and will match the same ordering as your</span>
<span class="sd">                  features .txt file (same as the order your model outputs</span>
<span class="sd">                  its predictions) and the matrix rows are the sequences.</span>
<span class="sd">                  Note that the row labels (FASTA description/IDs) will be</span>
<span class="sd">                  output as a separate .txt file (should match the ordering</span>
<span class="sd">                  of the sequences in the input FASTA).</span>

<span class="sd">        strand_index : int or None, optional</span>
<span class="sd">            Default is None. If the trained model makes strand-specific</span>
<span class="sd">            predictions, your input file may include a column with strand</span>
<span class="sd">            information (strand must be one of {&#39;+&#39;, &#39;-&#39;, &#39;.&#39;}). Specify</span>
<span class="sd">            the index (0-based) to use it. Otherwise, by default &#39;+&#39; is used.</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes the output to file(s) in `output_dir`. Filename will</span>
<span class="sd">            match that specified in the filepath.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="n">output_prefix</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">seq_coords</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_sequences_from_bed_file</span><span class="p">(</span>
            <span class="n">input_path</span><span class="p">,</span>
            <span class="n">strand_index</span><span class="o">=</span><span class="n">strand_index</span><span class="p">,</span>
            <span class="n">output_NAs_to_file</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">.NA&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">)),</span>
            <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>

        <span class="n">reporter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;predictions&quot;</span><span class="p">],</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">),</span>
            <span class="n">output_format</span><span class="p">,</span>
            <span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;contains_unk&quot;</span><span class="p">],</span>
            <span class="n">output_size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">),</span>
            <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;prediction&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">coords</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">seq_coords</span><span class="p">)):</span>
            <span class="n">encoding</span><span class="p">,</span> <span class="n">contains_unk</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_encoding_from_coords_check_unk</span><span class="p">(</span>
                    <span class="o">*</span><span class="n">coords</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">sequences</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="o">*</span><span class="n">encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">preds</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>
                <span class="n">reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">preds</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>
                <span class="n">sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="o">*</span><span class="n">encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
                <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">sequences</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">encoding</span>
            <span class="n">batch_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">label</span><span class="o">+</span><span class="p">(</span><span class="n">contains_unk</span><span class="p">,))</span>
            <span class="k">if</span> <span class="n">contains_unk</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">((</span><span class="s2">&quot;For region </span><span class="si">{0}</span><span class="s2">, &quot;</span>
                               <span class="s2">&quot;reference sequence contains unknown &quot;</span>
                               <span class="s2">&quot;base(s). --will be marked `True` in the &quot;</span>
                               <span class="s2">&quot;`contains_unk` column of the .tsv or &quot;</span>
                               <span class="s2">&quot;row_labels .txt file.&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">label</span><span class="p">))</span>

        <span class="n">sequences</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[:</span><span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
        <span class="n">preds</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>
        <span class="n">reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">preds</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>
        <span class="n">reporter</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">()</span></div>


<div class="viewcode-block" id="AnalyzeSequences.get_predictions_for_fasta_file"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.get_predictions_for_fasta_file">[docs]</a>    <span class="k">def</span> <span class="nf">get_predictions_for_fasta_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                       <span class="n">input_path</span><span class="p">,</span>
                                       <span class="n">output_dir</span><span class="p">,</span>
                                       <span class="n">output_format</span><span class="o">=</span><span class="s2">&quot;tsv&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get model predictions for sequences in a FASTA file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : str</span>
<span class="sd">            Input path to the FASTA file.</span>
<span class="sd">        output_dir : str</span>
<span class="sd">            Output directory to write the model predictions.</span>
<span class="sd">        output_format : {&#39;tsv&#39;, &#39;hdf5&#39;}, optional</span>
<span class="sd">            Default is &#39;tsv&#39;. Choose whether to save TSV or HDF5 output files.</span>
<span class="sd">            TSV is easier to access (i.e. open with text editor/Excel) and</span>
<span class="sd">            quickly peruse, whereas HDF5 files must be accessed through</span>
<span class="sd">            specific packages/viewers that support this format (e.g. h5py</span>
<span class="sd">            Python package). Choose</span>

<span class="sd">                * &#39;tsv&#39; if your list of sequences is relatively small</span>
<span class="sd">                  (:math:`10^4` or less in order of magnitude) and/or your</span>
<span class="sd">                  model has a small number of features (&lt;1000).</span>
<span class="sd">                * &#39;hdf5&#39; for anything larger and/or if you would like to</span>
<span class="sd">                  access the predictions/scores as a matrix that you can</span>
<span class="sd">                  easily filter, apply computations, or use in a subsequent</span>
<span class="sd">                  classifier/model. In this case, you may access the matrix</span>
<span class="sd">                  using `mat[&quot;data&quot;]` after opening the HDF5 file using</span>
<span class="sd">                  `mat = h5py.File(&quot;&lt;output.h5&gt;&quot;, &#39;r&#39;)`. The matrix columns</span>
<span class="sd">                  are the features and will match the same ordering as your</span>
<span class="sd">                  features .txt file (same as the order your model outputs</span>
<span class="sd">                  its predictions) and the matrix rows are the sequences.</span>
<span class="sd">                  Note that the row labels (FASTA description/IDs) will be</span>
<span class="sd">                  output as a separate .txt file (should match the ordering</span>
<span class="sd">                  of the sequences in the input FASTA).</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes the output to file(s) in `output_dir`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">_</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="n">output_prefix</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">fasta_file</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="n">reporter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;predictions&quot;</span><span class="p">],</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">),</span>
            <span class="n">output_format</span><span class="p">,</span>
            <span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">],</span>
            <span class="n">output_size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">fasta_file</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
            <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;prediction&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span>
                              <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
                              <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)))</span>
        <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">fasta_record</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">):</span>
            <span class="n">cur_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pad_or_truncate_sequence</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">fasta_record</span><span class="p">))</span>
            <span class="n">cur_sequence_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">i</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">preds</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>
                <span class="n">sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
                    <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="o">*</span><span class="n">cur_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
                <span class="n">reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">preds</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>
                <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="n">batch_ids</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">i</span><span class="p">,</span> <span class="n">fasta_record</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>
            <span class="n">sequences</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">cur_sequence_encoding</span>

        <span class="n">sequences</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[:</span><span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
        <span class="n">preds</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>
        <span class="n">reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">preds</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>

        <span class="n">fasta_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">reporter</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">()</span></div>


<div class="viewcode-block" id="AnalyzeSequences.get_predictions"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.get_predictions">[docs]</a>    <span class="k">def</span> <span class="nf">get_predictions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                        <span class="n">input_path</span><span class="p">,</span>
                        <span class="n">output_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">output_format</span><span class="o">=</span><span class="s2">&quot;tsv&quot;</span><span class="p">,</span>
                        <span class="n">strand_index</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get model predictions for sequences specified as a raw sequence,</span>
<span class="sd">        FASTA, or BED file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : str</span>
<span class="sd">            A single sequence, or a path to the FASTA or BED file input.</span>
<span class="sd">        output_dir : str, optional</span>
<span class="sd">            Default is None. Output directory to write the model predictions.</span>
<span class="sd">            If this is left blank a raw sequence input will be assumed, though</span>
<span class="sd">            an output directory is required for FASTA and BED inputs.</span>
<span class="sd">        output_format : {&#39;tsv&#39;, &#39;hdf5&#39;}, optional</span>
<span class="sd">            Default is &#39;tsv&#39;. Choose whether to save TSV or HDF5 output files.</span>
<span class="sd">            TSV is easier to access (i.e. open with text editor/Excel) and</span>
<span class="sd">            quickly peruse, whereas HDF5 files must be accessed through</span>
<span class="sd">            specific packages/viewers that support this format (e.g. h5py</span>
<span class="sd">            Python package). Choose</span>

<span class="sd">                * &#39;tsv&#39; if your list of sequences is relatively small</span>
<span class="sd">                  (:math:`10^4` or less in order of magnitude) and/or your</span>
<span class="sd">                  model has a small number of features (&lt;1000).</span>
<span class="sd">                * &#39;hdf5&#39; for anything larger and/or if you would like to</span>
<span class="sd">                  access the predictions/scores as a matrix that you can</span>
<span class="sd">                  easily filter, apply computations, or use in a subsequent</span>
<span class="sd">                  classifier/model. In this case, you may access the matrix</span>
<span class="sd">                  using `mat[&quot;data&quot;]` after opening the HDF5 file using</span>
<span class="sd">                  `mat = h5py.File(&quot;&lt;output.h5&gt;&quot;, &#39;r&#39;)`. The matrix columns</span>
<span class="sd">                  are the features and will match the same ordering as your</span>
<span class="sd">                  features .txt file (same as the order your model outputs</span>
<span class="sd">                  its predictions) and the matrix rows are the sequences.</span>
<span class="sd">                  Note that the row labels (FASTA description/IDs) will be</span>
<span class="sd">                  output as a separate .txt file (should match the ordering</span>
<span class="sd">                  of the sequences in the input FASTA).</span>

<span class="sd">        strand_index : int or None, optional</span>
<span class="sd">            Default is None. If the trained model makes strand-specific</span>
<span class="sd">            predictions, your input BED file may include a column with strand</span>
<span class="sd">            information (strand must be one of {&#39;+&#39;, &#39;-&#39;, &#39;.&#39;}). Specify</span>
<span class="sd">            the index (0-based) to use it. Otherwise, by default &#39;+&#39; is used.</span>
<span class="sd">            (This parameter is ignored if FASTA file is used as input.)</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes the output to file(s) in `output_dir`. Filename will</span>
<span class="sd">            match that specified in the filepath. In addition, if any base</span>
<span class="sd">            in the given or retrieved sequence is unknown, the row labels .txt file</span>
<span class="sd">            or .tsv file will mark this sequence or region as `contains_unk = True`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pad_or_truncate_sequence</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
            <span class="n">seq_enc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
            <span class="n">seq_enc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">expand_dims</span><span class="p">(</span><span class="n">seq_enc</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># add batch size of 1</span>
            <span class="k">return</span> <span class="n">predict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">seq_enc</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">input_path</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fa&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">input_path</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_predictions_for_fasta_file</span><span class="p">(</span>
                <span class="n">input_path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">output_format</span><span class="o">=</span><span class="n">output_format</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_predictions_for_bed_file</span><span class="p">(</span>
                <span class="n">input_path</span><span class="p">,</span>
                <span class="n">output_dir</span><span class="p">,</span>
                <span class="n">output_format</span><span class="o">=</span><span class="n">output_format</span><span class="p">,</span>
                <span class="n">strand_index</span><span class="o">=</span><span class="n">strand_index</span><span class="p">)</span>

        <span class="k">return</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="AnalyzeSequences.in_silico_mutagenesis_predict"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_predict">[docs]</a>    <span class="k">def</span> <span class="nf">in_silico_mutagenesis_predict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                      <span class="n">sequence</span><span class="p">,</span>
                                      <span class="n">base_preds</span><span class="p">,</span>
                                      <span class="n">mutations_list</span><span class="p">,</span>
                                      <span class="n">reporters</span><span class="o">=</span><span class="p">[]):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the predictions for all specified mutations applied</span>
<span class="sd">        to a given sequence and, if applicable, compute the scores</span>
<span class="sd">        (&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;) for these mutations.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            The sequence to mutate.</span>
<span class="sd">        base_preds : numpy.ndarray</span>
<span class="sd">            The model&#39;s prediction for `sequence`.</span>
<span class="sd">        mutations_list : list(list(tuple))</span>
<span class="sd">            The mutations to apply to the sequence. Each element in</span>
<span class="sd">            `mutations_list` is a list of tuples, where each tuple</span>
<span class="sd">            specifies the `int` position in the sequence to mutate and what</span>
<span class="sd">            `str` base to which the position is mutated (e.g. (1, &#39;A&#39;)).</span>
<span class="sd">        reporters : list(PredictionsHandler)</span>
<span class="sd">            The list of reporters, where each reporter handles the predictions</span>
<span class="sd">            made for each mutated sequence. Will collect, compute scores</span>
<span class="sd">            (e.g. `AbsDiffScoreHandler` computes the absolute difference</span>
<span class="sd">            between `base_preds` and the predictions for the mutated</span>
<span class="sd">            sequence), and output these as a file at the end.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes results to files corresponding to each reporter in</span>
<span class="sd">            `reporters`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">current_sequence_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations_list</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">):</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">i</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations_list</span><span class="p">))</span>

            <span class="n">mutated_sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
                <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="o">*</span><span class="n">current_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>

            <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">mutation_info</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mutations_list</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]):</span>
                <span class="n">mutated_seq</span> <span class="o">=</span> <span class="n">mutate_sequence</span><span class="p">(</span>
                    <span class="n">current_sequence_encoding</span><span class="p">,</span> <span class="n">mutation_info</span><span class="p">,</span>
                    <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>
                <span class="n">mutated_sequences</span><span class="p">[</span><span class="n">ix</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">mutated_seq</span>
                <span class="n">batch_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_ism_sample_id</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">mutation_info</span><span class="p">))</span>
            <span class="n">outputs</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">mutated_sequences</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">needs_base_pred</span><span class="p">:</span>
                    <span class="n">r</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">outputs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">,</span> <span class="n">base_preds</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">r</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">outputs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
            <span class="n">r</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">()</span></div>

<div class="viewcode-block" id="AnalyzeSequences.in_silico_mutagenesis"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis">[docs]</a>    <span class="k">def</span> <span class="nf">in_silico_mutagenesis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                              <span class="n">sequence</span><span class="p">,</span>
                              <span class="n">save_data</span><span class="p">,</span>
                              <span class="n">output_path_prefix</span><span class="o">=</span><span class="s2">&quot;ism&quot;</span><span class="p">,</span>
                              <span class="n">mutate_n_bases</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                              <span class="n">output_format</span><span class="o">=</span><span class="s2">&quot;tsv&quot;</span><span class="p">,</span>
                              <span class="n">start_position</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                              <span class="n">end_position</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Applies *in silico* mutagenesis to a sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            The sequence to mutate.</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_path_prefix : str, optional</span>
<span class="sd">            The path to which the data files are written. If directories in</span>
<span class="sd">            the path do not yet exist they will be automatically created.</span>
<span class="sd">        mutate_n_bases : int, optional</span>
<span class="sd">            The number of bases to mutate at one time. We recommend leaving</span>
<span class="sd">            this parameter set to `1` at this time, as we have not yet</span>
<span class="sd">            optimized operations for double and triple mutations.</span>
<span class="sd">        output_format : {&#39;tsv&#39;, &#39;hdf5&#39;}, optional</span>
<span class="sd">            Default is &#39;tsv&#39;. The desired output format.</span>
<span class="sd">        start_position : int, optional</span>
<span class="sd">            Default is 0. The starting position of the subsequence to be</span>
<span class="sd">            mutated.</span>
<span class="sd">        end_position : int or None, optional</span>
<span class="sd">            Default is None. The ending position of the subsequence to be</span>
<span class="sd">            mutated. If left as `None`, then `self.sequence_length` will be</span>
<span class="sd">            used.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Outputs data files from *in silico* mutagenesis to `output_dir`.</span>
<span class="sd">            For HDF5 output and &#39;predictions&#39; in `save_data`, an additional</span>
<span class="sd">            file named `*_ref_predictions.h5` will be outputted with the</span>
<span class="sd">            model prediction for the original input sequence.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the value of `start_position` or `end_position` is negative.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If there are fewer than `mutate_n_bases` between `start_position`</span>
<span class="sd">            and `end_position`.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If `start_position` is greater or equal to `end_position`.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If `start_position` is not less than `self.sequence_length`.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If `end_position` is greater than `self.sequence_length`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">end_position</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">end_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
        <span class="k">if</span> <span class="n">start_position</span> <span class="o">&gt;=</span> <span class="n">end_position</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Starting positions must be less than the ending &quot;</span>
                              <span class="s2">&quot;positions. Found a starting position of </span><span class="si">{0}</span><span class="s2"> with &quot;</span>
                              <span class="s2">&quot;an ending position of </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">start_position</span><span class="p">,</span>
                                                                   <span class="n">end_position</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">start_position</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Negative starting positions are not supported.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">end_position</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Negative ending positions are not supported.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">start_position</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Starting positions must be less than the sequence length.&quot;</span>
                              <span class="s2">&quot; Found a starting position of </span><span class="si">{0}</span><span class="s2"> with a sequence length &quot;</span>
                              <span class="s2">&quot;of </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">start_position</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">end_position</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Ending positions must be less than or equal to the sequence &quot;</span>
                              <span class="s2">&quot;length. Found an ending position of </span><span class="si">{0}</span><span class="s2"> with a sequence &quot;</span>
                              <span class="s2">&quot;length of </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">end_position</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">))</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">end_position</span> <span class="o">-</span> <span class="n">start_position</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">mutate_n_bases</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Fewer bases exist in the substring specified by the starting &quot;</span>
                              <span class="s2">&quot;and ending positions than need to be mutated. There are only &quot;</span>
                              <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> currently, but </span><span class="si">{1}</span><span class="s2"> bases must be mutated at a &quot;</span>
                              <span class="s2">&quot;time&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">end_position</span> <span class="o">-</span> <span class="n">start_position</span><span class="p">,</span> <span class="n">mutate_n_bases</span><span class="p">))</span>

        <span class="n">path_dirs</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">path_dirs</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path_dirs</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span> <span class="c1"># Pad string length as necessary.</span>
             <span class="n">diff</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="n">n</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span>
             <span class="n">pad_l</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">diff</span><span class="p">))</span>
             <span class="n">pad_r</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span>
             <span class="n">sequence</span> <span class="o">=</span> <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">pad_l</span><span class="p">)</span> <span class="o">+</span>
                         <span class="n">sequence</span> <span class="o">+</span>
                         <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">pad_r</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>  <span class="c1"># Extract center substring of proper length.</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">n</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>

        <span class="n">sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="n">mutated_sequences</span> <span class="o">=</span> <span class="n">in_silico_mutagenesis_sequences</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">,</span> <span class="n">mutate_n_bases</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">,</span>
            <span class="n">start_position</span><span class="o">=</span><span class="n">start_position</span><span class="p">,</span>
            <span class="n">end_position</span><span class="o">=</span><span class="n">end_position</span><span class="p">)</span>
        <span class="n">reporters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
            <span class="n">save_data</span><span class="p">,</span>
            <span class="n">output_path_prefix</span><span class="p">,</span>
            <span class="n">output_format</span><span class="p">,</span>
            <span class="n">ISM_COLS</span><span class="p">,</span>
            <span class="n">output_size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">mutated_sequences</span><span class="p">))</span>

        <span class="n">current_sequence_encoding</span> <span class="o">=</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

        <span class="n">current_sequence_encoding</span> <span class="o">=</span> <span class="n">current_sequence_encoding</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
            <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">*</span><span class="n">current_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
        <span class="n">base_preds</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">current_sequence_encoding</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span> <span class="ow">and</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s1">&#39;hdf5&#39;</span><span class="p">:</span>
            <span class="n">ref_reporter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
                <span class="p">[</span><span class="s2">&quot;predictions&quot;</span><span class="p">],</span>
                <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_ref&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">),</span>
                <span class="n">output_format</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="n">output_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">ref_reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span>
                <span class="n">base_preds</span><span class="p">,</span> <span class="p">[[</span><span class="s2">&quot;input_sequence&quot;</span><span class="p">]])</span>
            <span class="n">ref_reporter</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">()</span>
        <span class="k">elif</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span> <span class="ow">and</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s1">&#39;tsv&#39;</span><span class="p">:</span>
            <span class="n">reporters</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span>
                <span class="n">base_preds</span><span class="p">,</span> <span class="p">[[</span><span class="s2">&quot;input_sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">]])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">in_silico_mutagenesis_predict</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">,</span>
            <span class="n">base_preds</span><span class="p">,</span>
            <span class="n">mutated_sequences</span><span class="p">,</span>
            <span class="n">reporters</span><span class="o">=</span><span class="n">reporters</span><span class="p">)</span></div>

<div class="viewcode-block" id="AnalyzeSequences.in_silico_mutagenesis_from_file"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_from_file">[docs]</a>    <span class="k">def</span> <span class="nf">in_silico_mutagenesis_from_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                        <span class="n">input_path</span><span class="p">,</span>
                                        <span class="n">save_data</span><span class="p">,</span>
                                        <span class="n">output_dir</span><span class="p">,</span>
                                        <span class="n">mutate_n_bases</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                        <span class="n">use_sequence_name</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                        <span class="n">output_format</span><span class="o">=</span><span class="s2">&quot;tsv&quot;</span><span class="p">,</span>
                                        <span class="n">start_position</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                        <span class="n">end_position</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Apply *in silico* mutagenesis to all sequences in a FASTA file.</span>

<span class="sd">        Please note that we have not parallelized this function yet, so runtime</span>
<span class="sd">        increases exponentially when you increase `mutate_n_bases`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path: str</span>
<span class="sd">            The path to the FASTA file of sequences.</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_dir : str</span>
<span class="sd">            The path to the output directory. Directories in the path will be</span>
<span class="sd">            created if they do not currently exist.</span>
<span class="sd">        mutate_n_bases : int, optional</span>
<span class="sd">            Default is 1. The number of bases to mutate at one time in</span>
<span class="sd">            *in silico* mutagenesis.</span>
<span class="sd">        use_sequence_name : bool, optional.</span>
<span class="sd">            Default is True. If `use_sequence_name`, output files are prefixed</span>
<span class="sd">            by the sequence name/description corresponding to each sequence</span>
<span class="sd">            in the FASTA file. Spaces in the sequence name are replaced with</span>
<span class="sd">            underscores &#39;_&#39;. If not `use_sequence_name`, output files are</span>
<span class="sd">            prefixed with an index :math:`i` (starting with 0) corresponding</span>
<span class="sd">            to the :math:`i`th sequence in the FASTA file.</span>
<span class="sd">        output_format : {&#39;tsv&#39;, &#39;hdf5&#39;}, optional</span>
<span class="sd">            Default is &#39;tsv&#39;. The desired output format. Each sequence in</span>
<span class="sd">            the FASTA file will have its own set of output files, where</span>
<span class="sd">            the number of output files depends on the number of `save_data`</span>
<span class="sd">            predictions/scores specified.</span>
<span class="sd">        start_position : int, optional</span>
<span class="sd">            Default is 0. The starting position of the subsequence to be</span>
<span class="sd">            mutated.</span>
<span class="sd">        end_position : int or None, optional</span>
<span class="sd">            Default is None. The ending position of the subsequence to be</span>
<span class="sd">            mutated. If left as `None`, then `self.sequence_length` will be</span>
<span class="sd">            used.</span>




<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Outputs data files from *in silico* mutagenesis to `output_dir`.</span>
<span class="sd">            For HDF5 output and &#39;predictions&#39; in `save_data`, an additional</span>
<span class="sd">            file named `*_ref_predictions.h5` will be outputted with the</span>
<span class="sd">            model prediction for the original input sequence.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the value of `start_position` or `end_position` is negative.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If there are fewer than `mutate_n_bases` between `start_position`</span>
<span class="sd">            and `end_position`.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If `start_position` is greater or equal to `end_position`.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If `start_position` is not less than `self.sequence_length`.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If `end_position` is greater than `self.sequence_length`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">end_position</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">end_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
        <span class="k">if</span> <span class="n">start_position</span> <span class="o">&gt;=</span> <span class="n">end_position</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Starting positions must be less than the ending &quot;</span>
                              <span class="s2">&quot;positions. Found a starting position of </span><span class="si">{0}</span><span class="s2"> with &quot;</span>
                              <span class="s2">&quot;an ending position of </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">start_position</span><span class="p">,</span>
                                                                   <span class="n">end_position</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">start_position</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Negative starting positions are not supported.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">end_position</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Negative ending positions are not supported.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">start_position</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Starting positions must be less than the sequence length.&quot;</span>
                              <span class="s2">&quot; Found a starting position of </span><span class="si">{0}</span><span class="s2"> with a sequence length &quot;</span>
                              <span class="s2">&quot;of </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">start_position</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">end_position</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Ending positions must be less than or equal to the sequence &quot;</span>
                              <span class="s2">&quot;length. Found an ending position of </span><span class="si">{0}</span><span class="s2"> with a sequence &quot;</span>
                              <span class="s2">&quot;length of </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">end_position</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">))</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">end_position</span> <span class="o">-</span> <span class="n">start_position</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">mutate_n_bases</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">((</span><span class="s2">&quot;Fewer bases exist in the substring specified by the starting &quot;</span>
                              <span class="s2">&quot;and ending positions than need to be mutated. There are only &quot;</span>
                              <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> currently, but </span><span class="si">{1}</span><span class="s2"> bases must be mutated at a &quot;</span>
                              <span class="s2">&quot;time&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">end_position</span> <span class="o">-</span> <span class="n">start_position</span><span class="p">,</span> <span class="n">mutate_n_bases</span><span class="p">))</span>

        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">fasta_file</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">fasta_record</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">):</span>
            <span class="n">cur_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pad_or_truncate_sequence</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">fasta_record</span><span class="p">)))</span>

            <span class="c1"># Generate mut sequences and base preds.</span>
            <span class="n">mutated_sequences</span> <span class="o">=</span> <span class="n">in_silico_mutagenesis_sequences</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">,</span>
                <span class="n">mutate_n_bases</span><span class="o">=</span><span class="n">mutate_n_bases</span><span class="p">,</span>
                <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">,</span>
                <span class="n">start_position</span><span class="o">=</span><span class="n">start_position</span><span class="p">,</span>
                <span class="n">end_position</span><span class="o">=</span><span class="n">end_position</span><span class="p">)</span>
            <span class="n">cur_sequence_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">)</span>
            <span class="n">base_encoding</span> <span class="o">=</span> <span class="n">cur_sequence_encoding</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
                <span class="mi">1</span><span class="p">,</span> <span class="o">*</span><span class="n">cur_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
            <span class="n">base_preds</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">base_encoding</span><span class="p">,</span> <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>

            <span class="n">file_prefix</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">use_sequence_name</span><span class="p">:</span>
                <span class="n">file_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">output_dir</span><span class="p">,</span> <span class="n">fasta_record</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">file_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">output_dir</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
            <span class="c1"># Write base to file, and make mut preds.</span>
            <span class="n">reporters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
                <span class="n">save_data</span><span class="p">,</span>
                <span class="n">file_prefix</span><span class="p">,</span>
                <span class="n">output_format</span><span class="p">,</span>
                <span class="n">ISM_COLS</span><span class="p">,</span>
                <span class="n">output_size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">mutated_sequences</span><span class="p">))</span>

            <span class="k">if</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span> <span class="ow">and</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s1">&#39;hdf5&#39;</span><span class="p">:</span>
                <span class="n">ref_reporter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
                    <span class="p">[</span><span class="s2">&quot;predictions&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_ref&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_prefix</span><span class="p">),</span>
                    <span class="n">output_format</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="n">output_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">ref_reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span>
                    <span class="n">base_preds</span><span class="p">,</span> <span class="p">[[</span><span class="s2">&quot;input_sequence&quot;</span><span class="p">]])</span>
                <span class="n">ref_reporter</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">()</span>
            <span class="k">elif</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span> <span class="ow">and</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s1">&#39;tsv&#39;</span><span class="p">:</span>
                <span class="n">reporters</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span>
                    <span class="n">base_preds</span><span class="p">,</span> <span class="p">[[</span><span class="s2">&quot;input_sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">]])</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">in_silico_mutagenesis_predict</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">,</span> <span class="n">base_preds</span><span class="p">,</span> <span class="n">mutated_sequences</span><span class="p">,</span>
                <span class="n">reporters</span><span class="o">=</span><span class="n">reporters</span><span class="p">)</span>
        <span class="n">fasta_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="AnalyzeSequences.variant_effect_prediction"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.variant_effect_prediction">[docs]</a>    <span class="k">def</span> <span class="nf">variant_effect_prediction</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                  <span class="n">vcf_file</span><span class="p">,</span>
                                  <span class="n">save_data</span><span class="p">,</span>
                                  <span class="n">output_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                  <span class="n">output_format</span><span class="o">=</span><span class="s2">&quot;tsv&quot;</span><span class="p">,</span>
                                  <span class="n">strand_index</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                  <span class="n">require_strand</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get model predictions and scores for a list of variants.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vcf_file : str</span>
<span class="sd">            Path to a VCF file. Must contain the columns</span>
<span class="sd">            [#CHROM, POS, ID, REF, ALT], in order. Column header does not need</span>
<span class="sd">            to be present.</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_dir : str or None, optional</span>
<span class="sd">            Default is None. Path to the output directory. If no path is</span>
<span class="sd">            specified, will save files corresponding to the options in</span>
<span class="sd">            `save_data` to the current working directory.</span>
<span class="sd">        output_format : {&#39;tsv&#39;, &#39;hdf5&#39;}, optional</span>
<span class="sd">            Default is &#39;tsv&#39;. Choose whether to save TSV or HDF5 output files.</span>
<span class="sd">            TSV is easier to access (i.e. open with text editor/Excel) and</span>
<span class="sd">            quickly peruse, whereas HDF5 files must be accessed through</span>
<span class="sd">            specific packages/viewers that support this format (e.g. h5py</span>
<span class="sd">            Python package). Choose</span>

<span class="sd">                * &#39;tsv&#39; if your list of variants is relatively small</span>
<span class="sd">                  (:math:`10^4` or less in order of magnitude) and/or your</span>
<span class="sd">                  model has a small number of features (&lt;1000).</span>
<span class="sd">                * &#39;hdf5&#39; for anything larger and/or if you would like to</span>
<span class="sd">                  access the predictions/scores as a matrix that you can</span>
<span class="sd">                  easily filter, apply computations, or use in a subsequent</span>
<span class="sd">                  classifier/model. In this case, you may access the matrix</span>
<span class="sd">                  using `mat[&quot;data&quot;]` after opening the HDF5 file using</span>
<span class="sd">                  `mat = h5py.File(&quot;&lt;output.h5&gt;&quot;, &#39;r&#39;)`. The matrix columns</span>
<span class="sd">                  are the features and will match the same ordering as your</span>
<span class="sd">                  features .txt file (same as the order your model outputs</span>
<span class="sd">                  its predictions) and the matrix rows are the sequences.</span>
<span class="sd">                  Note that the row labels (chrom, pos, id, ref, alt) will be</span>
<span class="sd">                  output as a separate .txt file.</span>
<span class="sd">        strand_index : int or None, optional.</span>
<span class="sd">            Default is None. If applicable, specify the column index (0-based)</span>
<span class="sd">            in the VCF file that contains strand information for each variant.</span>
<span class="sd">        require_strand : bool, optional.</span>
<span class="sd">            Default is False. Whether strand can be specified as &#39;.&#39;. If False,</span>
<span class="sd">            Selene accepts strand value to be &#39;+&#39;, &#39;-&#39;, or &#39;.&#39; and automatically</span>
<span class="sd">            treats &#39;.&#39; as &#39;+&#39;. If True, Selene skips any variant with strand &#39;.&#39;.</span>
<span class="sd">            This parameter assumes that `strand_index` has been set.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Saves all files to `output_dir`. If any bases in the &#39;ref&#39; column</span>
<span class="sd">            of the VCF do not match those at the specified position in the</span>
<span class="sd">            reference genome, the row labels .txt file will mark this variant</span>
<span class="sd">            as `ref_match = False`. If most of your variants do not match</span>
<span class="sd">            the reference genome, please check that the reference genome</span>
<span class="sd">            you specified matches the version with which the variants were</span>
<span class="sd">            called. The predictions can used directly if you have verified that</span>
<span class="sd">            the &#39;ref&#39; bases specified for these variants are correct (Selene</span>
<span class="sd">            will have substituted these bases for those in the reference</span>
<span class="sd">            genome). In addition, if any base in the retrieved reference</span>
<span class="sd">            sequence is unknown, the row labels .txt file will mark this variant</span>
<span class="sd">            as `contains_unk = True`. Finally, some variants may show up in an</span>
<span class="sd">            &#39;NA&#39; file. This is because the surrounding sequence context ended up</span>
<span class="sd">            being out of bounds or overlapping with blacklist regions  or the</span>
<span class="sd">            chromosome containing the variant did not show up in the reference</span>
<span class="sd">            genome FASTA file.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: GIVE USER MORE CONTROL OVER PREFIX.</span>
        <span class="n">path</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">)</span>
        <span class="n">output_path_prefix</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">output_dir</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output_dir</span> <span class="o">=</span> <span class="n">path</span>

        <span class="n">output_path_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">output_path_prefix</span><span class="p">)</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="n">read_vcf_file</span><span class="p">(</span>
            <span class="n">vcf_file</span><span class="p">,</span>
            <span class="n">strand_index</span><span class="o">=</span><span class="n">strand_index</span><span class="p">,</span>
            <span class="n">require_strand</span><span class="o">=</span><span class="n">require_strand</span><span class="p">,</span>
            <span class="n">output_NAs_to_file</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">.NA&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">),</span>
            <span class="n">seq_context</span><span class="o">=</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span><span class="p">),</span>
            <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>
        <span class="n">reporters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
            <span class="n">save_data</span><span class="p">,</span>
            <span class="n">output_path_prefix</span><span class="p">,</span>
            <span class="n">output_format</span><span class="p">,</span>
            <span class="n">VARIANTEFFECT_COLS</span><span class="p">,</span>
            <span class="n">output_size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">variants</span><span class="p">),</span>
            <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;varianteffect&quot;</span><span class="p">)</span>

        <span class="n">batch_ref_seqs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">batch_alt_seqs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">t_i</span> <span class="o">=</span> <span class="n">time</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">variants</span><span class="p">):</span>
            <span class="c1"># centers the sequence containing the ref allele based on the size</span>
            <span class="c1"># of ref</span>
            <span class="n">center</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">center</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">center</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span>
            <span class="n">ref_sequence_encoding</span><span class="p">,</span> <span class="n">contains_unk</span> <span class="o">=</span> \
                <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_encoding_from_coords_check_unk</span><span class="p">(</span>
                    <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>

            <span class="n">ref_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span>
            <span class="n">alt_sequence_encoding</span> <span class="o">=</span> <span class="n">_process_alt</span><span class="p">(</span>
                <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
                <span class="n">ref_sequence_encoding</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>

            <span class="n">match</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">seq_at_ref</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="n">match</span><span class="p">,</span> <span class="n">ref_sequence_encoding</span><span class="p">,</span> <span class="n">seq_at_ref</span> <span class="o">=</span> <span class="n">_handle_standard_ref</span><span class="p">(</span>
                    <span class="n">ref_encoding</span><span class="p">,</span>
                    <span class="n">ref_sequence_encoding</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="n">match</span><span class="p">,</span> <span class="n">ref_sequence_encoding</span><span class="p">,</span> <span class="n">seq_at_ref</span> <span class="o">=</span> <span class="n">_handle_long_ref</span><span class="p">(</span>
                    <span class="n">ref_encoding</span><span class="p">,</span>
                    <span class="n">ref_sequence_encoding</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">contains_unk</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;For variant (</span><span class="si">{0}</span><span class="s2">, </span><span class="si">{1}</span><span class="s2">, </span><span class="si">{2}</span><span class="s2">, </span><span class="si">{3}</span><span class="s2">, </span><span class="si">{4}</span><span class="s2">, </span><span class="si">{5}</span><span class="s2">), &quot;</span>
                           <span class="s2">&quot;reference sequence contains unknown base(s)&quot;</span>
                           <span class="s2">&quot;--will be marked `True` in the `contains_unk` column &quot;</span>
                           <span class="s2">&quot;of the .tsv or the row_labels .txt file.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                             <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">match</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;For variant (</span><span class="si">{0}</span><span class="s2">, </span><span class="si">{1}</span><span class="s2">, </span><span class="si">{2}</span><span class="s2">, </span><span class="si">{3}</span><span class="s2">, </span><span class="si">{4}</span><span class="s2">, </span><span class="si">{5}</span><span class="s2">), &quot;</span>
                              <span class="s2">&quot;reference does not match the reference genome. &quot;</span>
                              <span class="s2">&quot;Reference genome contains </span><span class="si">{6}</span><span class="s2"> instead. &quot;</span>
                              <span class="s2">&quot;Predictions/scores associated with this &quot;</span>
                              <span class="s2">&quot;variant--where we use &#39;</span><span class="si">{3}</span><span class="s2">&#39; in the input &quot;</span>
                              <span class="s2">&quot;sequence--will be marked `False` in the `ref_match` &quot;</span>
                              <span class="s2">&quot;column of the .tsv or the row_labels .txt file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                  <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">seq_at_ref</span><span class="p">))</span>
            <span class="n">batch_ids</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">match</span><span class="p">,</span> <span class="n">contains_unk</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
                <span class="n">ref_sequence_encoding</span> <span class="o">=</span> <span class="n">get_reverse_complement_encoding</span><span class="p">(</span>
                    <span class="n">ref_sequence_encoding</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">COMPLEMENTARY_BASE_DICT</span><span class="p">)</span>
                <span class="n">alt_sequence_encoding</span> <span class="o">=</span> <span class="n">get_reverse_complement_encoding</span><span class="p">(</span>
                    <span class="n">alt_sequence_encoding</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">COMPLEMENTARY_BASE_DICT</span><span class="p">)</span>
            <span class="n">batch_ref_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref_sequence_encoding</span><span class="p">)</span>
            <span class="n">batch_alt_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt_sequence_encoding</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">batch_ref_seqs</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">:</span>
                <span class="n">_handle_ref_alt_predictions</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span>
                    <span class="n">batch_ref_seqs</span><span class="p">,</span>
                    <span class="n">batch_alt_seqs</span><span class="p">,</span>
                    <span class="n">batch_ids</span><span class="p">,</span>
                    <span class="n">reporters</span><span class="p">,</span>
                    <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>
                <span class="n">batch_ref_seqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">batch_alt_seqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">if</span> <span class="n">ix</span> <span class="ow">and</span> <span class="n">ix</span> <span class="o">%</span> <span class="mi">10000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;[STEP </span><span class="si">{0}</span><span class="s2">]: </span><span class="si">{1}</span><span class="s2"> s to process 10000 variants.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">ix</span><span class="p">,</span> <span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">t_i</span><span class="p">))</span>
                <span class="n">t_i</span> <span class="o">=</span> <span class="n">time</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">batch_ref_seqs</span><span class="p">:</span>
            <span class="n">_handle_ref_alt_predictions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span>
                <span class="n">batch_ref_seqs</span><span class="p">,</span>
                <span class="n">batch_alt_seqs</span><span class="p">,</span>
                <span class="n">batch_ids</span><span class="p">,</span>
                <span class="n">reporters</span><span class="p">,</span>
                <span class="n">use_cuda</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
            <span class="n">r</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="nf">_pad_or_truncate_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">_pad_sequence</span><span class="p">(</span>
                <span class="n">sequence</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">UNK_BASE</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">_truncate_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">sequence</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>