

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene_sdk.targets.genomic_features &mdash; Selene master documentation</title>
  

  
  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.4.8)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-selene-with-pip">Installing selene with pip:</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene_sdk.targets.genomic_features</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene_sdk.targets.genomic_features</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This class contains methods to query a file of genomic coordinates,</span>
<span class="sd">where each row of [start, end) coordinates corresponds to a genomic feature</span>
<span class="sd">in the sequence.</span>

<span class="sd">It accepts the path to a tabix-indexed .bed.gz file of genomic coordinates.</span>

<span class="sd">This .tsv/.bed file must contain the following columns, in order:</span>
<span class="sd">    chrom (&#39;1&#39;, &#39;2&#39;, ..., &#39;X&#39;, &#39;Y&#39;), start (0-based), end, feature</span>
<span class="sd">Additionally, the column names should be omitted from the file itself</span>
<span class="sd">(i.e. there is no header and the first line in the file is the first</span>
<span class="sd">row of genome coordinates for a feature).</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">types</span>

<span class="kn">import</span> <span class="nn">tabix</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">.target</span> <span class="kn">import</span> <span class="n">Target</span>
<span class="kn">from</span> <span class="nn">._genomic_features</span> <span class="kn">import</span> <span class="n">_fast_get_feature_data</span>


<span class="k">def</span> <span class="nf">_any_positive_rows</span><span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">thresholds</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Searches through a set of feature annotations for positive examples</span>
<span class="sd">    according to a threshold specific to each feature. For each feature</span>
<span class="sd">    in `rows`, the overlap between the feature and the query region must</span>
<span class="sd">    be greater than that feature&#39;s threshold to be considered positive.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    rows : list(tuple(int, int, str)) or None</span>
<span class="sd">        A list of tuples of the form `(start, end, feature_name)`, or</span>
<span class="sd">        `None`.</span>
<span class="sd">    start : int</span>
<span class="sd">        The 0-based start coordinate of the region to query.</span>
<span class="sd">    end : int</span>
<span class="sd">        One past the last coordinate of the region to query.</span>
<span class="sd">    thresholds : dict</span>
<span class="sd">        A dictionary mapping feature names (`str`) to</span>
<span class="sd">        thresholds (`float`), where the threshold is the minimum</span>
<span class="sd">        fraction of a region that must overlap with a label for it to be</span>
<span class="sd">        considered a positive example of that label.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool</span>
<span class="sd">        `True` if there is at least one feature in `rows` that meets its</span>
<span class="sd">        feature-specific cutoff. `False` otherwise, or if `rows==None`.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">rows</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>  <span class="c1"># features within [start, end)</span>
        <span class="n">is_positive</span> <span class="o">=</span> <span class="n">_is_positive_row</span><span class="p">(</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]),</span> <span class="n">thresholds</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]])</span>
        <span class="k">if</span> <span class="n">is_positive</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="kc">False</span>


<span class="k">def</span> <span class="nf">_is_positive_row</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
                     <span class="n">feature_start</span><span class="p">,</span> <span class="n">feature_end</span><span class="p">,</span>
                     <span class="n">threshold</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determine if a feature annotation overlaps enough with a query</span>
<span class="sd">    region to meet a minimal overlap threshold and be considered a</span>
<span class="sd">    positive example of said feature.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    start : int</span>
<span class="sd">        The 0-based start coordinate of the query region.</span>
<span class="sd">    end : int</span>
<span class="sd">        One past the last coordinate of the query region.</span>
<span class="sd">    feature_start : int</span>
<span class="sd">        The 0-based start coordinate of the feature.</span>
<span class="sd">    feature_end : int</span>
<span class="sd">        One past the last coordinate of the feature.</span>
<span class="sd">    threshold : float</span>
<span class="sd">        The minimum fraction of the query region that a feature must</span>
<span class="sd">        overlap with to be considered positive.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool</span>
<span class="sd">        `True` if the feature is a positive example and meets the</span>
<span class="sd">        overlap threshold. Otherwise `False`.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">overlap_start</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">feature_start</span><span class="p">,</span> <span class="n">start</span><span class="p">)</span>
    <span class="n">overlap_end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">feature_end</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
    <span class="n">min_overlap_needed</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span>
        <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span> <span class="o">*</span> <span class="n">threshold</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">min_overlap_needed</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">min_overlap_needed</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">overlap_end</span> <span class="o">-</span> <span class="n">overlap_start</span> <span class="o">&gt;</span> <span class="n">min_overlap_needed</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>


<span class="k">def</span> <span class="nf">_get_feature_data</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
                      <span class="n">thresholds</span><span class="p">,</span> <span class="n">feature_index_dict</span><span class="p">,</span> <span class="n">get_feature_rows</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a target vector for the given query region.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">     chrom : str</span>
<span class="sd">        The name of the region (e.g. &#39;chr1&#39;, &#39;chr2&#39;, ..., &#39;chrX&#39;,</span>
<span class="sd">        &#39;chrY&#39;) to query inside of.</span>
<span class="sd">    start : int</span>
<span class="sd">        The 0-based start coordinate of the region to query.</span>
<span class="sd">    end : int</span>
<span class="sd">        One past the last coordinate of the region to query.</span>
<span class="sd">    thresholds : np.ndarray, dtype=numpy.float32</span>
<span class="sd">        An array of feature thresholds, where the value in position</span>
<span class="sd">        `i` corresponds to the threshold for the feature name that is</span>
<span class="sd">        mapped to index `i` by `feature_index_dict`.</span>
<span class="sd">    feature_index_dict : dict</span>
<span class="sd">        A dictionary mapping feature names (`str`) to indices (`int`),</span>
<span class="sd">        where the index is the position of the feature in `features`.</span>
<span class="sd">    get_feature_rows : types.FunctionType</span>
<span class="sd">        A function that takes coordinates and returns rows</span>
<span class="sd">        (`list(tuple(int, int, str))`).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    numpy.ndarray, dtype=int</span>
<span class="sd">        A target vector where the `i`th position is equal to one if the</span>
<span class="sd">        `i`th feature is positive, and zero otherwise.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="n">get_feature_rows</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">_fast_get_feature_data</span><span class="p">(</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">thresholds</span><span class="p">,</span> <span class="n">feature_index_dict</span><span class="p">,</span> <span class="n">rows</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_define_feature_thresholds</span><span class="p">(</span><span class="n">feature_thresholds</span><span class="p">,</span> <span class="n">features</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Defines the minimal overlap thresholds for the various features.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    feature_thresholds : float or dict or type.FunctionType</span>
<span class="sd">        Either a function that takes a feature name (`str`) and returns</span>
<span class="sd">        a threshold (`float`) or a constant `float` that will be used</span>
<span class="sd">        for all features.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        A list of feature names.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    feature_thresholds_dict, feature_thresholds_vec : \</span>
<span class="sd">    tuple(dict, numpy.ndarray)</span>
<span class="sd">        A tuple. The first element, `feature_thresholds_dict`, is a</span>
<span class="sd">        dictionary that maps feature names (`str`) to thresholds</span>
<span class="sd">        (`float`). The second element, `feature_thresholds_vec`, is an</span>
<span class="sd">        array of the thresholds (numpy.`float32`) where the `i`th value</span>
<span class="sd">        corresponds to the threshold for the `i`th feature from</span>
<span class="sd">        the `features` input.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">feature_thresholds_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">feature_thresholds_vec</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">features</span><span class="p">))</span>
    <span class="k">if</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">feature_thresholds</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="ow">or</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="n">feature_thresholds</span><span class="p">,</span> <span class="nb">int</span><span class="p">))</span> <span class="ow">and</span> <span class="n">feature_thresholds</span> <span class="o">&lt;=</span> <span class="mi">1</span> \
            <span class="ow">and</span> <span class="n">feature_thresholds</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">feature_thresholds_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="n">feature_thresholds</span><span class="p">)</span>
        <span class="n">feature_thresholds_vec</span> <span class="o">+=</span> <span class="n">feature_thresholds</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">feature_thresholds</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="c1"># assign the default value to everything first</span>
        <span class="n">feature_thresholds_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span>
            <span class="n">features</span><span class="p">,</span> <span class="n">feature_thresholds</span><span class="p">[</span><span class="s2">&quot;default&quot;</span><span class="p">])</span>
        <span class="n">feature_thresholds_vec</span> <span class="o">+=</span> <span class="n">feature_thresholds</span><span class="p">[</span><span class="s2">&quot;default&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">feature_thresholds</span><span class="p">:</span>
                <span class="n">feature_thresholds_dict</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature_thresholds</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
                <span class="n">feature_thresholds_vec</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature_thresholds</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
    <span class="c1"># this branch will not be accessed if you use a config.yml file to</span>
    <span class="c1"># specify input parameters</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">feature_thresholds</span><span class="p">,</span> <span class="n">types</span><span class="o">.</span><span class="n">FunctionType</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features</span><span class="p">):</span>
            <span class="n">threshold</span> <span class="o">=</span> <span class="n">feature_thresholds</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">feature_thresholds_dict</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">threshold</span>
            <span class="n">feature_thresholds_vec</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">threshold</span>
    <span class="n">feature_thresholds_vec</span> <span class="o">=</span> <span class="n">feature_thresholds_vec</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">feature_thresholds_dict</span><span class="p">,</span> <span class="n">feature_thresholds_vec</span>


<div class="viewcode-block" id="GenomicFeatures"><a class="viewcode-back" href="../../../targets.html#selene_sdk.targets.GenomicFeatures">[docs]</a><span class="k">class</span> <span class="nc">GenomicFeatures</span><span class="p">(</span><span class="n">Target</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Stores the dataset specifying sequence regions and features.</span>
<span class="sd">    Accepts a tabix-indexed `*.bed` file with the following columns,</span>
<span class="sd">    in order:</span>
<span class="sd">    ::</span>
<span class="sd">        [chrom, start, end, strand, feature]</span>


<span class="sd">    Note that `chrom` is interchangeable with any sort of region (e.g.</span>
<span class="sd">    a protein in a FAA file). Further, `start` is 0-based. Lastly, any</span>
<span class="sd">    addition columns following the five shown above will be ignored.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input_path : str</span>
<span class="sd">        Path to the tabix-indexed dataset. Note that for the file to</span>
<span class="sd">        be tabix-indexed, it must have been compressed with `bgzip`.</span>
<span class="sd">        Thus, `input_path` should be a `*.gz` file with a</span>
<span class="sd">        corresponding `*.tbi` file in the same directory.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        The non-redundant list of genomic features (i.e. labels)</span>
<span class="sd">        that will be predicted.</span>
<span class="sd">    feature_thresholds : float or dict or types.FunctionType or None</span>
<span class="sd">        Default is None. A genomic region is determined to be a</span>
<span class="sd">        positive sample if at least one genomic feature peak takes</span>
<span class="sd">        up a proportion of the region greater than or equal to</span>
<span class="sd">        the threshold specified for that feature.</span>

<span class="sd">        * `None` - No thresholds specified. All features found in\</span>
<span class="sd">                   a query region are annotated to that region.</span>
<span class="sd">        * `float` - A single threshold applies to all the features\</span>
<span class="sd">                    in the dataset.</span>
<span class="sd">        * `dict` - A dictionary mapping feature names (`str`) to \</span>
<span class="sd">                   threshold values (`float`), which thereby assigns\</span>
<span class="sd">                   different thresholds to different features. If a\</span>
<span class="sd">                   feature&#39;s threshold is not specified in this \</span>
<span class="sd">                   dictionary, then we assume that a key `&quot;default&quot;`\</span>
<span class="sd">                   exists in the dictionary that has the default \</span>
<span class="sd">                   threshold value we should assign to the feature \</span>
<span class="sd">                   name that is absent from the dictionary keys.</span>
<span class="sd">        * `types.FunctionType` - define a function that takes as \</span>
<span class="sd">                                 input the feature name and returns\</span>
<span class="sd">                                 the feature&#39;s threshold.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    data : tabix.open</span>
<span class="sd">        The data stored in a tabix-indexed `*.bed` file.</span>
<span class="sd">    n_features : int</span>
<span class="sd">        The number of distinct features.</span>
<span class="sd">    feature_index_dict : dict</span>
<span class="sd">        A dictionary mapping feature names (`str`) to indices (`int`),</span>
<span class="sd">        where the index is the position of the feature in `features`.</span>
<span class="sd">    index_feature_dict : dict</span>
<span class="sd">        A dictionary mapping indices (`int`) to feature names (`str`),</span>
<span class="sd">        where the index is the position of the feature in the input</span>
<span class="sd">        features.</span>
<span class="sd">    feature_thresholds : dict or None</span>

<span class="sd">        * `dict` - A dictionary mapping feature names (`str`) to thresholds\</span>
<span class="sd">        (`float`), where the threshold is the minimum overlap that a\</span>
<span class="sd">        feature annotation must have with a query region to be\</span>
<span class="sd">        considered a positive example of that feature.</span>
<span class="sd">        * `None` - No threshold specifications. Assumes that all features\</span>
<span class="sd">        returned by a tabix query are annotated to the query region.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">,</span> <span class="n">features</span><span class="p">,</span> <span class="n">feature_thresholds</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructs a new `GenomicFeatures` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">tabix</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">n_features</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">feature_index_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">[(</span><span class="n">feat</span><span class="p">,</span> <span class="n">index</span><span class="p">)</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">feat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features</span><span class="p">)])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">index_feature_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">features</span><span class="p">)))</span>

        <span class="k">if</span> <span class="n">feature_thresholds</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">feature_thresholds</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_feature_thresholds_vec</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">feature_thresholds</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_feature_thresholds_vec</span> <span class="o">=</span> \
                <span class="n">_define_feature_thresholds</span><span class="p">(</span><span class="n">feature_thresholds</span><span class="p">,</span> <span class="n">features</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_query_tabix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Queries a tabix-indexed `*.bed` file for features falling into</span>
<span class="sd">        the specified region.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            The name of the region (e.g. &#39;1&#39;, &#39;2&#39;, ..., &#39;X&#39;, &#39;Y&#39;) to</span>
<span class="sd">            query in.</span>
<span class="sd">        start : int</span>
<span class="sd">            The 0-based start position of the query coordinates.</span>
<span class="sd">        end : int</span>
<span class="sd">            One past the last position of the query coordinates.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list(list(str)) or None</span>
<span class="sd">            A list, wherein each sub-list corresponds to a line from the</span>
<span class="sd">            tabix-indexed file, and each value in a sub-list corresponds</span>
<span class="sd">            to a column in that row. If a `tabix.TabixError` is caught,</span>
<span class="sd">            we assume it was because there were no features present in</span>
<span class="sd">            the query region, and return `None`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">tabix</span><span class="o">.</span><span class="n">TabixError</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

<div class="viewcode-block" id="GenomicFeatures.is_positive"><a class="viewcode-back" href="../../../targets.html#selene_sdk.targets.GenomicFeatures.is_positive">[docs]</a>    <span class="k">def</span> <span class="nf">is_positive</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determines whether the query the `chrom` queried contains any</span>
<span class="sd">        genomic features within the :math:`[start, end)` region. If so,</span>
<span class="sd">        the query is considered positive.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            The name of the region (e.g. &#39;1&#39;, &#39;2&#39;, ..., &#39;X&#39;, &#39;Y&#39;).</span>
<span class="sd">        start : int</span>
<span class="sd">            The 0-based first position in the region.</span>
<span class="sd">        end : int</span>
<span class="sd">            One past the 0-based last position in the region.</span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bool</span>
<span class="sd">            `True` if this meets the criterion for a positive example,</span>
<span class="sd">            `False` otherwise.</span>
<span class="sd">            Note that if we catch a `tabix.TabixError` exception, we</span>
<span class="sd">            assume the error was the result of no features being present</span>
<span class="sd">            in the queried region and return `False`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">rows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_query_tabix</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">_any_positive_rows</span><span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_thresholds</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicFeatures.get_feature_data"><a class="viewcode-back" href="../../../targets.html#selene_sdk.targets.GenomicFeatures.get_feature_data">[docs]</a>    <span class="k">def</span> <span class="nf">get_feature_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        For a sequence of length :math:`L = end - start`, return the</span>
<span class="sd">        features&#39; one-hot encoding corresponding to that region. For</span>
<span class="sd">        instance, for `n_features`, each position in that sequence will</span>
<span class="sd">        have a binary vector specifying whether the genomic feature&#39;s</span>
<span class="sd">        coordinates overlap with that position.</span>
<span class="sd">        @TODO: Clarify with an example, as this is hard to read right now.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            The name of the region (e.g. &#39;1&#39;, &#39;2&#39;, ..., &#39;X&#39;, &#39;Y&#39;).</span>
<span class="sd">        start : int</span>
<span class="sd">            The 0-based first position in the region.</span>
<span class="sd">        end : int</span>
<span class="sd">            One past the 0-based last position in the region.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        numpy.ndarray</span>
<span class="sd">            :math:`L \\times N` array, where :math:`L = end - start`</span>
<span class="sd">            and :math:`N =` `self.n_features`. Note that if we catch a</span>
<span class="sd">            `tabix.TabixError`, we assume the error was the result of</span>
<span class="sd">            there being no features present in the queried region and</span>
<span class="sd">            return a `numpy.ndarray` of zeros.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_feature_thresholds_vec</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">features</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">))</span>
            <span class="n">rows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_query_tabix</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">rows</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">features</span>
            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
                <span class="n">feature</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
                <span class="n">ix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_index_map</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span>
                <span class="n">features</span><span class="p">[</span><span class="n">ix</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="n">features</span>
        <span class="k">return</span> <span class="n">_get_feature_data</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_feature_thresholds_vec</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">feature_index_dict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_query_tabix</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>