

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>selene_sdk.interpret.vis &mdash; Selene 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-selene-with-pip">Installing selene with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene_sdk.interpret.vis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene_sdk.interpret.vis</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module provides the methods for visualizing different ouputs</span>
<span class="sd">from selene analysis methods.</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">copy</span> <span class="kn">import</span> <span class="n">deepcopy</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">transforms</span>
<span class="kn">from</span> <span class="nn">matplotlib.path</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">matplotlib.patches</span> <span class="kn">import</span> <span class="n">PathPatch</span>
<span class="kn">import</span> <span class="nn">matplotlib.patheffects</span>
<span class="kn">from</span> <span class="nn">matplotlib.text</span> <span class="kn">import</span> <span class="n">TextPath</span>
<span class="kn">import</span> <span class="nn">pkg_resources</span>
<span class="kn">from</span> <span class="nn">plotly.offline</span> <span class="kn">import</span> <span class="n">download_plotlyjs</span><span class="p">,</span> <span class="n">plot</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objs</span> <span class="k">as</span> <span class="nn">go</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">tabix</span>

<span class="kn">from</span> <span class="nn">..sequences</span> <span class="kn">import</span> <span class="n">Genome</span>


<span class="n">_SVG_PATHS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="s2">&quot;M 0,100 l 100, 0 l 0,-25 l -37.5, 0 l 0,-75 l -25, 0 &quot;</span> <span class="o">+</span>
                   <span class="s2">&quot;l 0,75 l -37.5,0 l 0,25 z&quot;</span><span class="p">,</span>
              <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;M 100,12.5 l 0,25 c 0,0 -25,-15 -50,-12.5 &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;c 0,0 -25,0 -25,25 c 0,0 0,25 25,25 c 0,0 25,2.5 50,-15&quot;</span> <span class="o">+</span>
                    <span class="s2">&quot; l 0, 25 C 100,87.5 75,100 50,100 C 50,100 0,100 0,50 &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;C 0,50 0,0 50,0 C 50,0 75,0 100,12.5 z&quot;</span><span class="p">),</span>
              <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;M 100,12.5 l 0,25 c 0,0 -25,-15 -50,-12.5 &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;c 0,0 -25,0 -25,25 c 0,0 0,25 25,25 c 0,0 25,2.5 50,-15&quot;</span> <span class="o">+</span>
                    <span class="s2">&quot; l 0, 25 C 100,87.5 75,100 50,100 C 50,100 0,100 0,50 &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;C 0,50 0,0 50,0 C 50,0 75,0 100,12.5 M 100,37.5 &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;l 0,17.5 l -50,0 l 0,-17 l 25,0 l 0,-25 l 25,0 z&quot;</span><span class="p">),</span>
              <span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;M 0,0 l 37.5,100 l 25,0 l 37.5,-100 l -25,0 l -9.375,25&quot;</span> <span class="o">+</span>
                    <span class="s2">&quot; l -31.25,0 l -9.375,-25 l -25,0 z 0,0 M 43.75, 50 &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;l 12.5,0 l -5.859375,15.625 l -5.859375,-15.625 z&quot;</span><span class="p">),</span>
              <span class="s1">&#39;U&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;M 0,100 l 25,0 l 0,-50 C 25,50 25,25 50,25&quot;</span> <span class="o">+</span>
                    <span class="s2">&quot; C 50,25 75,25 75,50 l 0,50 l 25,0 L 100,50 &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;C 100,50 100,0, 50,0 C 50,0 0,0 0,50 l 0,50 z&quot;</span><span class="p">)}</span>


<span class="k">def</span> <span class="nf">_svg_parse</span><span class="p">(</span><span class="n">path_string</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Functionality for parsing a string from source vector graphics.</span>

<span class="sd">    Source is from `matplotlib.org/2.1.1/gallery/showcase/firefox.html`</span>
<span class="sd">    with minor modifications.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    path_string : str</span>
<span class="sd">        String containing the path code from an SVG file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list(numpy.uint8), numpy.ndarray, dtype=np.float32</span>
<span class="sd">        A 2-tuple containing code types and coordinates for a matplotlib</span>
<span class="sd">        path.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">commands</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">Path</span><span class="o">.</span><span class="n">MOVETO</span><span class="p">,),</span>
                <span class="s1">&#39;L&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">Path</span><span class="o">.</span><span class="n">LINETO</span><span class="p">,),</span>
                <span class="s1">&#39;Q&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">Path</span><span class="o">.</span><span class="n">CURVE3</span><span class="p">,)</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span>
                <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">Path</span><span class="o">.</span><span class="n">CURVE4</span><span class="p">,)</span><span class="o">*</span><span class="mi">3</span><span class="p">,</span>
                <span class="s1">&#39;Z&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">Path</span><span class="o">.</span><span class="n">CLOSEPOLY</span><span class="p">,)}</span>
    <span class="n">path_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;([MLHVCSQTAZ])([^MLHVCSQTAZ]+)&#39;</span><span class="p">,</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>
    <span class="n">float_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(?:[\s,]*)([+-]?\d+(?:\.\d+)?)&#39;</span><span class="p">)</span>
    <span class="n">vertices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">codes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">last</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">cmd</span><span class="p">,</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">path_re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">path_string</span><span class="p">):</span>
        <span class="n">points</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">float_re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">values</span><span class="p">)]</span>
        <span class="n">points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">points</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">points</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">cmd</span><span class="o">.</span><span class="n">islower</span><span class="p">():</span>
            <span class="n">points</span> <span class="o">+=</span> <span class="n">last</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="n">cmd</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">points</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">last</span> <span class="o">=</span> <span class="n">points</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">codes</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">commands</span><span class="p">[</span><span class="n">cmd</span><span class="p">])</span>
        <span class="n">vertices</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">points</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">vertices</span><span class="p">),</span> <span class="n">codes</span>


<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">_SVG_PATHS</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">_SVG_PATHS</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">_svg_parse</span><span class="p">(</span><span class="n">_SVG_PATHS</span><span class="p">[</span><span class="n">k</span><span class="p">])</span>


<span class="k">class</span> <span class="nc">_TextPathRenderingEffect</span><span class="p">(</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">patheffects</span><span class="o">.</span><span class="n">AbstractPathEffect</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;This class provides an effect for continuously rendering a text</span>
<span class="sd">    path over another path.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bar</span><span class="p">,</span> <span class="n">x_translation</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span> <span class="n">y_translation</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span>
                 <span class="n">x_scale</span><span class="o">=</span><span class="mf">1.</span><span class="p">,</span> <span class="n">y_scale</span><span class="o">=</span><span class="mf">1.</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;This is a class for re-rendering text paths and preserving</span>
<span class="sd">        their scale.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        bar : matplotlib.patches.Patch</span>
<span class="sd">            The patch where the letter is.</span>
<span class="sd">        x_translation : float, optional</span>
<span class="sd">            Default is 0. Amount by which to translate the x coordinate.</span>
<span class="sd">        y_translation : float, optional</span>
<span class="sd">            Default is 0. Amount by which to translate the y coordinate.</span>
<span class="sd">        x_scale : float, optional</span>
<span class="sd">            Default is 1. Amount by which to scale the width.</span>
<span class="sd">        y_scale : float, optional</span>
<span class="sd">            Default is 1. Amount by which to scale the height.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_bar</span> <span class="o">=</span> <span class="n">bar</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_x_translation</span> <span class="o">=</span> <span class="n">x_translation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_y_translation</span> <span class="o">=</span> <span class="n">y_translation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_x_scale</span> <span class="o">=</span> <span class="n">x_scale</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_y_scale</span> <span class="o">=</span> <span class="n">y_scale</span>

    <span class="k">def</span> <span class="nf">draw_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">renderer</span><span class="p">,</span> <span class="n">gc</span><span class="p">,</span> <span class="n">tpath</span><span class="p">,</span> <span class="n">affine</span><span class="p">,</span> <span class="n">rgbFace</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Redraws the path.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">b_x</span><span class="p">,</span> <span class="n">b_y</span><span class="p">,</span> <span class="n">b_w</span><span class="p">,</span> <span class="n">b_h</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bar</span><span class="o">.</span><span class="n">get_extents</span><span class="p">()</span><span class="o">.</span><span class="n">bounds</span>
        <span class="n">t_x</span><span class="p">,</span> <span class="n">t_y</span><span class="p">,</span> <span class="n">t_w</span><span class="p">,</span> <span class="n">t_h</span> <span class="o">=</span> <span class="n">tpath</span><span class="o">.</span><span class="n">get_extents</span><span class="p">()</span><span class="o">.</span><span class="n">bounds</span>
        <span class="n">translation</span> <span class="o">=</span> <span class="p">[</span><span class="n">b_x</span> <span class="o">-</span> <span class="n">t_x</span><span class="p">,</span> <span class="n">b_y</span> <span class="o">-</span> <span class="n">t_y</span><span class="p">]</span>
        <span class="n">translation</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_x_translation</span>
        <span class="n">translation</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_y_translation</span>
        <span class="n">scale</span> <span class="o">=</span> <span class="p">[</span><span class="n">b_w</span> <span class="o">/</span> <span class="n">t_w</span><span class="p">,</span> <span class="n">b_h</span> <span class="o">/</span> <span class="n">t_h</span><span class="p">]</span>
        <span class="n">scale</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_x_scale</span>
        <span class="n">scale</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_y_scale</span>
        <span class="n">affine</span> <span class="o">=</span> <span class="n">affine</span><span class="o">.</span><span class="n">identity</span><span class="p">()</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="o">*</span><span class="n">scale</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="o">*</span><span class="n">translation</span><span class="p">)</span>
        <span class="n">renderer</span><span class="o">.</span><span class="n">draw_path</span><span class="p">(</span><span class="n">gc</span><span class="p">,</span> <span class="n">tpath</span><span class="p">,</span> <span class="n">affine</span><span class="p">,</span> <span class="n">rgbFace</span><span class="p">)</span>


<div class="viewcode-block" id="sequence_logo"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.sequence_logo">[docs]</a><span class="k">def</span> <span class="nf">sequence_logo</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="s2">&quot;value&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                  <span class="n">sequence_type</span><span class="o">=</span><span class="n">Genome</span><span class="p">,</span> <span class="n">font_properties</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                  <span class="n">color_scheme</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                  <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Plots a sequence logo for visualizing motifs.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    score_matrix : np.ndarray</span>
<span class="sd">        An :math:`L \\times N` array (where :math:`L` is the length of</span>
<span class="sd">        the sequence, and :math:`N` is the size of the alphabet)</span>
<span class="sd">        containing the scores for each base occuring at each position.</span>
<span class="sd">    order : {&#39;alpha&#39;, &#39;value&#39;}</span>
<span class="sd">        The manner by which to sort the bases stacked at each position</span>
<span class="sd">        in the sequence logo plot.</span>

<span class="sd">            * &#39;alpha&#39; - Bases go in the order they are found in the\</span>
<span class="sd">                      sequence alphabet.</span>
<span class="sd">            * &#39;value&#39; - Bases go in the order of their value, with the\</span>
<span class="sd">                        largest at the bottom.</span>
<span class="sd">    width : float, optional</span>
<span class="sd">        Default is 1. The width of each character in the plotted logo.</span>
<span class="sd">        A value of 1 will mean that there is no gap between each the</span>
<span class="sd">        characters at each position. A value of 0 will not draw any</span>
<span class="sd">        characters.</span>
<span class="sd">    ax : matplotlib.pyplot.Axes or None, optional</span>
<span class="sd">        Default is `None`. The axes to plot on. If left as `None`, a new</span>
<span class="sd">        axis will be created.</span>
<span class="sd">    sequence_type : class, optional</span>
<span class="sd">        Default is `selene_sdk.sequences.Genome`. The type of sequence that</span>
<span class="sd">        the *in silico* mutagenesis results are associated with. This</span>
<span class="sd">        should generally be a subclass of `selene_sdk.sequences.Sequence`.</span>
<span class="sd">    font_properties : matplotlib.font_manager.FontProperties or None, optional</span>
<span class="sd">        Default is `None`. A `matplotlib.font_manager.FontProperties`</span>
<span class="sd">        object that specifies the properties of the font to use for</span>
<span class="sd">        plotting the motif. If `None`, no font will be used, and the</span>
<span class="sd">        text will be rendered by a path. This method of rendering paths</span>
<span class="sd">        is  preferred, as it ensures all character heights correspond to</span>
<span class="sd">        the actual values, and that there are no extra gaps between</span>
<span class="sd">        the tops and bottoms of characters at each position in the</span>
<span class="sd">        sequence logo. If the user opts to use a value other</span>
<span class="sd">        than `None`, then no such guarantee can be made.</span>
<span class="sd">    color_scheme : list(str) or None, optional</span>
<span class="sd">        Default is `None`. A list containing the hex codes or names of</span>
<span class="sd">        colors to use, appearing in the order of the bases of the</span>
<span class="sd">        sequence type. If left as `None`, a default palette will be made</span>
<span class="sd">        with `seaborn.color_palette`, and will have as many</span>
<span class="sd">        colors as there are characters in the input sequence alphabet.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    matplotlib.pyplot.Axes</span>
<span class="sd">        The axes containing the sequence logo plot.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the number of columns in `score_matrix` does not match the</span>
<span class="sd">        number of characters in the alphabet of `sequence_type`.</span>

<span class="sd">    ValueError</span>
<span class="sd">        If the number of colors in `color_palette` does not match the</span>
<span class="sd">        number of characters in the alphabet of `sequence_type`.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    We have included an example of the output from a `sequence_logo`</span>
<span class="sd">    plot below:</span>

<span class="sd">    .. image:: ../../docs/source/_static/img/sequence_logo_example.png</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Note that everything will break if we do not deepcopy.</span>
    <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">)</span>

    <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">font_properties</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Specifying a value for `font_properties` (other than `None`) &quot;</span>
            <span class="s2">&quot;will use the `matplotlib`-based character paths, and causes &quot;</span>
            <span class="s2">&quot;distortions in the plotted motif. We recommend leaving &quot;</span>
            <span class="s2">&quot;`font_properties=None`. See the documentation for details.&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">color_scheme</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">color_scheme</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;Set1&quot;</span><span class="p">,</span>
                                         <span class="n">n_colors</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">))</span>
        <span class="n">color_scheme</span> <span class="o">=</span> <span class="n">color_scheme</span><span class="o">.</span><span class="n">as_hex</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">color_scheme</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Color scheme is shorter than number of bases in sequence.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Got score with </span><span class="si">{0}</span><span class="s2"> bases for sequence with </span><span class="si">{1}</span><span class="s2"> bases.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)))</span>
    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

    <span class="c1"># Determine offsets depending on sort order.</span>
    <span class="n">positive_offsets</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">)</span>
    <span class="n">negative_offsets</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">)</span>
    <span class="n">bases</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>
    <span class="n">bases</span><span class="p">[:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="s2">&quot;?&quot;</span>  <span class="c1"># This ensures blanks are visually obvious.</span>

    <span class="c1"># Change ordering of things based on input arguments.</span>
    <span class="k">if</span> <span class="n">order</span> <span class="o">==</span> <span class="s2">&quot;alpha&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">bases</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span>

    <span class="k">elif</span> <span class="n">order</span> <span class="o">==</span> <span class="s2">&quot;value&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">score_matrix</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">sorted_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="c1"># Sort the negative values and put them at bottom.</span>
                <span class="n">div</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.</span><span class="p">)</span>
                <span class="n">negative_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.</span><span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
                <span class="n">negative_sort_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">[</span><span class="n">negative_idx</span><span class="p">,</span> <span class="n">j</span><span class="p">],</span>
                                               <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">sorted_scores</span><span class="p">[:</span><span class="n">div</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="p">[</span>
                    <span class="n">negative_idx</span><span class="p">[</span><span class="n">negative_sort_idx</span><span class="p">],</span> <span class="n">j</span><span class="p">]</span>
                <span class="n">bases</span><span class="p">[:</span><span class="n">div</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)[</span>
                    <span class="n">negative_idx</span><span class="p">[</span><span class="n">negative_sort_idx</span><span class="p">]]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

                <span class="c1"># Sort the positive values and stack atop the negatives.</span>
                <span class="n">positive_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.</span><span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
                <span class="n">positive_sort_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">[</span><span class="n">positive_idx</span><span class="p">,</span> <span class="n">j</span><span class="p">],</span>
                                               <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">sorted_scores</span><span class="p">[</span><span class="n">div</span><span class="p">:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="p">[</span>
                    <span class="n">positive_idx</span><span class="p">[</span><span class="n">positive_sort_idx</span><span class="p">],</span> <span class="n">j</span><span class="p">]</span>
                <span class="n">bases</span><span class="p">[</span><span class="n">div</span><span class="p">:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)[</span>
                    <span class="n">positive_idx</span><span class="p">[</span><span class="n">positive_sort_idx</span><span class="p">]]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
            <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">sorted_scores</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="n">sort_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">[:,</span> <span class="n">j</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">bases</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)[</span><span class="n">sort_idx</span><span class="p">]</span>
                <span class="n">score_matrix</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="p">[</span><span class="n">sort_idx</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>

    <span class="c1"># Create offsets for each bar.</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">y_coords</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">negative_offsets</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">negative_offsets</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span>
            <span class="n">positive_offsets</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">positive_offsets</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span>
        <span class="n">neg_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">y_coords</span> <span class="o">&lt;</span> <span class="mf">0.</span><span class="p">)</span>
        <span class="n">pos_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">y_coords</span> <span class="o">&gt;=</span> <span class="mf">0.</span><span class="p">)</span>
        <span class="n">negative_offsets</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">neg_idx</span><span class="p">]</span> <span class="o">+=</span> <span class="n">y_coords</span><span class="p">[</span><span class="n">neg_idx</span><span class="p">]</span>
        <span class="n">positive_offsets</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">pos_idx</span><span class="p">]</span> <span class="o">+=</span> <span class="n">y_coords</span><span class="p">[</span><span class="n">pos_idx</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">x_coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mf">0.5</span>
        <span class="n">y_coords</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span>

        <span class="c1"># Manage negatives and positives separately.</span>
        <span class="n">offsets</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">negative_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">y_coords</span> <span class="o">&lt;</span> <span class="mf">0.</span><span class="p">)</span>
        <span class="n">positive_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">y_coords</span> <span class="o">&gt;=</span> <span class="mf">0.</span><span class="p">)</span>
        <span class="n">offsets</span><span class="p">[</span><span class="n">negative_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">negative_offsets</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">negative_idx</span><span class="p">]</span>
        <span class="n">offsets</span><span class="p">[</span><span class="n">positive_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">positive_offsets</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">positive_idx</span><span class="p">]</span>
        <span class="n">bars</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x_coords</span><span class="p">,</span> <span class="n">y_coords</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span>
                      <span class="n">bottom</span><span class="o">=</span><span class="n">offsets</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">bar</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bars</span><span class="p">):</span>
            <span class="n">base</span> <span class="o">=</span> <span class="n">bases</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
            <span class="n">bar</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="n">color_scheme</span><span class="p">[</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASE_TO_INDEX</span><span class="p">[</span><span class="n">base</span><span class="p">]])</span>
            <span class="n">bar</span><span class="o">.</span><span class="n">set_edgecolor</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="c1"># Iterate over the barplot&#39;s bars and turn them into letters.</span>
    <span class="n">new_patches</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bar</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">patches</span><span class="p">):</span>
        <span class="n">base_idx</span> <span class="o">=</span> <span class="n">i</span> <span class="o">//</span> <span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">seq_idx</span> <span class="o">=</span> <span class="n">i</span> <span class="o">%</span> <span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">base</span> <span class="o">=</span> <span class="n">bases</span><span class="p">[</span><span class="n">base_idx</span><span class="p">,</span> <span class="n">seq_idx</span><span class="p">]</span>
        <span class="c1"># We construct a text path that tracks the bars in the barplot.</span>
        <span class="c1"># Thus, the barplot takes care of scaling and translation,</span>
        <span class="c1">#  and we just copy it.</span>
        <span class="k">if</span> <span class="n">font_properties</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">text</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">_SVG_PATHS</span><span class="p">[</span><span class="n">base</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">_SVG_PATHS</span><span class="p">[</span><span class="n">base</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">text</span> <span class="o">=</span> <span class="n">TextPath</span><span class="p">((</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">0.</span><span class="p">),</span> <span class="n">base</span><span class="p">,</span> <span class="n">fontproperties</span><span class="o">=</span><span class="n">font_properties</span><span class="p">)</span>
        <span class="n">b_x</span><span class="p">,</span> <span class="n">b_y</span><span class="p">,</span> <span class="n">b_w</span><span class="p">,</span> <span class="n">b_h</span> <span class="o">=</span> <span class="n">bar</span><span class="o">.</span><span class="n">get_extents</span><span class="p">()</span><span class="o">.</span><span class="n">bounds</span>
        <span class="n">t_x</span><span class="p">,</span> <span class="n">t_y</span><span class="p">,</span> <span class="n">t_w</span><span class="p">,</span> <span class="n">t_h</span> <span class="o">=</span> <span class="n">text</span><span class="o">.</span><span class="n">get_extents</span><span class="p">()</span><span class="o">.</span><span class="n">bounds</span>
        <span class="n">scale</span> <span class="o">=</span> <span class="p">(</span><span class="n">b_w</span> <span class="o">/</span> <span class="n">t_w</span><span class="p">,</span> <span class="n">b_h</span> <span class="o">/</span> <span class="n">t_h</span><span class="p">)</span>
        <span class="n">translation</span> <span class="o">=</span> <span class="p">(</span><span class="n">b_x</span> <span class="o">-</span> <span class="n">t_x</span><span class="p">,</span> <span class="n">b_y</span> <span class="o">-</span> <span class="n">t_y</span><span class="p">)</span>
        <span class="n">text</span> <span class="o">=</span> <span class="n">PathPatch</span><span class="p">(</span><span class="n">text</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="n">bar</span><span class="o">.</span><span class="n">get_facecolor</span><span class="p">(),</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.</span><span class="p">)</span>
        <span class="n">bar</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">(</span><span class="s2">&quot;none&quot;</span><span class="p">)</span>
        <span class="n">text</span><span class="o">.</span><span class="n">set_path_effects</span><span class="p">([</span><span class="n">_TextPathRenderingEffect</span><span class="p">(</span><span class="n">bar</span><span class="p">)])</span>
        <span class="n">transform</span> <span class="o">=</span> <span class="n">transforms</span><span class="o">.</span><span class="n">Affine2D</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="o">*</span><span class="n">translation</span><span class="p">)</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="o">*</span><span class="n">scale</span><span class="p">)</span>
        <span class="n">text</span><span class="o">.</span><span class="n">set_transform</span><span class="p">(</span><span class="n">transform</span><span class="p">)</span>
        <span class="n">new_patches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">text</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">patch</span> <span class="ow">in</span> <span class="n">new_patches</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">patch</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">score_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="k">return</span> <span class="n">ax</span></div>


<div class="viewcode-block" id="rescale_score_matrix"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.rescale_score_matrix">[docs]</a><span class="k">def</span> <span class="nf">rescale_score_matrix</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">base_scaling</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">,</span>
                         <span class="n">position_scaling</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Performs base-wise and position-wise scaling of a score matrix for a</span>
<span class="sd">    feature, usually produced from an *in silico* mutagenesis experiment.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    score_matrix : numpy.ndarray</span>
<span class="sd">        An :math:`L \\times N` matrix containing the scores for each</span>
<span class="sd">        position, where :math:`L` is the length of the sequence, and</span>
<span class="sd">        :math:`N` is the number of characters in the alphabet.</span>
<span class="sd">    base_scaling : {&#39;identity&#39;, &#39;probability&#39;, &#39;max_effect&#39;}</span>
<span class="sd">        The type of scaling performed on each base at a given position.</span>

<span class="sd">            * &#39;identity&#39; - No transformation will be applied to the\</span>
<span class="sd">                           data.</span>
<span class="sd">            * &#39;probability&#39; - The relative sizes of the bases will be\</span>
<span class="sd">                              the original input probabilities.</span>
<span class="sd">            * &#39;max_effect&#39; - The relative sizes of the bases will be\</span>
<span class="sd">                             the max effect of the original input\</span>
<span class="sd">                             values.</span>
<span class="sd">    position_scaling : {&#39;identity&#39;, &#39;probability&#39;, &#39;max_effect&#39;}</span>
<span class="sd">        The type of scaling performed on each position.</span>

<span class="sd">            * &#39;identity&#39;    - No transformation will be applied to the data.</span>
<span class="sd">            * &#39;probability&#39; - The sum of values at a position will be\</span>
<span class="sd">                              equal to the sum of the original input\</span>
<span class="sd">                              values at that position.</span>
<span class="sd">            * &#39;max_effect&#39;  - The sum of values at a position will be\</span>
<span class="sd">                              equal to the sum of the max effect values\</span>
<span class="sd">                              of the original input values at that\</span>
<span class="sd">                              position.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    numpy.ndarray</span>
<span class="sd">        The transformed score matrix.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If an unsupported `base_scaling` or `position_scaling` is</span>
<span class="sd">        entered.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Note that things can break if we do not deepcopy.</span>
    <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">)</span>

    <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="n">rescaled_scores</span> <span class="o">=</span> <span class="n">score_matrix</span>

    <span class="c1"># Scale individual bases.</span>
    <span class="k">if</span> <span class="n">base_scaling</span> <span class="o">==</span> <span class="s2">&quot;identity&quot;</span> <span class="ow">or</span> <span class="n">base_scaling</span> <span class="o">==</span> <span class="s2">&quot;probability&quot;</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">elif</span> <span class="n">base_scaling</span> <span class="o">==</span> <span class="s2">&quot;max_effect&quot;</span><span class="p">:</span>
        <span class="n">rescaled_scores</span> <span class="o">=</span> <span class="n">score_matrix</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Could not find base scaling </span><span class="se">\&quot;</span><span class="si">{0}</span><span class="se">\&quot;</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">base_scaling</span><span class="p">))</span>

    <span class="c1"># Scale each position</span>
    <span class="k">if</span> <span class="n">position_scaling</span> <span class="o">==</span> <span class="s2">&quot;max_effect&quot;</span><span class="p">:</span>
        <span class="n">max_effects</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span>
                                                            <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">rescaled_scores</span> <span class="o">/=</span> <span class="n">rescaled_scores</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)[</span><span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">,</span> <span class="p">:]</span>
        <span class="n">rescaled_scores</span> <span class="o">*=</span> <span class="n">max_effects</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">,</span> <span class="p">:]</span>
    <span class="k">elif</span> <span class="n">position_scaling</span> <span class="o">==</span> <span class="s2">&quot;probability&quot;</span><span class="p">:</span>
        <span class="n">rescaled_scores</span> <span class="o">/=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)[</span><span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">,</span> <span class="p">:]</span>
    <span class="k">elif</span> <span class="n">position_scaling</span> <span class="o">!=</span> <span class="s2">&quot;identity&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Could not find position scaling </span><span class="se">\&quot;</span><span class="si">{0}</span><span class="se">\&quot;</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">position_scaling</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">rescaled_scores</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span></div>


<div class="viewcode-block" id="heatmap"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.heatmap">[docs]</a><span class="k">def</span> <span class="nf">heatmap</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sequence_type</span><span class="o">=</span><span class="n">Genome</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Plots the input matrix of scores, generally those produced by an</span>
<span class="sd">    *in silico* mutagenesis experiment, on a heatmap.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    score_matrix : numpy.ndarray</span>
<span class="sd">        An :math:`L \\times N` array (where :math:`L` is the length of</span>
<span class="sd">        the sequence, and :math:`N` is the size of the alphabet)</span>
<span class="sd">        containing the scores for each base change at each position.</span>
<span class="sd">    mask : numpy.ndarray, dtype=bool or None, optional</span>
<span class="sd">        Default is `None`. An :math:`L \\times N` array (where :math:`L`</span>
<span class="sd">        is the length of the sequence, and :math:`N` is the size of the</span>
<span class="sd">        alphabet)  containing `True` at positions in the heatmap to</span>
<span class="sd">        mask. If `None`, no masking will occur.</span>
<span class="sd">    sequence_type : class, optional</span>
<span class="sd">        Default is `selene_sdk.sequences.Genome`. The class of sequence that</span>
<span class="sd">        the *in silico* mutagenesis results are associated with. This is</span>
<span class="sd">        generally a sub-class of `selene_sdk.sequences.Sequence`.</span>
<span class="sd">    **kwargs : dict</span>
<span class="sd">        Keyword arguments to pass to `seaborn.heatmap`. Some useful ones</span>
<span class="sd">        to remember are:</span>

<span class="sd">            * cbar_kws - Keyword arguments to forward to the colorbar.</span>
<span class="sd">            * yticklabels - Manipulate the tick labels on the y axis.</span>
<span class="sd">            * cbar - If `False`, hide the color bar. If `True`, show\</span>
<span class="sd">                     the colorbar.</span>
<span class="sd">            * cmap - The color map to use for the heatmap.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    matplotlib.pyplot.Axes</span>
<span class="sd">        The axes containing the heatmap plot.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Note that some things can break if we do not deepcopy.</span>
    <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">)</span>

    <span class="c1"># This flipping is so that ordering is consistent with ordering</span>
    <span class="c1"># in the sequence logo.</span>
    <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">score_matrix</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="n">score_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;yticklabels&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">yticklabels</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;yticklabels&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">yticklabels</span> <span class="o">=</span> <span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="s2">&quot;cbar_kws&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">cbar_kws</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;cbar_kws&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cbar_kws</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">use_gridspec</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="s2">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;cmap&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">cmap</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;cmap&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;Blues&quot;</span>
    <span class="n">ret</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="n">yticklabels</span><span class="p">,</span>
                      <span class="n">cbar_kws</span><span class="o">=</span><span class="n">cbar_kws</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">ret</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">ret</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ret</span></div>


<div class="viewcode-block" id="load_variant_abs_diff_scores"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.load_variant_abs_diff_scores">[docs]</a><span class="k">def</span> <span class="nf">load_variant_abs_diff_scores</span><span class="p">(</span><span class="n">input_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads the variant data, labels, and feature names from a diff scores</span>
<span class="sd">    file output from variant effect prediction.</span>

<span class="sd">    TODO: should we move this out of `vis.py`?</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input_path : str</span>
<span class="sd">        Path to the input file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple(np.ndarray, list(tuple(str)), list(str))</span>

<span class="sd">        * `tuple[0]` is the matrix of absolute difference scores. The rows\</span>
<span class="sd">          are the variants and the columns are the features for which the\</span>
<span class="sd">          model makes predictions.</span>
<span class="sd">        * `tuple[1]` is the list of variant labels. Each tuple contains\</span>
<span class="sd">          (chrom, pos, name, ref, alt).</span>
<span class="sd">        * `tuple[2]` is the list of features.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">diffs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
        <span class="n">colnames</span> <span class="o">=</span> <span class="n">file_handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
        <span class="n">features</span> <span class="o">=</span> <span class="n">colnames</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">5</span><span class="p">:]</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_handle</span><span class="p">:</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">scores</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">[</span><span class="mi">5</span><span class="p">:]]</span>
            <span class="n">label</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">cols</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
            <span class="n">diffs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
            <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
    <span class="n">diffs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">diffs</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">diffs</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">features</span><span class="p">)</span></div>


<div class="viewcode-block" id="sort_standard_chrs"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.sort_standard_chrs">[docs]</a><span class="k">def</span> <span class="nf">sort_standard_chrs</span><span class="p">(</span><span class="n">chrom</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the value on which the</span>
<span class="sd">    standard chromosomes can be sorted.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    chrom : str</span>
<span class="sd">        The chromosome</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        The value on which to sort</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
    <span class="k">if</span> <span class="n">chrom</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
        <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s1">&#39;X&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">23</span>
    <span class="k">elif</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">24</span>
    <span class="k">elif</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s1">&#39;M&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">25</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># unknown chr</span>
        <span class="k">return</span> <span class="mi">26</span></div>


<div class="viewcode-block" id="ordered_variants_and_indices"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.ordered_variants_and_indices">[docs]</a><span class="k">def</span> <span class="nf">ordered_variants_and_indices</span><span class="p">(</span><span class="n">labels</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the ordered variant labels, where the labels are sorted by chromosome</span>
<span class="sd">    and position, and the indices corresponding to the sort,</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    labels : list(tuple(str))</span>
<span class="sd">        The list of variant labels. Each label is a tuple of</span>
<span class="sd">        (chrom, pos, name, ref, alt).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple(list(tuple), list(int))</span>
<span class="sd">        The first value is the ordered list of labels. Each label</span>
<span class="sd">        is a tuple of (chrom, pos, ref, alt). The second value</span>
<span class="sd">        is the ordered list of label indices.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">labels_dict</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">labels</span><span class="p">):</span>
        <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">l</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
        <span class="n">info</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span>
        <span class="n">labels_dict</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">labels_list</span> <span class="ow">in</span> <span class="n">labels_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">labels_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
    <span class="n">ordered_keys</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">labels_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>
                          <span class="n">key</span><span class="o">=</span><span class="n">sort_standard_chrs</span><span class="p">)</span>

    <span class="n">ordered_labels</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">ordered_label_indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">ordered_keys</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">labels_dict</span><span class="p">[</span><span class="n">chrom</span><span class="p">]:</span>
            <span class="n">index</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">l</span>
            <span class="n">ordered_label_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
            <span class="n">ordered_labels</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">))</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">ordered_labels</span><span class="p">,</span> <span class="n">ordered_label_indices</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">_label_tuple_to_text</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">diff</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Converts the variant label tuple to a string.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    label : tuple(str)</span>
<span class="sd">        A tuple of (chrom, pos, ref, alt).</span>
<span class="sd">    diff : float</span>
<span class="sd">        The max difference score across some or all features in the model.</span>
<span class="sd">    genes : list(str) or None, optional</span>
<span class="sd">        Default is None. If the closest protein-coding `genes` are specified,</span>
<span class="sd">        will display them in the label text.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        The label text.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">label</span>
    <span class="k">if</span> <span class="n">genes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">genes_str</span> <span class="o">=</span> <span class="s2">&quot;none found&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">genes_str</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">text</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;max diff score: </span><span class="si">{0}</span><span class="s2">&lt;br /&gt;</span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2">, </span><span class="si">{3}</span><span class="s2">/</span><span class="si">{4}</span><span class="s2">&lt;br /&gt;&quot;</span>
                <span class="s2">&quot;closest protein-coding gene(s): </span><span class="si">{5}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">diff</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">genes_str</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;max diff score: </span><span class="si">{0}</span><span class="s2">&lt;br /&gt;</span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2">, </span><span class="si">{3}</span><span class="s2">/</span><span class="si">{4}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">diff</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">text</span>


<span class="k">def</span> <span class="nf">_variant_closest_genes</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">tabix_fh</span><span class="p">,</span> <span class="n">chrs_gene_intervals</span><span class="p">):</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="n">label</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">label</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">closest_genes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">overlaps</span> <span class="o">=</span> <span class="n">tabix_fh</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">overlaps</span><span class="p">:</span>
            <span class="n">closest_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">o</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">except</span> <span class="n">tabix</span><span class="o">.</span><span class="n">TabixError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">closest_genes</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">closest_genes</span>
    <span class="n">gene_intervals</span> <span class="o">=</span> <span class="n">chrs_gene_intervals</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
    <span class="n">closest</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span> <span class="ow">in</span> <span class="n">gene_intervals</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">pos</span> <span class="ow">and</span> <span class="n">closest</span> <span class="ow">and</span> <span class="n">closest</span> <span class="o">==</span> <span class="n">pos</span> <span class="o">-</span> <span class="n">end</span><span class="p">:</span>
            <span class="n">closest_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">pos</span> <span class="ow">and</span> <span class="p">(</span><span class="n">closest</span> <span class="ow">is</span> <span class="kc">None</span>
                <span class="ow">or</span> <span class="n">closest</span> <span class="o">&gt;</span> <span class="n">pos</span> <span class="o">-</span> <span class="n">end</span><span class="p">):</span>
            <span class="n">closest</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">-</span> <span class="n">end</span>
            <span class="n">closest_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">pos</span> <span class="ow">and</span> <span class="n">closest</span> <span class="ow">and</span> <span class="n">closest</span> <span class="o">==</span> <span class="n">start</span> <span class="o">-</span> <span class="n">pos</span><span class="p">:</span>
            <span class="n">closest_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">pos</span> <span class="ow">and</span> <span class="p">(</span><span class="n">closest</span> <span class="ow">is</span> <span class="kc">None</span>
                <span class="ow">or</span> <span class="n">closest</span> <span class="o">&gt;</span> <span class="n">start</span> <span class="o">-</span> <span class="n">pos</span><span class="p">):</span>
            <span class="n">closest</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="n">pos</span>
            <span class="n">closest_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">closest_genes</span>


<span class="k">def</span> <span class="nf">_load_chrs_gene_intervals</span><span class="p">(</span><span class="n">gene_intervals_bed</span><span class="p">):</span>
    <span class="n">chrs_gene_intervals</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gene_intervals_bed</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_handle</span><span class="p">:</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
            <span class="n">chrs_gene_intervals</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">gene</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">chrs_gene_intervals</span>


<span class="k">def</span> <span class="nf">_variants_closest_protein_coding_gene</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">):</span>
    <span class="n">gene_intervals_tabix</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span>
        <span class="s2">&quot;selene_sdk&quot;</span><span class="p">,</span>
        <span class="p">(</span><span class="s2">&quot;interpret/data/gencode_v28_</span><span class="si">{0}</span><span class="s2">/&quot;</span>
         <span class="s2">&quot;protein_coding_l12_genes.bed.gz&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">))</span>
    <span class="n">gene_intervals_bed</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span>
        <span class="s2">&quot;selene_sdk&quot;</span><span class="p">,</span>
        <span class="p">(</span><span class="s2">&quot;interpret/data/gencode_v28_</span><span class="si">{0}</span><span class="s2">/&quot;</span>
         <span class="s2">&quot;protein_coding_l12_genes.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">))</span>

    <span class="n">tabix_fh</span> <span class="o">=</span> <span class="n">tabix</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">gene_intervals_tabix</span><span class="p">)</span>
    <span class="n">chrs_gene_intervals</span> <span class="o">=</span> <span class="n">_load_chrs_gene_intervals</span><span class="p">(</span><span class="n">gene_intervals_bed</span><span class="p">)</span>

    <span class="n">labels_gene_information</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
        <span class="n">labels_gene_information</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_variant_closest_genes</span><span class="p">(</span>
            <span class="n">l</span><span class="p">,</span> <span class="n">tabix_fh</span><span class="p">,</span> <span class="n">chrs_gene_intervals</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">labels_gene_information</span>


<div class="viewcode-block" id="variant_diffs_scatter_plot"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.variant_diffs_scatter_plot">[docs]</a><span class="k">def</span> <span class="nf">variant_diffs_scatter_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span>
                               <span class="n">labels</span><span class="p">,</span>
                               <span class="n">features</span><span class="p">,</span>
                               <span class="n">output_path</span><span class="p">,</span>
                               <span class="n">filter_features</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">labels_sort_fn</span><span class="o">=</span><span class="n">ordered_variants_and_indices</span><span class="p">,</span>
                               <span class="n">nth_percentile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">hg_reference_version</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">threshold_line</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">auto_open</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Displays each variant&#39;s max probability difference across features</span>
<span class="sd">    as a point in a scatter plot. The points in the scatter plot are</span>
<span class="sd">    ordered by the variant chromosome and position by default. Variants can</span>
<span class="sd">    be sorted differently by passing in a new `labels_sort_fn`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data : np.ndarray</span>
<span class="sd">        Absolute difference scores for variants across all features that a</span>
<span class="sd">        model predicts. This is the first value in the tuple returned by</span>
<span class="sd">        `load_variant_abs_diff_scores`.</span>
<span class="sd">    labels : list(tuple(str))</span>
<span class="sd">        A list of variant labels. This is the second value in the tuple</span>
<span class="sd">        returned by `load_variant_abs_diff_scores`.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        A list of the features the model predicts. This is the third value</span>
<span class="sd">        in the tuple returned by `load_variant_abs_diff_scores`.</span>
<span class="sd">    output_path : str</span>
<span class="sd">        Path to output file. Must have &#39;.html&#39; extension.</span>
<span class="sd">    filter_features : types.FunctionType or None, optional</span>
<span class="sd">        Default is None. A function that takes in a `list(str)` of features</span>
<span class="sd">        and returns the `list(int)` of feature indices over which we would</span>
<span class="sd">        compute the max(probability difference) for each variant. For example,</span>
<span class="sd">        a user may only want to visualize the max probability difference</span>
<span class="sd">        for TF binding features. If `None`, uses all the features.</span>
<span class="sd">    labels_sort_fn : types.FunctionType, optional</span>
<span class="sd">        Default is `ordered_variants_and_indices`. A function that takes</span>
<span class="sd">        in a `list(tuple(str))` of labels corresponding to the rows in `data`</span>
<span class="sd">        and returns a `tuple(list(tuple), list(int))`, where the first value</span>
<span class="sd">        is the ordered list of variant labels and the second value is the</span>
<span class="sd">        ordered list of indices for those variant labels. By default,</span>
<span class="sd">        variants are sorted by chromosome and position.</span>
<span class="sd">    nth_percentile : int [0, 100] or None, optional</span>
<span class="sd">        Default is None. If `nth_percentile` is not None, only displays the</span>
<span class="sd">        variants with a max absolute difference score within the</span>
<span class="sd">        `nth_percentile` of scores.</span>
<span class="sd">    hg_reference_version : str {&quot;hg19&quot;, &quot;hg38&quot;} or None, optional</span>
<span class="sd">        Default is None. On hover, we can display the gene(s) closest to</span>
<span class="sd">        each variant if `hg_reference_version` is not None, where closest</span>
<span class="sd">        can be a variant within a gene interval (where genes and</span>
<span class="sd">        their coordinates are taken from level 1 &amp; 2 protein-coding genes</span>
<span class="sd">        in gencode v28) or near a gene. In the future, we will allow users</span>
<span class="sd">        to specify their own genome file so that this information can</span>
<span class="sd">        be annotated to variants from other organisms, other genome versions,</span>
<span class="sd">        etc.</span>
<span class="sd">    threshold_line : float or None, optional</span>
<span class="sd">        Default is None. If `threshold_line` is not None, draws a horizontal</span>
<span class="sd">        line at the specified threshold. Helps focus the visual on variants</span>
<span class="sd">        above a certain threshold.</span>
<span class="sd">    auto_open : bool, optional</span>
<span class="sd">        Default is False. If `auto_open`, will automatically open a web</span>
<span class="sd">        browser that displays the plotted HTML file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    plotly.graph_objs.graph_objs.Figure</span>
<span class="sd">        The generated Plotly figure.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">labels_ordered</span><span class="p">,</span> <span class="n">label_indices</span> <span class="o">=</span> <span class="n">labels_sort_fn</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
    <span class="n">variant_closest_genes</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">hg_reference_version</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">variant_closest_genes</span> <span class="o">=</span> <span class="n">_variants_closest_protein_coding_gene</span><span class="p">(</span>
            <span class="n">labels_ordered</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">hg_reference_version</span><span class="p">)</span>
    <span class="n">ordered_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">label_indices</span><span class="p">,</span> <span class="p">:]</span>

    <span class="n">feature_indices</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">filter_features</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">feature_indices</span> <span class="o">=</span> <span class="n">filter_features</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
        <span class="n">ordered_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">feature_indices</span><span class="p">]</span>
    <span class="n">variants_max_diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">amax</span><span class="p">(</span><span class="n">ordered_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">display_labels</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">nth_percentile</span><span class="p">:</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">variants_max_diff</span><span class="p">,</span> <span class="n">nth_percentile</span><span class="p">)</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">variants_max_diff</span> <span class="o">&gt;=</span> <span class="n">p</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> variants with max abs diff score above </span><span class="si">{1}</span><span class="s2"> are in the &quot;</span>
              <span class="s2">&quot;</span><span class="si">{2}</span><span class="s2">th percentile.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                  <span class="nb">len</span><span class="p">(</span><span class="n">keep</span><span class="p">),</span> <span class="n">p</span><span class="p">,</span> <span class="n">nth_percentile</span><span class="p">))</span>
        <span class="n">variants_max_diff</span> <span class="o">=</span> <span class="n">variants_max_diff</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>
        <span class="n">display_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">labels_ordered</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">keep</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">display_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">l</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">display_labels</span> <span class="o">=</span> <span class="n">labels_ordered</span>

    <span class="k">if</span> <span class="n">variant_closest_genes</span><span class="p">:</span>
        <span class="n">text_labels</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">_label_tuple_to_text</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">d</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span> <span class="k">for</span> <span class="n">l</span><span class="p">,</span> <span class="n">d</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span>
            <span class="nb">zip</span><span class="p">(</span><span class="n">display_labels</span><span class="p">,</span> <span class="n">variants_max_diff</span><span class="p">,</span> <span class="n">variant_closest_genes</span><span class="p">)]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">text_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">_label_tuple_to_text</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">d</span><span class="p">)</span> <span class="k">for</span> <span class="n">l</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span>
                       <span class="nb">zip</span><span class="p">(</span><span class="n">display_labels</span><span class="p">,</span> <span class="n">variants_max_diff</span><span class="p">)]</span>

    <span class="n">label_x</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">str</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">display_labels</span><span class="p">]</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">label_x</span><span class="p">,</span>
                       <span class="n">y</span><span class="o">=</span><span class="n">variants_max_diff</span><span class="p">,</span>
                       <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers&#39;</span><span class="p">,</span>
                       <span class="n">marker</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
                           <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;#39CCCC&quot;</span><span class="p">,</span>
                           <span class="n">line</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)),</span>
                       <span class="n">text</span><span class="o">=</span><span class="n">text_labels</span><span class="p">,</span>
                       <span class="n">hoverinfo</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)]</span>

    <span class="n">go_layout</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Max probability difference scores&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hovermode&quot;</span><span class="p">:</span> <span class="s2">&quot;closest&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hoverlabel&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;font&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;size&quot;</span><span class="p">:</span> <span class="mi">16</span><span class="p">}</span>
            <span class="p">},</span>
        <span class="s2">&quot;xaxis&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Genome coordinates&quot;</span><span class="p">,</span>
            <span class="s2">&quot;showticklabels&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;tickangle&quot;</span><span class="p">:</span> <span class="mi">35</span><span class="p">,</span>
            <span class="s2">&quot;titlefont&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;family&quot;</span><span class="p">:</span> <span class="s2">&quot;Arial, sans-serif&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;size&quot;</span><span class="p">:</span> <span class="mi">16</span><span class="p">},</span>
            <span class="s2">&quot;nticks&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">,</span>
            <span class="s2">&quot;tickmode&quot;</span><span class="p">:</span> <span class="s2">&quot;auto&quot;</span><span class="p">,</span>
            <span class="s2">&quot;automargin&quot;</span><span class="p">:</span> <span class="kc">True</span>
            <span class="p">},</span>
        <span class="s2">&quot;yaxis&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Absolute difference&quot;</span><span class="p">,</span>
            <span class="s2">&quot;titlefont&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;family&quot;</span><span class="p">:</span> <span class="s2">&quot;Arial, sans-serif&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;size&quot;</span><span class="p">:</span> <span class="mi">16</span><span class="p">}</span>
            <span class="p">}</span>
        <span class="p">}</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">threshold_line</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
        <span class="n">layout</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Layout</span><span class="p">(</span>
            <span class="o">**</span><span class="n">go_layout</span><span class="p">,</span>
            <span class="n">shapes</span><span class="o">=</span><span class="p">[</span>
                <span class="p">{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;line&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;x0&quot;</span><span class="p">:</span> <span class="n">label_x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                 <span class="s2">&quot;y0&quot;</span><span class="p">:</span> <span class="n">threshold_line</span><span class="p">,</span>
                 <span class="s2">&quot;x1&quot;</span><span class="p">:</span> <span class="n">label_x</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
                 <span class="s2">&quot;y1&quot;</span><span class="p">:</span> <span class="n">threshold_line</span><span class="p">,</span>
                 <span class="s2">&quot;line&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;rgba(255, 99, 71, 1)&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">2</span>
                    <span class="p">}</span>
                <span class="p">}</span>
            <span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">layout</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Layout</span><span class="p">(</span><span class="o">**</span><span class="n">go_layout</span><span class="p">)</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="n">layout</span><span class="p">)</span>
    <span class="n">path</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plot</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">output_path</span><span class="p">,</span> <span class="n">auto_open</span><span class="o">=</span><span class="n">auto_open</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">fig</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2018, Selene Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>