

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>selene_sdk.interpret &mdash; Selene 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="_static/css/selene_docs_theme.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="selene_sdk.predict" href="predict.html" />
    <link rel="prev" title="selene_sdk" href="selene.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Selene
          

          
            
            <img src="_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-selene-with-pip">Installing selene with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>selene_sdk.interpret</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/interpret.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene_sdk.interpret">
<span id="selene-sdk-interpret"></span><h1>selene_sdk.interpret<a class="headerlink" href="#module-selene_sdk.interpret" title="Permalink to this headline">¶</a></h1>
<p>This module provides functions and classes for interpreting modules
trained with Selene.</p>
<div class="section" id="sequence-logo">
<h2>sequence_logo<a class="headerlink" href="#sequence-logo" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.interpret.sequence_logo">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">sequence_logo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">score_matrix</span></em>, <em class="sig-param"><span class="pre">order='value'</span></em>, <em class="sig-param"><span class="pre">width=1.0</span></em>, <em class="sig-param"><span class="pre">ax=None</span></em>, <em class="sig-param"><span class="pre">sequence_type=&lt;class</span> <span class="pre">'selene_sdk.sequences.genome.Genome'&gt;</span></em>, <em class="sig-param"><span class="pre">font_properties=None</span></em>, <em class="sig-param"><span class="pre">color_scheme=None</span></em>, <em class="sig-param"><span class="pre">**kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/vis.html#sequence_logo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.sequence_logo" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots a sequence logo for visualizing motifs.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>score_matrix</strong> (<em>np.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span> is the length of
the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the alphabet)
containing the scores for each base occuring at each position.</p></li>
<li><p><strong>order</strong> (<em>{‘alpha’, ‘value’}</em>) – The manner by which to sort the bases stacked at each position
in the sequence logo plot.</p>
<blockquote>
<div><ul class="simple">
<li><p>‘alpha’ - Bases go in the order they are found in the                      sequence alphabet.</p></li>
<li><p>‘value’ - Bases go in the order of their value, with the                        largest at the bottom.</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>width</strong> (<em>float, optional</em>) – Default is 1. The width of each character in the plotted logo.
A value of 1 will mean that there is no gap between each the
characters at each position. A value of 0 will not draw any
characters.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.pyplot.Axes or None, optional</em>) – Default is <cite>None</cite>. The axes to plot on. If left as <cite>None</cite>, a new
axis will be created.</p></li>
<li><p><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene_sdk.sequences.Genome</cite>. The type of sequence that
the <em>in silico</em> mutagenesis results are associated with. This
should generally be a subclass of <cite>selene_sdk.sequences.Sequence</cite>.</p></li>
<li><p><strong>font_properties</strong> (<em>matplotlib.font_manager.FontProperties or None, optional</em>) – Default is <cite>None</cite>. A <cite>matplotlib.font_manager.FontProperties</cite>
object that specifies the properties of the font to use for
plotting the motif. If <cite>None</cite>, no font will be used, and the
text will be rendered by a path. This method of rendering paths
is  preferred, as it ensures all character heights correspond to
the actual values, and that there are no extra gaps between
the tops and bottoms of characters at each position in the
sequence logo. If the user opts to use a value other
than <cite>None</cite>, then no such guarantee can be made.</p></li>
<li><p><strong>color_scheme</strong> (<em>list(str) or None, optional</em>) – Default is <cite>None</cite>. A list containing the hex codes or names of
colors to use, appearing in the order of the bases of the
sequence type. If left as <cite>None</cite>, a default palette will be made
with <cite>seaborn.color_palette</cite>, and will have as many
colors as there are characters in the input sequence alphabet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The axes containing the sequence logo plot.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>matplotlib.pyplot.Axes</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.10)"><strong>ValueError</strong></a> – If the number of columns in <cite>score_matrix</cite> does not match the
    number of characters in the alphabet of <cite>sequence_type</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.10)"><strong>ValueError</strong></a> – If the number of colors in <cite>color_palette</cite> does not match the
    number of characters in the alphabet of <cite>sequence_type</cite>.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>We have included an example of the output from a <cite>sequence_logo</cite>
plot below:</p>
<img alt="_images/sequence_logo_example.png" src="_images/sequence_logo_example.png" />
</dd></dl>

</div>
<div class="section" id="heatmap">
<h2>heatmap<a class="headerlink" href="#heatmap" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.interpret.heatmap">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">score_matrix</span></em>, <em class="sig-param"><span class="pre">mask=None</span></em>, <em class="sig-param"><span class="pre">sequence_type=&lt;class</span> <span class="pre">'selene_sdk.sequences.genome.Genome'&gt;</span></em>, <em class="sig-param"><span class="pre">**kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/vis.html#heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the input matrix of scores, generally those produced by an
<em>in silico</em> mutagenesis experiment, on a heatmap.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>score_matrix</strong> (<em>numpy.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span> is the length of
the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the alphabet)
containing the scores for each base change at each position.</p></li>
<li><p><strong>mask</strong> (<em>numpy.ndarray, dtype=bool or None, optional</em>) – Default is <cite>None</cite>. An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span>
is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the
alphabet)  containing <cite>True</cite> at positions in the heatmap to
mask. If <cite>None</cite>, no masking will occur.</p></li>
<li><p><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene_sdk.sequences.Genome</cite>. The class of sequence that
the <em>in silico</em> mutagenesis results are associated with. This is
generally a sub-class of <cite>selene_sdk.sequences.Sequence</cite>.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Keyword arguments to pass to <cite>seaborn.heatmap</cite>. Some useful ones
to remember are:</p>
<blockquote>
<div><ul class="simple">
<li><p>cbar_kws - Keyword arguments to forward to the colorbar.</p></li>
<li><p>yticklabels - Manipulate the tick labels on the y axis.</p></li>
<li><p>cbar - If <cite>False</cite>, hide the color bar. If <cite>True</cite>, show                     the colorbar.</p></li>
<li><p>cmap - The color map to use for the heatmap.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The axes containing the heatmap plot.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>matplotlib.pyplot.Axes</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="ismresult">
<h2>ISMResult<a class="headerlink" href="#ismresult" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="selene_sdk.interpret.ISMResult">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">ISMResult</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">data_frame</span></em>, <em class="sig-param"><span class="pre">sequence_type=&lt;class</span> <span class="pre">'selene_sdk.sequences.genome.Genome'&gt;</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/ism_result.html#ISMResult"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.ISMResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>An object storing the results of an <em>in silico</em> mutagenesis
experiment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_frame</strong> (<em>pandas.DataFrame</em>) – The data frame with the results from the <em>in silico</em>
mutagenesis experiments.</p></li>
<li><p><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene_sdk.sequences.Genome</cite>. The type of sequence
that the <em>in silico</em> mutagenesis results are associated
with. This should generally be a subclass of
<cite>selene_sdk.sequences.Sequence</cite></p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.10)"><strong>ValueError</strong></a> – If the input data frame contains a base not included in the
    alphabet of <cite>sequence_type</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.10)"><strong>Exception</strong></a> – If multiple reference positions are specified in the input
    data frame.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.10)"><strong>Exception</strong></a> – If the input data does not contain scores for every mutation
    at every position.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.interpret.ISMResult.from_file">
<em class="property"><span class="pre">static</span> </em><span class="sig-name descname"><span class="pre">from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">input_path</span></em>, <em class="sig-param"><span class="pre">sequence_type=&lt;class</span> <span class="pre">'selene_sdk.sequences.genome.Genome'&gt;</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/ism_result.html#ISMResult.from_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.ISMResult.from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads a <cite>selene_sdk.interpret.ISMResult</cite> from a <cite>pandas.DataFrame</cite>
stored in a file of comma separated values (CSV).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_path</strong> (<em>str</em>) – A path to the file of comma separated input values.</p></li>
<li><p><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene_sdk.sequences.Genome</cite>. The type of sequence
that the <em>in silico</em> mutagenesis results are associated
with. This should generally be a subclass of
<cite>selene_sdk.sequences.Sequence</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <em>in silico</em> mutagenesis results that were stored in the
specified input file.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#selene_sdk.interpret.ISMResult" title="selene_sdk.interpret.ISMResult">selene_sdk.interpret.ISMResult</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.interpret.ISMResult.get_score_matrix_for">
<span class="sig-name descname"><span class="pre">get_score_matrix_for</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">feature</span></em>, <em class="sig-param"><span class="pre">reference_mask=None</span></em>, <em class="sig-param"><span class="pre">dtype=&lt;class</span> <span class="pre">'numpy.float64'&gt;</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/ism_result.html#ISMResult.get_score_matrix_for"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.ISMResult.get_score_matrix_for" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a feature from the <em>in silico</em> mutagenesis results
as a matrix, where the reference base positions hold the value
for the reference prediction, and alternative positions hold the
results for making a one-base change from the reference base to
the specified alternative base.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>feature</strong> (<em>str</em>) – The name of the feature to extract as a matrix.</p></li>
<li><p><strong>reference_mask</strong> (<em>float or None, optional</em>) – Default is <cite>None</cite>. A value to mask the reference entries
with. If left as <cite>None</cite>, then no masking will be performed
on the reference positions.</p></li>
<li><p><strong>dtype</strong> (<em>numpy.dtype, optional</em>) – Default is <cite>numpy.float64</cite>. The data type to use for the
returned matrix.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <span class="math notranslate nohighlight">\(L \times N\)</span> shaped array (where <span class="math notranslate nohighlight">\(L\)</span> is the
sequence length, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the alphabet
of <cite>sequence_type</cite>) that holds the results from the
<em>in silico</em> mutagenesis experiment for the specified
feature. The elements will be of type <cite>dtype</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.10)"><strong>ValueError</strong></a> – If the input data frame contains a base not included in the
    alphabet of <cite>sequence_type</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="selene_sdk.interpret.ISMResult.reference_sequence">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">reference_sequence</span></span><a class="headerlink" href="#selene_sdk.interpret.ISMResult.reference_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>The reference sequence that the <em>in silico</em> mutagenesis
experiment was performed on.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The reference sequence (i.e. non-mutated input) as a
string of characters.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="selene_sdk.interpret.ISMResult.sequence_type">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">sequence_type</span></span><a class="headerlink" href="#selene_sdk.interpret.ISMResult.sequence_type" title="Permalink to this definition">¶</a></dt>
<dd><p>The type of underlying sequence. This should generally be a
subclass of <cite>selene_sdk.sequences.Sequence</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The type of sequence that the <em>in silico</em> mutagenesis was
performed on.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>class</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="rescale-score-matrix">
<h2>rescale_score_matrix<a class="headerlink" href="#rescale-score-matrix" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.interpret.rescale_score_matrix">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">rescale_score_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">score_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_scaling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'identity'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_scaling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'identity'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/vis.html#rescale_score_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.rescale_score_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs base-wise and position-wise scaling of a score matrix for a
feature, usually produced from an <em>in silico</em> mutagenesis experiment.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>score_matrix</strong> (<em>numpy.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> matrix containing the scores for each
position, where <span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and
<span class="math notranslate nohighlight">\(N\)</span> is the number of characters in the alphabet.</p></li>
<li><p><strong>base_scaling</strong> (<em>{‘identity’, ‘probability’, ‘max_effect’}</em>) –</p>
<p>The type of scaling performed on each base at a given position.</p>
<blockquote>
<div><ul class="simple">
<li><p>‘identity’ - No transformation will be applied to the                           data.</p></li>
<li><p>‘probability’ - The relative sizes of the bases will be                              the original input probabilities.</p></li>
<li><p>‘max_effect’ - The relative sizes of the bases will be                             the max effect of the original input                             values.</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>position_scaling</strong> (<em>{‘identity’, ‘probability’, ‘max_effect’}</em>) –</p>
<p>The type of scaling performed on each position.</p>
<blockquote>
<div><ul class="simple">
<li><p>‘identity’    - No transformation will be applied to the data.</p></li>
<li><p>‘probability’ - The sum of values at a position will be                              equal to the sum of the original input                              values at that position.</p></li>
<li><p>‘max_effect’  - The sum of values at a position will be                              equal to the sum of the max effect values                              of the original input values at that                              position.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The transformed score matrix.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.10)"><strong>ValueError</strong></a> – If an unsupported <cite>base_scaling</cite> or <cite>position_scaling</cite> is
    entered.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="load-variant-abs-diff-scores">
<h2>load_variant_abs_diff_scores<a class="headerlink" href="#load-variant-abs-diff-scores" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.interpret.load_variant_abs_diff_scores">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">load_variant_abs_diff_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/vis.html#load_variant_abs_diff_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.load_variant_abs_diff_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the variant data, labels, and feature names from a diff scores
file output from variant effect prediction.</p>
<p>TODO: should we move this out of <cite>vis.py</cite>?</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>input_path</strong> (<em>str</em>) – Path to the input file.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><cite>tuple[0]</cite> is the matrix of absolute difference scores. The rows          are the variants and the columns are the features for which the          model makes predictions.</p></li>
<li><p><cite>tuple[1]</cite> is the list of variant labels. Each tuple contains          (chrom, pos, name, ref, alt).</p></li>
<li><p><cite>tuple[2]</cite> is the list of features.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>(np.ndarray, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.10)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>)), <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.10)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>))</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="variant-diffs-scatter-plot">
<h2>variant_diffs_scatter_plot<a class="headerlink" href="#variant-diffs-scatter-plot" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.interpret.variant_diffs_scatter_plot">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">variant_diffs_scatter_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">data</span></em>, <em class="sig-param"><span class="pre">labels</span></em>, <em class="sig-param"><span class="pre">features</span></em>, <em class="sig-param"><span class="pre">output_path</span></em>, <em class="sig-param"><span class="pre">filter_features=None</span></em>, <em class="sig-param"><span class="pre">labels_sort_fn=&lt;function</span> <span class="pre">ordered_variants_and_indices&gt;</span></em>, <em class="sig-param"><span class="pre">nth_percentile=None</span></em>, <em class="sig-param"><span class="pre">hg_reference_version=None</span></em>, <em class="sig-param"><span class="pre">threshold_line=None</span></em>, <em class="sig-param"><span class="pre">auto_open=False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/vis.html#variant_diffs_scatter_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.variant_diffs_scatter_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Displays each variant’s max probability difference across features
as a point in a scatter plot. The points in the scatter plot are
ordered by the variant chromosome and position by default. Variants can
be sorted differently by passing in a new <cite>labels_sort_fn</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>np.ndarray</em>) – Absolute difference scores for variants across all features that a
model predicts. This is the first value in the tuple returned by
<cite>load_variant_abs_diff_scores</cite>.</p></li>
<li><p><strong>labels</strong> (<em>list(tuple(str))</em>) – A list of variant labels. This is the second value in the tuple
returned by <cite>load_variant_abs_diff_scores</cite>.</p></li>
<li><p><strong>features</strong> (<em>list(str)</em>) – A list of the features the model predicts. This is the third value
in the tuple returned by <cite>load_variant_abs_diff_scores</cite>.</p></li>
<li><p><strong>output_path</strong> (<em>str</em>) – Path to output file. Must have ‘.html’ extension.</p></li>
<li><p><strong>filter_features</strong> (<em>types.FunctionType or None, optional</em>) – Default is None. A function that takes in a <cite>list(str)</cite> of features
and returns the <cite>list(int)</cite> of feature indices over which we would
compute the max(probability difference) for each variant. For example,
a user may only want to visualize the max probability difference
for TF binding features. If <cite>None</cite>, uses all the features.</p></li>
<li><p><strong>labels_sort_fn</strong> (<em>types.FunctionType, optional</em>) – Default is <cite>ordered_variants_and_indices</cite>. A function that takes
in a <cite>list(tuple(str))</cite> of labels corresponding to the rows in <cite>data</cite>
and returns a <cite>tuple(list(tuple), list(int))</cite>, where the first value
is the ordered list of variant labels and the second value is the
ordered list of indices for those variant labels. By default,
variants are sorted by chromosome and position.</p></li>
<li><p><strong>nth_percentile</strong> (<em>int [0, 100] or None, optional</em>) – Default is None. If <cite>nth_percentile</cite> is not None, only displays the
variants with a max absolute difference score within the
<cite>nth_percentile</cite> of scores.</p></li>
<li><p><strong>hg_reference_version</strong> (<em>str {“hg19”, “hg38”} or None, optional</em>) – Default is None. On hover, we can display the gene(s) closest to
each variant if <cite>hg_reference_version</cite> is not None, where closest
can be a variant within a gene interval (where genes and
their coordinates are taken from level 1 &amp; 2 protein-coding genes
in gencode v28) or near a gene. In the future, we will allow users
to specify their own genome file so that this information can
be annotated to variants from other organisms, other genome versions,
etc.</p></li>
<li><p><strong>threshold_line</strong> (<em>float or None, optional</em>) – Default is None. If <cite>threshold_line</cite> is not None, draws a horizontal
line at the specified threshold. Helps focus the visual on variants
above a certain threshold.</p></li>
<li><p><strong>auto_open</strong> (<em>bool, optional</em>) – Default is False. If <cite>auto_open</cite>, will automatically open a web
browser that displays the plotted HTML file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The generated Plotly figure.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotly.graph_objs.graph_objs.Figure</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="ordered-variants-and-indices">
<h2>ordered_variants_and_indices<a class="headerlink" href="#ordered-variants-and-indices" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.interpret.ordered_variants_and_indices">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">ordered_variants_and_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/vis.html#ordered_variants_and_indices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.ordered_variants_and_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the ordered variant labels, where the labels are sorted by chromosome
and position, and the indices corresponding to the sort,</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>labels</strong> (<em>list(tuple(str))</em>) – The list of variant labels. Each label is a tuple of
(chrom, pos, name, ref, alt).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The first value is the ordered list of labels. Each label
is a tuple of (chrom, pos, ref, alt). The second value
is the ordered list of label indices.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.10)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>), <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.10)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)">int</a>))</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="sort-standard-chrs">
<h2>sort_standard_chrs<a class="headerlink" href="#sort-standard-chrs" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.interpret.sort_standard_chrs">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.interpret.</span></span><span class="sig-name descname"><span class="pre">sort_standard_chrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/interpret/vis.html#sort_standard_chrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.interpret.sort_standard_chrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the value on which the
standard chromosomes can be sorted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>chrom</strong> (<em>str</em>) – The chromosome</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The value on which to sort</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)">int</a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="predict.html" class="btn btn-neutral float-right" title="selene_sdk.predict" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="selene.html" class="btn btn-neutral float-left" title="selene_sdk" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2018, Selene Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>