

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene_sdk.predict &mdash; Selene 0.5.dev0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="selene_sdk.samplers" href="samplers.html" />
    <link rel="prev" title="selene_sdk.interpret" href="interpret.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Selene
          

          
            
            <img src="_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-selene-with-pip">Installing selene with pip:</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>selene_sdk.predict</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/predict.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene_sdk.predict">
<span id="selene-sdk-predict"></span><h1>selene_sdk.predict<a class="headerlink" href="#module-selene_sdk.predict" title="Permalink to this headline">¶</a></h1>
<p>This module contains classes and methods for making and analyzing
predictions with models that have already been trained.</p>
<div class="section" id="analyzesequences">
<h2>AnalyzeSequences<a class="headerlink" href="#analyzesequences" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="selene_sdk.predict.AnalyzeSequences">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">selene_sdk.predict.</span></code><code class="sig-name descname"><span class="pre">AnalyzeSequences</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="pre">model</span></em>, <em class="sig-param"><span class="pre">trained_model_path</span></em>, <em class="sig-param"><span class="pre">sequence_length</span></em>, <em class="sig-param"><span class="pre">features</span></em>, <em class="sig-param"><span class="pre">batch_size=64</span></em>, <em class="sig-param"><span class="pre">use_cuda=False</span></em>, <em class="sig-param"><span class="pre">data_parallel=False</span></em>, <em class="sig-param"><span class="pre">reference_sequence=&lt;class</span> <span class="pre">'selene_sdk.sequences.genome.Genome'&gt;</span></em>, <em class="sig-param"><span class="pre">write_mem_limit=1500</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Score sequences and their variants using the predictions made
by a trained model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>torch.nn.Module</em>) – A sequence-based model architecture.</p></li>
<li><p><strong>trained_model_path</strong> (<em>str or list(str)</em>) – The path(s) to the weights file for a trained sequence-based model.
For a single path, the model architecture must match <cite>model</cite>. For
a list of paths, assumes that the <cite>model</cite> passed in is of type
<cite>selene_sdk.utils.MultiModelWrapper</cite>, which takes in a list of
models. The paths must be ordered the same way the models
are ordered in that list. <cite>list(str)</cite> input is an API-only function–
Selene’s config file CLI does not support the <cite>MultiModelWrapper</cite>
functionality at this time.</p></li>
<li><p><strong>sequence_length</strong> (<em>int</em>) – The length of sequences that the model is expecting.</p></li>
<li><p><strong>features</strong> (<em>list(str)</em>) – The names of the features that the model is predicting.</p></li>
<li><p><strong>batch_size</strong> (<em>int, optional</em>) – Default is 64. The size of the mini-batches to use.</p></li>
<li><p><strong>use_cuda</strong> (<em>bool, optional</em>) – Default is <cite>False</cite>. Specifies whether CUDA-enabled GPUs are available
for torch to use.</p></li>
<li><p><strong>data_parallel</strong> (<em>bool, optional</em>) – Default is <cite>False</cite>. Specify whether multiple GPUs are available for
torch to use during training.</p></li>
<li><p><strong>reference_sequence</strong> (<em>class, optional</em>) – Default is <cite>selene_sdk.sequences.Genome</cite>. The type of sequence on
which this analysis will be performed. Please note that if you need
to use variant effect prediction, you cannot only pass in the
class–you must pass in the constructed <cite>selene_sdk.sequences.Sequence</cite>
object with a particular sequence version (e.g. <cite>Genome(“hg19.fa”)</cite>).
This version does NOT have to be the same sequence version that the
model was trained on. That is, if the sequences in your variants file
are hg19 but your model was trained on hg38 sequences, you should pass
in hg19.</p></li>
<li><p><strong>write_mem_limit</strong> (<em>int, optional</em>) – Default is 5000. Specify, in MB, the amount of memory you want to
allocate to storing model predictions/scores. When running one of
_in <a href="#id3"><span class="problematic" id="id4">silico_</span></a> mutagenesis, variant effect prediction, or prediction,
prediction/score handlers will accumulate data in memory and only
write this data to files periodically. By default, Selene will write
to files when the total amount of data (across all handlers) takes up
5000MB of space. Please keep in mind that Selene will not monitor the
memory needed to actually carry out the operations (e.g. variant effect
prediction) or load the model, so <cite>write_mem_limit</cite> should always be
less than the total amount of CPU memory you have available on your
machine. For example, for variant effect prediction, we load all
the variants in 1 file into memory before getting the predictions, so
your machine must have enough memory to accommodate that. Another
possible consideration is your model size and whether you are
using it on the CPU or a CUDA-enabled GPU (i.e. setting
<cite>use_cuda</cite> to True).</p></li>
</ul>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>~AnalyzeSequences.model</strong> (<a class="reference external" href="https://pytorch.org/docs/master/generated/torch.nn.Module.html#torch.nn.Module" title="(in PyTorch vmaster (1.9.0a0+git26116b3 ))"><em>torch.nn.Module</em></a>) – A sequence-based model that has already been trained.</p></li>
<li><p><strong>~AnalyzeSequences.sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The length of sequences that the model is expecting.</p></li>
<li><p><strong>~AnalyzeSequences.batch_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The size of the mini-batches to use.</p></li>
<li><p><strong>~AnalyzeSequences.features</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – The names of the features that the model is predicting.</p></li>
<li><p><strong>~AnalyzeSequences.use_cuda</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Specifies whether to use a CUDA-enabled GPU or not.</p></li>
<li><p><strong>~AnalyzeSequences.data_parallel</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to use multiple GPUs or not.</p></li>
<li><p><strong>~AnalyzeSequences.reference_sequence</strong> (<em>class</em>) – The type of sequence on which this analysis will be performed.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="selene_sdk.predict.AnalyzeSequences.get_predictions">
<code class="sig-name descname"><span class="pre">get_predictions</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tsv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences.get_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences.get_predictions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get model predictions for sequences specified as a raw sequence,
FASTA, or BED file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>input_path</strong> (<em>str</em>) – A single sequence, or a path to the FASTA or BED file input.</p></li>
<li><p><strong>output_dir</strong> (<em>str, optional</em>) – Default is None. Output directory to write the model predictions.
If this is left blank a raw sequence input will be assumed, though
an output directory is required for FASTA and BED inputs.</p></li>
<li><p><strong>output_format</strong> (<em>{‘tsv’, ‘hdf5’}, optional</em>) – Default is ‘tsv’. Choose whether to save TSV or HDF5 output files.
TSV is easier to access (i.e. open with text editor/Excel) and
quickly peruse, whereas HDF5 files must be accessed through
specific packages/viewers that support this format (e.g. h5py
Python package). Choose</p>
<blockquote>
<div><ul class="simple">
<li><p>‘tsv’ if your list of sequences is relatively small
(<span class="math notranslate nohighlight">\(10^4\)</span> or less in order of magnitude) and/or your
model has a small number of features (&lt;1000).</p></li>
<li><p>‘hdf5’ for anything larger and/or if you would like to
access the predictions/scores as a matrix that you can
easily filter, apply computations, or use in a subsequent
classifier/model. In this case, you may access the matrix
using <cite>mat[“data”]</cite> after opening the HDF5 file using
<cite>mat = h5py.File(“&lt;output.h5&gt;”, ‘r’)</cite>. The matrix columns
are the features and will match the same ordering as your
features .txt file (same as the order your model outputs
its predictions) and the matrix rows are the sequences.
Note that the row labels (FASTA description/IDs) will be
output as a separate .txt file (should match the ordering
of the sequences in the input FASTA).</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>strand_index</strong> (<em>int or None, optional</em>) – Default is None. If the trained model makes strand-specific
predictions, your input BED file may include a column with strand
information (strand must be one of {‘+’, ‘-‘, ‘.’}). Specify
the index (0-based) to use it. Otherwise, by default ‘+’ is used.
(This parameter is ignored if FASTA file is used as input.)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Writes the output to file(s) in <cite>output_dir</cite>. Filename will
match that specified in the filepath. In addition, if any base
in the given or retrieved sequence is unknown, the row labels .txt file
or .tsv file will mark this sequence or region as <cite>contains_unk = True</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="selene_sdk.predict.AnalyzeSequences.get_predictions_for_bed_file">
<code class="sig-name descname"><span class="pre">get_predictions_for_bed_file</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tsv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences.get_predictions_for_bed_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences.get_predictions_for_bed_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Get model predictions for sequences specified as genome coordinates
in a BED file. Coordinates do not need to be the same length as the
model expected sequence input–predictions will be centered at the
midpoint of the specified start and end coordinates.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>input_path</strong> (<em>str</em>) – Input path to the BED file.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Output directory to write the model predictions.</p></li>
<li><p><strong>output_format</strong> (<em>{‘tsv’, ‘hdf5’}, optional</em>) – Default is ‘tsv’. Choose whether to save TSV or HDF5 output files.
TSV is easier to access (i.e. open with text editor/Excel) and
quickly peruse, whereas HDF5 files must be accessed through
specific packages/viewers that support this format (e.g. h5py
Python package). Choose</p>
<blockquote>
<div><ul class="simple">
<li><p>‘tsv’ if your list of sequences is relatively small
(<span class="math notranslate nohighlight">\(10^4\)</span> or less in order of magnitude) and/or your
model has a small number of features (&lt;1000).</p></li>
<li><p>‘hdf5’ for anything larger and/or if you would like to
access the predictions/scores as a matrix that you can
easily filter, apply computations, or use in a subsequent
classifier/model. In this case, you may access the matrix
using <cite>mat[“data”]</cite> after opening the HDF5 file using
<cite>mat = h5py.File(“&lt;output.h5&gt;”, ‘r’)</cite>. The matrix columns
are the features and will match the same ordering as your
features .txt file (same as the order your model outputs
its predictions) and the matrix rows are the sequences.
Note that the row labels (FASTA description/IDs) will be
output as a separate .txt file (should match the ordering
of the sequences in the input FASTA).</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>strand_index</strong> (<em>int or None, optional</em>) – Default is None. If the trained model makes strand-specific
predictions, your input file may include a column with strand
information (strand must be one of {‘+’, ‘-‘, ‘.’}). Specify
the index (0-based) to use it. Otherwise, by default ‘+’ is used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Writes the output to file(s) in <cite>output_dir</cite>. Filename will
match that specified in the filepath.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="selene_sdk.predict.AnalyzeSequences.get_predictions_for_fasta_file">
<code class="sig-name descname"><span class="pre">get_predictions_for_fasta_file</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tsv'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences.get_predictions_for_fasta_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences.get_predictions_for_fasta_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Get model predictions for sequences in a FASTA file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>input_path</strong> (<em>str</em>) – Input path to the FASTA file.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Output directory to write the model predictions.</p></li>
<li><p><strong>output_format</strong> (<em>{‘tsv’, ‘hdf5’}, optional</em>) – Default is ‘tsv’. Choose whether to save TSV or HDF5 output files.
TSV is easier to access (i.e. open with text editor/Excel) and
quickly peruse, whereas HDF5 files must be accessed through
specific packages/viewers that support this format (e.g. h5py
Python package). Choose</p>
<blockquote>
<div><ul class="simple">
<li><p>‘tsv’ if your list of sequences is relatively small
(<span class="math notranslate nohighlight">\(10^4\)</span> or less in order of magnitude) and/or your
model has a small number of features (&lt;1000).</p></li>
<li><p>‘hdf5’ for anything larger and/or if you would like to
access the predictions/scores as a matrix that you can
easily filter, apply computations, or use in a subsequent
classifier/model. In this case, you may access the matrix
using <cite>mat[“data”]</cite> after opening the HDF5 file using
<cite>mat = h5py.File(“&lt;output.h5&gt;”, ‘r’)</cite>. The matrix columns
are the features and will match the same ordering as your
features .txt file (same as the order your model outputs
its predictions) and the matrix rows are the sequences.
Note that the row labels (FASTA description/IDs) will be
output as a separate .txt file (should match the ordering
of the sequences in the input FASTA).</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Writes the output to file(s) in <cite>output_dir</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis">
<code class="sig-name descname"><span class="pre">in_silico_mutagenesis</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_path_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ism'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_n_bases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tsv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences.in_silico_mutagenesis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies <em>in silico</em> mutagenesis to a sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – The sequence to mutate.</p></li>
<li><p><strong>save_data</strong> (<em>list(str)</em>) – A list of the data files to output. Must input 1 or more of the
following options: [“abs_diffs”, “diffs”, “logits”, “predictions”].</p></li>
<li><p><strong>output_path_prefix</strong> (<em>str, optional</em>) – The path to which the data files are written. If directories in
the path do not yet exist they will be automatically created.</p></li>
<li><p><strong>mutate_n_bases</strong> (<em>int, optional</em>) – The number of bases to mutate at one time. We recommend leaving
this parameter set to <cite>1</cite> at this time, as we have not yet
optimized operations for double and triple mutations.</p></li>
<li><p><strong>output_format</strong> (<em>{‘tsv’, ‘hdf5’}, optional</em>) – Default is ‘tsv’. The desired output format.</p></li>
<li><p><strong>start_position</strong> (<em>int, optional</em>) – Default is 0. The starting position of the subsequence to be
mutated.</p></li>
<li><p><strong>end_position</strong> (<em>int or None, optional</em>) – Default is None. The ending position of the subsequence to be
mutated. If left as <cite>None</cite>, then <cite>self.sequence_length</cite> will be
used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Outputs data files from <em>in silico</em> mutagenesis to <cite>output_dir</cite>.
For HDF5 output and ‘predictions’ in <cite>save_data</cite>, an additional
file named <cite>*_ref_predictions.h5</cite> will be outputted with the
model prediction for the original input sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If the value of <cite>start_position</cite> or <cite>end_position</cite> is negative.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If there are fewer than <cite>mutate_n_bases</cite> between <cite>start_position</cite>
    and <cite>end_position</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If <cite>start_position</cite> is greater or equal to <cite>end_position</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If <cite>start_position</cite> is not less than <cite>self.sequence_length</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If <cite>end_position</cite> is greater than <cite>self.sequence_length</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_from_file">
<code class="sig-name descname"><span class="pre">in_silico_mutagenesis_from_file</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_n_bases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_sequence_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tsv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences.in_silico_mutagenesis_from_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply <em>in silico</em> mutagenesis to all sequences in a FASTA file.</p>
<p>Please note that we have not parallelized this function yet, so runtime
increases exponentially when you increase <cite>mutate_n_bases</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_path</strong> (<em>str</em>) – The path to the FASTA file of sequences.</p></li>
<li><p><strong>save_data</strong> (<em>list(str)</em>) – A list of the data files to output. Must input 1 or more of the
following options: [“abs_diffs”, “diffs”, “logits”, “predictions”].</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The path to the output directory. Directories in the path will be
created if they do not currently exist.</p></li>
<li><p><strong>mutate_n_bases</strong> (<em>int, optional</em>) – Default is 1. The number of bases to mutate at one time in
<em>in silico</em> mutagenesis.</p></li>
<li><p><strong>use_sequence_name</strong> (<em>bool, optional.</em>) – Default is True. If <cite>use_sequence_name</cite>, output files are prefixed
by the sequence name/description corresponding to each sequence
in the FASTA file. Spaces in the sequence name are replaced with
underscores ‘_’. If not <cite>use_sequence_name</cite>, output files are
prefixed with an index <span class="math notranslate nohighlight">\(i\)</span> (starting with 0) corresponding
to the :math:<a href="#id1"><span class="problematic" id="id2">`</span></a>i`th sequence in the FASTA file.</p></li>
<li><p><strong>output_format</strong> (<em>{‘tsv’, ‘hdf5’}, optional</em>) – Default is ‘tsv’. The desired output format. Each sequence in
the FASTA file will have its own set of output files, where
the number of output files depends on the number of <cite>save_data</cite>
predictions/scores specified.</p></li>
<li><p><strong>start_position</strong> (<em>int, optional</em>) – Default is 0. The starting position of the subsequence to be
mutated.</p></li>
<li><p><strong>end_position</strong> (<em>int or None, optional</em>) – Default is None. The ending position of the subsequence to be
mutated. If left as <cite>None</cite>, then <cite>self.sequence_length</cite> will be
used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Outputs data files from <em>in silico</em> mutagenesis to <cite>output_dir</cite>.
For HDF5 output and ‘predictions’ in <cite>save_data</cite>, an additional
file named <cite>*_ref_predictions.h5</cite> will be outputted with the
model prediction for the original input sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If the value of <cite>start_position</cite> or <cite>end_position</cite> is negative.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If there are fewer than <cite>mutate_n_bases</cite> between <cite>start_position</cite>
    and <cite>end_position</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If <cite>start_position</cite> is greater or equal to <cite>end_position</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If <cite>start_position</cite> is not less than <cite>self.sequence_length</cite>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If <cite>end_position</cite> is greater than <cite>self.sequence_length</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_predict">
<code class="sig-name descname"><span class="pre">in_silico_mutagenesis_predict</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_preds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutations_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reporters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences.in_silico_mutagenesis_predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the predictions for all specified mutations applied
to a given sequence and, if applicable, compute the scores
(“abs_diffs”, “diffs”, “logits”) for these mutations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – The sequence to mutate.</p></li>
<li><p><strong>base_preds</strong> (<em>numpy.ndarray</em>) – The model’s prediction for <cite>sequence</cite>.</p></li>
<li><p><strong>mutations_list</strong> (<em>list(list(tuple))</em>) – The mutations to apply to the sequence. Each element in
<cite>mutations_list</cite> is a list of tuples, where each tuple
specifies the <cite>int</cite> position in the sequence to mutate and what
<cite>str</cite> base to which the position is mutated (e.g. (1, ‘A’)).</p></li>
<li><p><strong>reporters</strong> (<em>list(PredictionsHandler)</em>) – The list of reporters, where each reporter handles the predictions
made for each mutated sequence. Will collect, compute scores
(e.g. <cite>AbsDiffScoreHandler</cite> computes the absolute difference
between <cite>base_preds</cite> and the predictions for the mutated
sequence), and output these as a file at the end.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Writes results to files corresponding to each reporter in
<cite>reporters</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="selene_sdk.predict.AnalyzeSequences.variant_effect_prediction">
<code class="sig-name descname"><span class="pre">variant_effect_prediction</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vcf_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tsv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">require_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/predict/model_predict.html#AnalyzeSequences.variant_effect_prediction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.predict.AnalyzeSequences.variant_effect_prediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Get model predictions and scores for a list of variants.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>vcf_file</strong> (<em>str</em>) – Path to a VCF file. Must contain the columns
[#CHROM, POS, ID, REF, ALT], in order. Column header does not need
to be present.</p></li>
<li><p><strong>save_data</strong> (<em>list(str)</em>) – A list of the data files to output. Must input 1 or more of the
following options: [“abs_diffs”, “diffs”, “logits”, “predictions”].</p></li>
<li><p><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. Path to the output directory. If no path is
specified, will save files corresponding to the options in
<cite>save_data</cite> to the current working directory.</p></li>
<li><p><strong>output_format</strong> (<em>{‘tsv’, ‘hdf5’}, optional</em>) – Default is ‘tsv’. Choose whether to save TSV or HDF5 output files.
TSV is easier to access (i.e. open with text editor/Excel) and
quickly peruse, whereas HDF5 files must be accessed through
specific packages/viewers that support this format (e.g. h5py
Python package). Choose</p>
<blockquote>
<div><ul class="simple">
<li><p>‘tsv’ if your list of variants is relatively small
(<span class="math notranslate nohighlight">\(10^4\)</span> or less in order of magnitude) and/or your
model has a small number of features (&lt;1000).</p></li>
<li><p>‘hdf5’ for anything larger and/or if you would like to
access the predictions/scores as a matrix that you can
easily filter, apply computations, or use in a subsequent
classifier/model. In this case, you may access the matrix
using <cite>mat[“data”]</cite> after opening the HDF5 file using
<cite>mat = h5py.File(“&lt;output.h5&gt;”, ‘r’)</cite>. The matrix columns
are the features and will match the same ordering as your
features .txt file (same as the order your model outputs
its predictions) and the matrix rows are the sequences.
Note that the row labels (chrom, pos, id, ref, alt) will be
output as a separate .txt file.</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>strand_index</strong> (<em>int or None, optional.</em>) – Default is None. If applicable, specify the column index (0-based)
in the VCF file that contains strand information for each variant.</p></li>
<li><p><strong>require_strand</strong> (<em>bool, optional.</em>) – Default is False. Whether strand can be specified as ‘.’. If False,
Selene accepts strand value to be ‘+’, ‘-‘, or ‘.’ and automatically
treats ‘.’ as ‘+’. If True, Selene skips any variant with strand ‘.’.
This parameter assumes that <cite>strand_index</cite> has been set.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Saves all files to <cite>output_dir</cite>. If any bases in the ‘ref’ column
of the VCF do not match those at the specified position in the
reference genome, the row labels .txt file will mark this variant
as <cite>ref_match = False</cite>. If most of your variants do not match
the reference genome, please check that the reference genome
you specified matches the version with which the variants were
called. The predictions can used directly if you have verified that
the ‘ref’ bases specified for these variants are correct (Selene
will have substituted these bases for those in the reference
genome). In addition, if any base in the retrieved reference
sequence is unknown, the row labels .txt file will mark this variant
as <cite>contains_unk = True</cite>. Finally, some variants may show up in an
‘NA’ file. This is because the surrounding sequence context ended up
being out of bounds or overlapping with blacklist regions  or the
chromosome containing the variant did not show up in the reference
genome FASTA file.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="samplers.html" class="btn btn-neutral float-right" title="selene_sdk.samplers" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="interpret.html" class="btn btn-neutral float-left" title="selene_sdk.interpret" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>