

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>selene_sdk.sequences &mdash; Selene 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="_static/css/selene_docs_theme.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="selene_sdk.targets" href="targets.html" />
    <link rel="prev" title="selene_sdk.samplers.file_samplers" href="samplers.file_samplers.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Selene
          

          
            
            <img src="_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-selene-with-pip">Installing selene with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="overview/cli.html#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>selene_sdk.sequences</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/sequences.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene_sdk.sequences">
<span id="selene-sdk-sequences"></span><h1>selene_sdk.sequences<a class="headerlink" href="#module-selene_sdk.sequences" title="Permalink to this headline">¶</a></h1>
<p>This module provides the types for representing biological sequences.</p>
<div class="section" id="sequence">
<h2>Sequence<a class="headerlink" href="#sequence" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">selene_sdk.sequences.</span></span><span class="sig-name descname"><span class="pre">Sequence</span></span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#Sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The abstract base class for biological sequence classes.</p>
<dl class="py property">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.BASES_ARR">
<em class="property"><span class="pre">abstract</span> <span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">BASES_ARR</span></span><a class="headerlink" href="#selene_sdk.sequences.Sequence.BASES_ARR" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an array with the alphabet (i.e. all possible symbols
that may occur in a sequence). We expect that
<cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all valid <cite>i</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The array of all members of the alphabet.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.BASE_TO_INDEX">
<em class="property"><span class="pre">abstract</span> <span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">BASE_TO_INDEX</span></span><a class="headerlink" href="#selene_sdk.sequences.Sequence.BASE_TO_INDEX" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping members of the alphabet (i.e. all
possible symbols that can occur in a sequence) to integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The dictionary mapping the alphabet to integers.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.INDEX_TO_BASE">
<em class="property"><span class="pre">abstract</span> <span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">INDEX_TO_BASE</span></span><a class="headerlink" href="#selene_sdk.sequences.Sequence.INDEX_TO_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping integers to members of the alphabet (i.e.
all possible symbols that can occur in a sequence). We expect
that <cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all
valid <cite>i</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The dictionary mapping integers to the alphabet.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.UNK_BASE">
<em class="property"><span class="pre">abstract</span> <span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">UNK_BASE</span></span><a class="headerlink" href="#selene_sdk.sequences.Sequence.UNK_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a base used to represent unknown positions. This is not
the same as a character from outside the sequence’s alphabet. A
character from outside the alphabet is an error. A position with
an unknown base signifies that the position is one of the bases
from the alphabet, but we are uncertain which.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The character representing an unknown base.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.coords_in_bounds">
<em class="property"><span class="pre">abstract</span> </em><span class="sig-name descname"><span class="pre">coords_in_bounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#Sequence.coords_in_bounds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Sequence.coords_in_bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if queried coordinates are valid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><cite>True</cite> if the coordinates are in bounds, otherwise <cite>False</cite>.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.encoding_to_sequence">
<em class="property"><span class="pre">abstract</span> <span class="pre">classmethod</span> </em><span class="sig-name descname"><span class="pre">encoding_to_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">encoding</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#Sequence.encoding_to_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Sequence.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms the input numerical representation of a sequence
into a string representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is
the size of the sequence type’s alphabet.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sequence of bases decoded from the input array. This
sequence will be of length <span class="math notranslate nohighlight">\(L\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.get_encoding_from_coords">
<em class="property"><span class="pre">abstract</span> </em><span class="sig-name descname"><span class="pre">get_encoding_from_coords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#Sequence.get_encoding_from_coords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Sequence.get_encoding_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts the numerical encoding for a sequence occurring at
the given coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence occuring
at queried coordinates, where <span class="math notranslate nohighlight">\(L\)</span> is the length of the
sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s
alphabet. Behavior is undefined for invalid coordinates.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.get_sequence_from_coords">
<em class="property"><span class="pre">abstract</span> </em><span class="sig-name descname"><span class="pre">get_sequence_from_coords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#Sequence.get_sequence_from_coords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Sequence.get_sequence_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a string representation of a sequence at the
given coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The sequence of bases occuring at the queried coordinates.
This sequence will be of length <span class="math notranslate nohighlight">\(L\)</span> normally, but
only if the coordinates are valid. Behavior is undefined for
invalid coordinates.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Sequence.sequence_to_encoding">
<em class="property"><span class="pre">abstract</span> <span class="pre">classmethod</span> </em><span class="sig-name descname"><span class="pre">sequence_to_encoding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#Sequence.sequence_to_encoding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Sequence.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms a biological sequence into a numerical
representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<em>str</em>) – The input sequence of characters.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is
the size of the sequence type’s alphabet.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="genome">
<h2>Genome<a class="headerlink" href="#genome" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">selene_sdk.sequences.</span></span><span class="sig-name descname"><span class="pre">Genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bases_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_unpicklable</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#selene_sdk.sequences.Sequence" title="selene_sdk.sequences.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.sequences.sequence.Sequence</span></code></a></p>
<p>This class provides access to an organism’s genomic sequence.</p>
<p>This class supports retrieving parts of the sequence and converting
these parts into their one-hot encodings. It is essentially a
wrapper class around the <cite>pyfaidx.Fasta</cite> class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_path</strong> (<em>str</em>) – Path to an indexed FASTA file, that is, a <cite>*.fasta</cite> file with
a corresponding <cite>*.fai</cite> file in the same directory. This file
should contain the target organism’s genome sequence.</p></li>
<li><p><strong>blacklist_regions</strong> (<em>str or None, optional</em>) – Default is None. Path to a tabix-indexed list of regions from
which we should not output sequences. This is used to ensure that
we are not sampling from areas where we will never collect
measurements. You can pass as input “hg19” or “hg38” to use the
blacklist regions released by ENCODE. You can also pass in your own
tabix-indexed .gz file.</p></li>
<li><p><strong>bases_order</strong> (<em>list(str) or None, optional</em>) – Default is None (use the default base ordering of
<cite>[‘A’, ‘C’, ‘G’, ‘T’]</cite>). Specify a different ordering of
DNA bases for one-hot encoding.</p></li>
<li><p><strong>init_unpicklable</strong> (<em>bool, optional</em>) – Default is False. Delays initialization until a relevant method
is called. This enables the object to be pickled after instantiation.
<cite>init_unpicklable</cite> must be <cite>False</cite> when multi-processing is needed e.g.
DataLoader. Set <cite>init_unpicklable</cite> to True if you are using this class
directly through Selene’s API and want to access class attributes
without having to call on a specific method in Genome.</p></li>
</ul>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>~Genome.genome</strong> (<em>pyfaidx.Fasta</em>) – The FASTA file containing the genome sequence.</p></li>
<li><p><strong>~Genome.chrs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – The list of chromosome names.</p></li>
<li><p><strong>~Genome.len_chrs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – A dictionary mapping the names of each chromosome in the file to
the length of said chromosome.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.BASES_ARR">
<span class="sig-name descname"><span class="pre">BASES_ARR</span></span><em class="property"> <span class="pre">=</span> <span class="pre">['A',</span> <span class="pre">'C',</span> <span class="pre">'G',</span> <span class="pre">'T']</span></em><a class="headerlink" href="#selene_sdk.sequences.Genome.BASES_ARR" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an array with the alphabet (i.e. all possible symbols
that may occur in a sequence). We expect that
<cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all valid <cite>i</cite>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.BASE_TO_INDEX">
<span class="sig-name descname"><span class="pre">BASE_TO_INDEX</span></span><em class="property"> <span class="pre">=</span> <span class="pre">{'A':</span> <span class="pre">0,</span> <span class="pre">'C':</span> <span class="pre">1,</span> <span class="pre">'G':</span> <span class="pre">2,</span> <span class="pre">'T':</span> <span class="pre">3,</span> <span class="pre">'a':</span> <span class="pre">0,</span> <span class="pre">'c':</span> <span class="pre">1,</span> <span class="pre">'g':</span> <span class="pre">2,</span> <span class="pre">'t':</span> <span class="pre">3}</span></em><a class="headerlink" href="#selene_sdk.sequences.Genome.BASE_TO_INDEX" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping members of the alphabet (i.e. all
possible symbols that can occur in a sequence) to integers.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.COMPLEMENTARY_BASE_DICT">
<span class="sig-name descname"><span class="pre">COMPLEMENTARY_BASE_DICT</span></span><em class="property"> <span class="pre">=</span> <span class="pre">{'A':</span> <span class="pre">'T',</span> <span class="pre">'C':</span> <span class="pre">'G',</span> <span class="pre">'G':</span> <span class="pre">'C',</span> <span class="pre">'N':</span> <span class="pre">'N',</span> <span class="pre">'T':</span> <span class="pre">'A',</span> <span class="pre">'a':</span> <span class="pre">'T',</span> <span class="pre">'c':</span> <span class="pre">'G',</span> <span class="pre">'g':</span> <span class="pre">'C',</span> <span class="pre">'n':</span> <span class="pre">'N',</span> <span class="pre">'t':</span> <span class="pre">'A'}</span></em><a class="headerlink" href="#selene_sdk.sequences.Genome.COMPLEMENTARY_BASE_DICT" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping each base to its complementary base.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.INDEX_TO_BASE">
<span class="sig-name descname"><span class="pre">INDEX_TO_BASE</span></span><em class="property"> <span class="pre">=</span> <span class="pre">{0:</span> <span class="pre">'A',</span> <span class="pre">1:</span> <span class="pre">'C',</span> <span class="pre">2:</span> <span class="pre">'G',</span> <span class="pre">3:</span> <span class="pre">'T'}</span></em><a class="headerlink" href="#selene_sdk.sequences.Genome.INDEX_TO_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping integers to members of the alphabet (i.e.
all possible symbols that can occur in a sequence). We expect
that <cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all
valid <cite>i</cite>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.UNK_BASE">
<span class="sig-name descname"><span class="pre">UNK_BASE</span></span><em class="property"> <span class="pre">=</span> <span class="pre">'N'</span></em><a class="headerlink" href="#selene_sdk.sequences.Genome.UNK_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a base used to represent unknown positions. This is not
the same as a character from outside the sequence’s alphabet. A
character from outside the alphabet is an error. A position with
an unknown base signifies that the position is one of the bases
from the alphabet, but we are uncertain which.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.coords_in_bounds">
<span class="sig-name descname"><span class="pre">coords_in_bounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.coords_in_bounds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.coords_in_bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the region we want to query is within the bounds of the
queried chromosome and non-overlapping with blacklist regions
(if given).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em>) – The name of the chromosomes, e.g. “chr1”.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the sequence.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Whether we can retrieve a sequence from the bounds specified
in the input.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.encoding_to_sequence">
<em class="property"><span class="pre">classmethod</span> </em><span class="sig-name descname"><span class="pre">encoding_to_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">encoding</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.encoding_to_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input one-hot encoding to its DNA sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – An <span class="math notranslate nohighlight">\(L \times 4\)</span> one-hot encoding of the sequence,
where <span class="math notranslate nohighlight">\(L\)</span> is the length of the output sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sequence of <span class="math notranslate nohighlight">\(L\)</span> nucleotides decoded from the
input array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.get_chr_lens">
<span class="sig-name descname"><span class="pre">get_chr_lens</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.get_chr_lens"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.get_chr_lens" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the name and length of each chromosome sequence in the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of tuples of the chromosome names and lengths.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>))</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.get_chrs">
<span class="sig-name descname"><span class="pre">get_chrs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.get_chrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.get_chrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the list of chromosome names.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of the chromosome names.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.get_encoding_from_coords">
<span class="sig-name descname"><span class="pre">get_encoding_from_coords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.get_encoding_from_coords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.get_encoding_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the one-hot encoding of the genomic sequence at the
queried coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em>) – The name of the chromosome or region, e.g. “chr1”.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</p></li>
<li><p><strong>strand</strong> (<em>{‘+’, ‘-‘, ‘.’}, optional</em>) – Default is ‘+’. The strand the sequence is located on. ‘.’ is
treated as ‘+’.</p></li>
<li><p><strong>pad</strong> (<em>bool, optional</em>) – Default is <cite>False</cite>. Pad the output sequence with ‘N’ if <cite>start</cite>
and/or <cite>end</cite> are out of bounds to return a sequence of length
<cite>end - start</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <span class="math notranslate nohighlight">\(L \times 4\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L = end - start\)</span>, unless <cite>chrom</cite> cannot be found
in the input FASTA, <cite>start</cite> or <cite>end</cite> are out of bounds,
or (if a blacklist exists) the region overlaps with a blacklist
region. In these cases, it will return an empty encoding–that is,
<cite>L</cite> = 0 for the NumPy array returned.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If the input char to <cite>strand</cite> is not one of the specified
    choices.
    (Raised in the call to <cite>self.get_sequence_from_coords</cite>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.get_encoding_from_coords_check_unk">
<span class="sig-name descname"><span class="pre">get_encoding_from_coords_check_unk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.get_encoding_from_coords_check_unk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.get_encoding_from_coords_check_unk" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the one-hot encoding of the genomic sequence at the
queried coordinates and check whether the sequence contains
unknown base(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em>) – The name of the chromosome or region, e.g. “chr1”.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</p></li>
<li><p><strong>strand</strong> (<em>{‘+’, ‘-‘, ‘.’}, optional</em>) – Default is ‘+’. The strand the sequence is located on. ‘.’ is
treated as ‘+’.</p></li>
<li><p><strong>pad</strong> (<em>bool, optional</em>) – Default is <cite>False</cite>. Pad the output sequence with ‘N’ if <cite>start</cite>
and/or <cite>end</cite> are out of bounds to return a sequence of length
<cite>end - start</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><cite>tuple[0]</cite> is the <span class="math notranslate nohighlight">\(L \times 4\)</span> encoding of the sequence</p></li>
</ul>
<p>containing data of <cite>numpy.float32</cite> type, where
<span class="math notranslate nohighlight">\(L = end - start\)</span>, unless <cite>chrom</cite> cannot be found
in the input FASTA, <cite>start</cite> or <cite>end</cite> are out of bounds,
or (if a blacklist exists) the region overlaps with a blacklist
region. In these cases, it will return an empty encoding–that is,
<cite>L</cite> = 0 for the NumPy array returned.
* <cite>tuple[1]</cite> is the boolean value that indicates whether the
sequence contains any unknown base(s) specified in self.UNK_BASE</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If the input char to <cite>strand</cite> is not one of the specified
    choices.
    (Raised in the call to <cite>self.get_sequence_from_coords</cite>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.get_sequence_from_coords">
<span class="sig-name descname"><span class="pre">get_sequence_from_coords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.get_sequence_from_coords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.get_sequence_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the queried chromosome’s sequence at the input coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em>) – The name of the chromosomes, e.g. “chr1”.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the sequence.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</p></li>
<li><p><strong>strand</strong> (<em>{‘+’, ‘-‘, ‘.’}, optional</em>) – Default is ‘+’. The strand the sequence is located on. ‘.’ is
treated as ‘.’.</p></li>
<li><p><strong>pad</strong> (<em>bool, optional</em>) – Default is <cite>False</cite>. Pad the output sequence with ‘N’ if <cite>start</cite>
and/or <cite>end</cite> are out of bounds to return a sequence of length
<cite>end - start</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The genomic sequence of length <span class="math notranslate nohighlight">\(L\)</span> where
<span class="math notranslate nohighlight">\(L = end - start\)</span>. If <cite>pad</cite> is <cite>False</cite> and one/both of
<cite>start</cite> and <cite>end</cite> are out of bounds, will return an empty string.
Also returns an empty string if <cite>chrom</cite> cannot be found in the
input FASTA file.
Otherwise, will return the sequence with padding at the start/end
if appropriate.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If the input char to <cite>strand</cite> is not one of the specified
    choices.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Genome.sequence_to_encoding">
<em class="property"><span class="pre">classmethod</span> </em><span class="sig-name descname"><span class="pre">sequence_to_encoding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/genome.html#Genome.sequence_to_encoding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Genome.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input sequence to its one-hot encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<em>str</em>) – A nucleotide sequence of length <span class="math notranslate nohighlight">\(L\)</span></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <span class="math notranslate nohighlight">\(L \times 4\)</span> one-hot encoding of the sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="proteome">
<h2>Proteome<a class="headerlink" href="#proteome" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">selene_sdk.sequences.</span></span><span class="sig-name descname"><span class="pre">Proteome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#selene_sdk.sequences.Sequence" title="selene_sdk.sequences.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.sequences.sequence.Sequence</span></code></a></p>
<p>Provides access to an organism’s proteomic sequence.</p>
<p>It supports retrieving parts of the sequence and converting these
parts into their one-hot encodings. It is essentially a wrapper
class around the <cite>pyfaidx.Fasta</cite> class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>input_path</strong> (<em>str</em>) – Path to an indexed FASTA file containing amino acid
sequences, that is, a <cite>*.fasta</cite> file with a corresponding
<cite>*.fai</cite> file in the same directory. File should contain the
sequences from which training examples will be created.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>~Proteome.proteome</strong> (<em>pyfaidx.Fasta</em>) – The FASTA or FAA file containing the protein sequences.</p></li>
<li><p><strong>~Proteome.prots</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – The list of protein names.</p></li>
<li><p><strong>~Proteome.len_prots</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – A dictionary that maps protein names to the lengths, and does so
for all protein sequences in the proteome.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.BASES_ARR">
<span class="sig-name descname"><span class="pre">BASES_ARR</span></span><em class="property"> <span class="pre">=</span> <span class="pre">['A',</span> <span class="pre">'R',</span> <span class="pre">'N',</span> <span class="pre">'D',</span> <span class="pre">'C',</span> <span class="pre">'E',</span> <span class="pre">'Q',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'L',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'F',</span> <span class="pre">'P',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'W',</span> <span class="pre">'Y',</span> <span class="pre">'V']</span></em><a class="headerlink" href="#selene_sdk.sequences.Proteome.BASES_ARR" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an array with the alphabet (i.e. all possible symbols
that may occur in a sequence). We expect that
<cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all valid <cite>i</cite>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.BASE_TO_INDEX">
<span class="sig-name descname"><span class="pre">BASE_TO_INDEX</span></span><em class="property"> <span class="pre">=</span> <span class="pre">{'A':</span> <span class="pre">0,</span> <span class="pre">'C':</span> <span class="pre">4,</span> <span class="pre">'D':</span> <span class="pre">3,</span> <span class="pre">'E':</span> <span class="pre">5,</span> <span class="pre">'F':</span> <span class="pre">13,</span> <span class="pre">'G':</span> <span class="pre">7,</span> <span class="pre">'H':</span> <span class="pre">8,</span> <span class="pre">'I':</span> <span class="pre">9,</span> <span class="pre">'K':</span> <span class="pre">11,</span> <span class="pre">'L':</span> <span class="pre">10,</span> <span class="pre">'M':</span> <span class="pre">12,</span> <span class="pre">'N':</span> <span class="pre">2,</span> <span class="pre">'P':</span> <span class="pre">14,</span> <span class="pre">'Q':</span> <span class="pre">6,</span> <span class="pre">'R':</span> <span class="pre">1,</span> <span class="pre">'S':</span> <span class="pre">15,</span> <span class="pre">'T':</span> <span class="pre">16,</span> <span class="pre">'V':</span> <span class="pre">19,</span> <span class="pre">'W':</span> <span class="pre">17,</span> <span class="pre">'Y':</span> <span class="pre">18}</span></em><a class="headerlink" href="#selene_sdk.sequences.Proteome.BASE_TO_INDEX" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping members of the alphabet (i.e. all
possible symbols that can occur in a sequence) to integers.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.INDEX_TO_BASE">
<span class="sig-name descname"><span class="pre">INDEX_TO_BASE</span></span><em class="property"> <span class="pre">=</span> <span class="pre">{0:</span> <span class="pre">'A',</span> <span class="pre">1:</span> <span class="pre">'R',</span> <span class="pre">2:</span> <span class="pre">'N',</span> <span class="pre">3:</span> <span class="pre">'D',</span> <span class="pre">4:</span> <span class="pre">'C',</span> <span class="pre">5:</span> <span class="pre">'E',</span> <span class="pre">6:</span> <span class="pre">'Q',</span> <span class="pre">7:</span> <span class="pre">'G',</span> <span class="pre">8:</span> <span class="pre">'H',</span> <span class="pre">9:</span> <span class="pre">'I',</span> <span class="pre">10:</span> <span class="pre">'L',</span> <span class="pre">11:</span> <span class="pre">'K',</span> <span class="pre">12:</span> <span class="pre">'M',</span> <span class="pre">13:</span> <span class="pre">'F',</span> <span class="pre">14:</span> <span class="pre">'P',</span> <span class="pre">15:</span> <span class="pre">'S',</span> <span class="pre">16:</span> <span class="pre">'T',</span> <span class="pre">17:</span> <span class="pre">'W',</span> <span class="pre">18:</span> <span class="pre">'Y',</span> <span class="pre">19:</span> <span class="pre">'V'}</span></em><a class="headerlink" href="#selene_sdk.sequences.Proteome.INDEX_TO_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping integers to members of the alphabet (i.e.
all possible symbols that can occur in a sequence). We expect
that <cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all
valid <cite>i</cite>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.UNK_BASE">
<span class="sig-name descname"><span class="pre">UNK_BASE</span></span><em class="property"> <span class="pre">=</span> <span class="pre">'X'</span></em><a class="headerlink" href="#selene_sdk.sequences.Proteome.UNK_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a base used to represent unknown positions. This is not
the same as a character from outside the sequence’s alphabet. A
character from outside the alphabet is an error. A position with
an unknown base signifies that the position is one of the bases
from the alphabet, but we are uncertain which.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.coords_in_bounds">
<span class="sig-name descname"><span class="pre">coords_in_bounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome.coords_in_bounds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome.coords_in_bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the coordinates we want to query is valid.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prot</strong> (<em>str</em>) – The name of the protein, e.g. “YFP”.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A boolean indicating whether we can retrieve a sequence from
the queried coordinates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.encoding_to_sequence">
<em class="property"><span class="pre">classmethod</span> </em><span class="sig-name descname"><span class="pre">encoding_to_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">encoding</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome.encoding_to_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input one-hot encoding to its amino acid sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – The <span class="math notranslate nohighlight">\(L \times 20\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the output amino acid sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sequence of <span class="math notranslate nohighlight">\(L\)</span> amino acids decoded from the
input array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.get_encoding_from_coords">
<span class="sig-name descname"><span class="pre">get_encoding_from_coords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome.get_encoding_from_coords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome.get_encoding_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the one-hot encoding of the protein’s sequence at the
input coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prot</strong> (<em>str</em>) – The name of the protein, e.g. “YFP”.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <span class="math notranslate nohighlight">\(L \times 20\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L = end - start\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.get_prot_lens">
<span class="sig-name descname"><span class="pre">get_prot_lens</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome.get_prot_lens"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome.get_prot_lens" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the name and length of each protein sequence in the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of tuples of protein names and protein lengths.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>))</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.get_prots">
<span class="sig-name descname"><span class="pre">get_prots</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome.get_prots"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome.get_prots" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the list of protein names.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of the protein names.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.get_sequence_from_coords">
<span class="sig-name descname"><span class="pre">get_sequence_from_coords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome.get_sequence_from_coords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome.get_sequence_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the queried protein sequence at the input coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prot</strong> (<em>str</em>) – The protein name, e.g. “YFP”.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sequence of <span class="math notranslate nohighlight">\(L\)</span> amino acids at the specified
coordinates, where <span class="math notranslate nohighlight">\(L = end - start\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="selene_sdk.sequences.Proteome.sequence_to_encoding">
<em class="property"><span class="pre">classmethod</span> </em><span class="sig-name descname"><span class="pre">sequence_to_encoding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/proteome.html#Proteome.sequence_to_encoding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.Proteome.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input sequence to its one-hot encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<em>str</em>) – The input sequence of amino acids of length <span class="math notranslate nohighlight">\(L\)</span>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <span class="math notranslate nohighlight">\(L \times 20\)</span> array, where <cite>L</cite> was the length of
the input sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="sequence-to-encoding">
<h2>sequence_to_encoding<a class="headerlink" href="#sequence-to-encoding" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.sequences.sequence_to_encoding">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.sequences.</span></span><span class="sig-name descname"><span class="pre">sequence_to_encoding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_to_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bases_arr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#sequence_to_encoding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input sequence to its one-hot encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – The input sequence of length <span class="math notranslate nohighlight">\(L\)</span>.</p></li>
<li><p><strong>base_to_index</strong> (<em>dict</em>) – A dict that maps input characters to indices, where the indices
specify the column to assign as 1 when a base exists at the
current position in the input. If a base does not exist at the
current position in the input, it’s corresponding column in the
encoding is set as zero. Note that the rows correspond directly
to the positions in the input sequence. For instance, with a
a genome you would have each of <cite>[‘A’, ‘C’, ‘G’, ‘T’]</cite> as keys,
mapping to values of <cite>[0, 1, 2, 3]</cite>.</p></li>
<li><p><strong>bases_arr</strong> (<em>list(str)</em>) – The characters in the sequence’s alphabet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the input sequence and <span class="math notranslate nohighlight">\(N\)</span> is
the size of the sequence alphabet.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="encoding-to-sequence">
<h2>encoding_to_sequence<a class="headerlink" href="#encoding-to-sequence" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.sequences.encoding_to_sequence">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.sequences.</span></span><span class="sig-name descname"><span class="pre">encoding_to_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">encoding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bases_arr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unk_base</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#encoding_to_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a sequence one-hot encoding to its string sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the
size of the sequence alphabet.</p></li>
<li><p><strong>bases_arr</strong> (<em>list(str)</em>) – A list of the bases in the sequence’s alphabet that corresponds
to the correct columns for those bases in the encoding.</p></li>
<li><p><strong>unk_base</strong> (<em>str</em>) – The base corresponding to the “unknown” character in this
encoding. See <cite>selene_sdk.sequences.Sequence.UNK_BASE</cite> for more
information.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sequence of <span class="math notranslate nohighlight">\(L\)</span> characters decoded from the
input array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="get-reverse-encoding">
<h2>get_reverse_encoding<a class="headerlink" href="#get-reverse-encoding" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="selene_sdk.sequences.get_reverse_encoding">
<span class="sig-prename descclassname"><span class="pre">selene_sdk.sequences.</span></span><span class="sig-name descname"><span class="pre">get_reverse_encoding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">encoding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bases_arr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_to_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">complementary_base_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/sequences/sequence.html#get_reverse_encoding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#selene_sdk.sequences.get_reverse_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>The Genome DNA bases encoding is created such that the reverse
encoding can be quickly computed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>encoding</strong> (<em>numpy.ndarray</em>)</p></li>
<li><p><strong>bases_arr</strong> (<em>list(str)</em>)</p></li>
<li><p><strong>base_to_index</strong> (<em>dict</em>)</p></li>
<li><p><strong>complementary_base_dict</strong> (<em>dict</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="targets.html" class="btn btn-neutral float-right" title="selene_sdk.targets" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="samplers.file_samplers.html" class="btn btn-neutral float-left" title="selene_sdk.samplers.file_samplers" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2018, Selene Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>