

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Selene CLI operations and outputs &mdash; Selene 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../_static/css/selene_docs_theme.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tutorials" href="tutorials.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html#help">Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#installing-selene-with-pip">Installing selene with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#additional-dependency-for-running-the-cli-versions-0-4-8-and-below">Additional dependency for running the CLI (versions 0.4.8 and below)</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id7">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id8"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id15">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id16">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Selene CLI operations and outputs</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/overview/cli.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="selene-cli-operations-and-outputs">
<h1>Selene CLI operations and outputs<a class="headerlink" href="#selene-cli-operations-and-outputs" title="Permalink to this headline">¶</a></h1>
<p>Selene provides a command-line interface (CLI) that takes in a user-specified configuration file containing the operations the user wants to run and the parameters required for these operations. It is automatically installed using a setuptools entrypoint so it can be called with <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">-m</span> <span class="pre">selene_sdk</span></code>.</p>
<p>The sections that follow describe in detail how the various components that make up the configuration file are specified. For operation-specific sections (e.g. training, evaluation), we also explain what the expected outputs are.</p>
<p>We strongly recommend you read through the first 4 sections (<a class="reference internal" href="#overview"><span class="std std-ref">Overview</span></a>, <a class="reference internal" href="#operations"><span class="std std-ref">Operations</span></a>, <a class="reference internal" href="#general-configurations"><span class="std std-ref">General configurations</span></a>, and <a class="reference internal" href="#model-architecture"><span class="std std-ref">Model architecture</span></a>) and then pick other sections based on your use case.</p>
<p>Join our <a class="reference external" href="https://groups.google.com/forum/#!forum/selene-sdk">Google group</a> if you have questions about the package, case studies, or model development.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Selene’s CLI accepts configuration files in the <a class="reference external" href="https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html">YAML</a> format that are composed of 4 main (high-level) groups:</p>
<ol class="arabic simple">
<li><p>list of operations</p></li>
<li><p>general configuration parameters</p></li>
<li><p>model configuration</p></li>
<li><p>operation-specific configurations</p></li>
</ol>
<p>“Operation-specific configurations” require you to specify the input parameters for different classes and methods that we have implemented in Selene. These configurations are parsed using code adapted from the <a class="reference external" href="http://deeplearning.net/software/pylearn2/yaml_tutorial/index.html#yaml-tutorial">Pylearn2 library</a> and will instantiate the appropriate Python object or function based on your inputs. You may use Selene’s <a class="reference external" href="https://selene.flatironinstitute.org">API documentation</a> to determine what parameters are accepted by the constructors/methods implemented in Selene. For your convenience, we have created this document to specifically describe the parameters necessary to build configuration files for the Selene CLI.</p>
<p>We recommend you start off by using one of the <a class="reference external" href="https://github.com/FunctionLab/selene/tree/master/config_examples">example configuration files</a> provided in the repository as a template for your own configuration file:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/config_examples/train.yml">Training configuration</a></p></li>
<li><p><a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/config_examples/evaluate_test_bed.yml">Evaluate with test BED file</a></p></li>
<li><p><a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/config_examples/evaluate_test_mat.yml">Evaluate with test matrix file</a></p></li>
<li><p><a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/config_examples/get_predictions.yml">Get predictions from trained model</a></p></li>
<li><p><a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/config_examples/in_silico_mutagenesis.yml">*In silico* mutagenesis</a></p></li>
<li><p><a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/config_examples/variant_effect_prediction.yml">Variant effect prediction</a></p></li>
</ul>
<p>There are also various configuration files associated with the Jupyter notebook <a class="reference external" href="https://github.com/FunctionLab/selene/tree/master/tutorials">tutorials</a> and <a class="reference external" href="https://github.com/FunctionLab/selene/tree/master/manuscript">manuscript</a> case studies that you may use as a starting point.</p>
</div>
<div class="section" id="operations">
<h2>Operations<a class="headerlink" href="#operations" title="Permalink to this headline">¶</a></h2>
<p>Every file should start with the operations that you want to run.</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">ops</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">train</span><span class="p p-Indicator">,</span> <span class="nv">evaluate</span><span class="p p-Indicator">,</span> <span class="nv">analyze</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">ops</span></code> key expects one or more of <code class="docutils literal notranslate"><span class="pre">[train,</span> <span class="pre">evaluate,</span> <span class="pre">analyze]</span></code> to be specified as a list. In addition to the general and model architecture configurations described in the next 2 sections, each of these operations will require some additional set of configurations attached to the following keys:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">train</span></code>: <code class="docutils literal notranslate"><span class="pre">train_model</span></code> (see <a class="reference internal" href="#train"><span class="std std-ref">Train</span></a>) and <code class="docutils literal notranslate"><span class="pre">sampler</span></code> (see <a class="reference internal" href="#samplers-used-for-training-and-evaluation-optionally"><span class="std std-ref">Samplers used for training (and evaluation, optionally)</span></a>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">evaluate</span></code>: <code class="docutils literal notranslate"><span class="pre">evaluate_model</span></code> (see <a class="reference internal" href="#evaluate"><span class="std std-ref">Evaluate</span></a>) and <code class="docutils literal notranslate"><span class="pre">sampler</span></code> (see <a class="reference internal" href="#samplers-used-for-evaluation"><span class="std std-ref">Samplers used for evaluation</span></a>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">analyze</span></code>: <code class="docutils literal notranslate"><span class="pre">analyze_sequences</span></code> (see <a class="reference internal" href="#analyze-sequences"><span class="std std-ref">Analyze sequences</span></a>)</p></li>
</ul>
<p><strong>Note</strong>: You should be able to use multiple operations (i.e. specify the necessary configuration keys for those operations in a single file). However, if <code class="docutils literal notranslate"><span class="pre">[train,</span> <span class="pre">evaluate]</span></code> are both specified, we expect that they will both rely on the same sampler. If you need to train and evaluate using different samplers, please create 2 separate YAML files.</p>
</div>
<div class="section" id="general-configurations">
<h2>General configurations<a class="headerlink" href="#general-configurations" title="Permalink to this headline">¶</a></h2>
<p>In addition to the <code class="docutils literal notranslate"><span class="pre">ops</span></code> key, you can specify the following parameters:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">random_seed</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">1337</span>
<span class="nt">output_dir</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">/absolute/path/to/output/dir</span>
<span class="nt">create_subdirectory</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="nt">lr</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.01</span>
<span class="nt">load_test_set</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<p>Note that there should not be any commas at the end of these lines.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">random_seed</span></code>: Set a random seed for <code class="docutils literal notranslate"><span class="pre">torch</span></code> and <code class="docutils literal notranslate"><span class="pre">torch.cuda</span></code> (if using CUDA-enabled GPUs) for reproducibility.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_dir</span></code>: The output directory to use for all operations. If no <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> is specified, Selene assumes that the <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> is specified in all relevant function-type values for operations in Selene. (More information on what function-type values are in later sections, see: <a class="reference internal" href="#a-note-for-the-following-sections"><span class="std std-ref">A note for the following sections</span></a>.) We recommend using this parameter for <code class="docutils literal notranslate"><span class="pre">train</span></code> and <code class="docutils literal notranslate"><span class="pre">evaluate</span></code> operations.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">create_subdirectory</span></code>: If True, creates a directory within <code class="docutils literal notranslate"><span class="pre">output_dir</span></code>   with the name formatted as <code class="docutils literal notranslate"><span class="pre">%Y-%m-%d-%H-%M-%S</span></code>—the date/time when Selene was run. (This is only applicable if <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> has been specified.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lr</span></code>:  The learning rate. If you use the CLI (<code class="docutils literal notranslate"><span class="pre">selene_sdk</span></code>), you can pass this in as a command-line argument rather than having it specified in the configuration file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">load_test_set</span></code>: This is only applicable if you have specified <cite>ops: [train, evaluate]</cite>. You can set this parameter to True (by default it is False and the test set is only loaded when training ends) if you would like to load the test set into memory before training begins—and therefore save the test data generated by a sampler to a .bed file. You would find this useful if you want to save a test dataset and you do not know if your model will finish training and evaluation within the allotted time that your job is run. You should also be running Selene on a machine that can support such an increase in memory usage (on the order of GBs, depending on how many classes your model predicts, how large the test dataset is, etc.).</p></li>
</ul>
</div>
<div class="section" id="model-architecture">
<h2>Model architecture<a class="headerlink" href="#model-architecture" title="Permalink to this headline">¶</a></h2>
<p>For all operations, Selene requires that you specify the model architecture, loss, and optimizer as inputs.</p>
<div class="section" id="expected-input-class-and-methods">
<h3>Expected input class and methods<a class="headerlink" href="#expected-input-class-and-methods" title="Permalink to this headline">¶</a></h3>
<p>There are two possible formats you can use to do this:</p>
<ul class="simple">
<li><p>single Python file: We expect that most people will start using Selene with model architectures in this format. In this case, you implement your architecture as a class and include 2 static methods, <code class="docutils literal notranslate"><span class="pre">criterion</span></code> and <code class="docutils literal notranslate"><span class="pre">get_optimizer</span></code> in the same file. See our <a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/models/deepsea.py">DeepSEA model file</a> as an example.</p>
<ul>
<li><p>The <code class="docutils literal notranslate"><span class="pre">criterion</span></code> method should not take any input arguments and must return a loss function object of type <code class="docutils literal notranslate"><span class="pre">torch.nn._Loss</span></code>.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">get_optimizer</span></code> method should accept a single input <code class="docutils literal notranslate"><span class="pre">lr</span></code>, the learning rate. (Note that this method is not used for the <code class="docutils literal notranslate"><span class="pre">evaluate</span></code> and <code class="docutils literal notranslate"><span class="pre">analyze</span></code> operations in Selene.) It returns a tuple, where <code class="docutils literal notranslate"><span class="pre">tuple[0]</span></code> is the optimizer class <code class="docutils literal notranslate"><span class="pre">torch.optim.Optimizer</span></code> and <code class="docutils literal notranslate"><span class="pre">tuple[1]</span></code> is a dictionary of any optional arguments with which Selene can then instantiate the class. Selene will first instantiate the model and then pass the required <code class="docutils literal notranslate"><span class="pre">model.parameters()</span></code> argument as input to the <code class="docutils literal notranslate"><span class="pre">torch.optim.Optimizer</span></code> class constructor.</p></li>
</ul>
</li>
<li><p>Python module: For more complicated architectures, you may want to write custom PyTorch modules and use them in your final architecture. In this case, it is likely your model architecture imports other custom classes. We ask that you then specify your architecture within a Python module. That is, the directory containing your architecture, loss, and optimizer must have a <code class="docutils literal notranslate"><span class="pre">__init__.py</span></code> that imports the architecture class, <code class="docutils literal notranslate"><span class="pre">criterion</span></code>, and <code class="docutils literal notranslate"><span class="pre">get_optimizer</span></code>.</p></li>
</ul>
</div>
<div class="section" id="model-architecture-configuration">
<h3>Model architecture configuration<a class="headerlink" href="#model-architecture-configuration" title="Permalink to this headline">¶</a></h3>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">model</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    path</span><span class="p">:</span> <span class="nv">/absolute/path/to/file_or_model</span><span class="p p-Indicator">,</span>
<span class="nt">    class</span><span class="p">:</span> <span class="nv">ModelArchitectureClassName</span><span class="p p-Indicator">,</span>
<span class="nt">    class_args</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">        arg1</span><span class="p">:</span> <span class="nv">val1</span><span class="p p-Indicator">,</span>
<span class="nt">        arg2</span><span class="p">:</span> <span class="nv">val2</span><span class="p p-Indicator">,</span>
        <span class="nv">...</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    non_strand_specific</span><span class="p">:</span> <span class="nv">mean</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">path</span></code>: This can be the path to a Python file or a Python module (directory). See the <a class="reference external" href="#expected-input-class-and-methods">previous section</a> for details.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">class</span></code>: The model architecture class name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">class_args</span></code>: The arguments needed to instantiate the class. In the case of <a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/models/deepsea.py">DeepSEA</a>, the <code class="docutils literal notranslate"><span class="pre">class_args</span></code> keys would be <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> and <code class="docutils literal notranslate"><span class="pre">n_genomic_features</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">non_strand_specific</span></code>: Optional, possible values are <code class="docutils literal notranslate"><span class="pre">mean</span></code> or <code class="docutils literal notranslate"><span class="pre">max</span></code> if you want to use this parameter. (Otherwise, do not use this key in your model configuration.) If your model does not need to train on strand-specific input sequences, we have implemented a class that will pass both the forward and reverse sequence to the model and either take the <code class="docutils literal notranslate"><span class="pre">mean</span></code> or the <code class="docutils literal notranslate"><span class="pre">max</span></code> value across the two sets of predictions for a sample.</p></li>
</ul>
</div>
</div>
<div class="section" id="a-note-for-the-following-sections">
<h2>A note for the following sections<a class="headerlink" href="#a-note-for-the-following-sections" title="Permalink to this headline">¶</a></h2>
<p>For training, evaluation, and analysis [of sequences using trained models], Selene requires that specific keys in the YAML file correspond to function-type values. The function-type value is used to construct an object that is a class in <code class="docutils literal notranslate"><span class="pre">selene_sdk</span></code>. Our <a class="reference external" href="https://selene.flatironinstitute.org/">documentation website</a> is an important resource for debugging configuration-related errors when you run Selene via the CLI.</p>
<p>We have covered the most common configurations in this document.</p>
</div>
<div class="section" id="train">
<h2>Train<a class="headerlink" href="#train" title="Permalink to this headline">¶</a></h2>
<p>An example configuration for training:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">train_model</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.TrainModel</span> <span class="p p-Indicator">{</span>
<span class="nt">    batch_size</span><span class="p">:</span> <span class="nv">64</span><span class="p p-Indicator">,</span>
<span class="nt">    max_steps</span><span class="p">:</span> <span class="nv">960000</span><span class="p p-Indicator">,</span>
<span class="nt">    report_stats_every_n_steps</span><span class="p">:</span> <span class="nv">32000</span><span class="p p-Indicator">,</span>
<span class="nt">    save_checkpoint_every_n_steps</span><span class="p">:</span> <span class="nv">1000</span><span class="p p-Indicator">,</span>
<span class="nt">    save_new_checkpoints_after_n_steps</span><span class="p">:</span> <span class="nv">640000</span><span class="p p-Indicator">,</span>
<span class="nt">    n_validation_samples</span><span class="p">:</span> <span class="nv">64000</span><span class="p p-Indicator">,</span>
<span class="nt">    n_test_samples</span><span class="p">:</span> <span class="nv">960000</span><span class="p p-Indicator">,</span>
<span class="nt">    cpu_n_threads</span><span class="p">:</span> <span class="nv">32</span><span class="p p-Indicator">,</span>
<span class="nt">    use_cuda</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    data_parallel</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    logging_verbosity</span><span class="p">:</span> <span class="nv">2</span><span class="p p-Indicator">,</span>
<span class="nt">    metrics</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">        roc_auc</span><span class="p">:</span> <span class="kt">!import</span> <span class="nv">sklearn.metrics.roc_auc_score</span><span class="p p-Indicator">,</span>
<span class="nt">        average_precision</span><span class="p">:</span> <span class="kt">!import</span> <span class="nv">sklearn.metrics.average_precision_score</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    checkpoint_resume</span><span class="p">:</span> <span class="nv">False</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="required-parameters">
<h3>Required parameters<a class="headerlink" href="#required-parameters" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">batch_size</span></code>:  Number of samples in one forward/backward pass (a single step).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">max_steps</span></code>: Total number of steps for which to train the model.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">report_stats_every_n_steps</span></code>: The frequency with which to report summary statistics. You can set this value to be equivalent to a training epoch (<code class="docutils literal notranslate"><span class="pre">n_steps</span> <span class="pre">*</span> <span class="pre">batch_size</span></code>) being the total number of samples seen by the model so far. Selene evaluates the model on the validation dataset every <code class="docutils literal notranslate"><span class="pre">report_stats_every_n_steps</span></code> and, if the model obtains the best performance so far (based on the user-specified loss function), Selene saves the model state to a file called <code class="docutils literal notranslate"><span class="pre">best_model.pth.tar</span></code> in <code class="docutils literal notranslate"><span class="pre">output_dir</span></code>.</p></li>
</ul>
</div>
<div class="section" id="optional-parameters">
<h3>Optional parameters<a class="headerlink" href="#optional-parameters" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">save_checkpoint_every_n_steps</span></code>: Default is 1000. The number of steps before Selene saves a new checkpoint model weights file. If this parameter is set to <code class="docutils literal notranslate"><span class="pre">None</span></code>, we will set it to the same value as <code class="docutils literal notranslate"><span class="pre">report_stats_every_n_steps</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">save_new_checkpoints_after_n_steps</span></code>: Default is None. The number of steps after which Selene will continually save new checkpoint model weights files (<code class="docutils literal notranslate"><span class="pre">checkpoint-&lt;TIMESTAMP&gt;.pth.tar</span></code>) every <code class="docutils literal notranslate"><span class="pre">save_checkpoint_every_n_steps</span></code>. Before this, the file <code class="docutils literal notranslate"><span class="pre">checkpoint.pth.tar</span></code> is overwritten every <code class="docutils literal notranslate"><span class="pre">save_checkpoint_every_n_steps</span></code> to limit the memory requirements.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_validation_samples</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. Specify the number of validation samples in the validation set. If <code class="docutils literal notranslate"><span class="pre">None</span></code></p>
<ul class="simple">
<li><p>and the data sampler you use is of type <code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.OnlineSampler</span></code>, we will by default retrieve 32000 validation samples.</p></li>
<li><p>and you are using a <code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.MultiFileSampler</span></code>, we will use all the validation samples available in the appropriate data file.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_test_samples</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. Specify the number of test samples in the test set. If <code class="docutils literal notranslate"><span class="pre">None</span></code> and</p>
<ul>
<li><p>the sampler you specified has no test partition, you should not specify <code class="docutils literal notranslate"><span class="pre">evaluate</span></code> as one of the operations in the <code class="docutils literal notranslate"><span class="pre">ops</span></code> list. That is, Selene will not automatically evaluate your trained model on a test dataset, because the sampler you are using does not have any test data.</p></li>
<li><p>the sampler you use is of type <code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.OnlineSampler</span></code> (and the test partition exists), we will retrieve 640000 test samples.</p></li>
<li><p>the sampler you use is of type <code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.MultiFileSampler</span></code> (and the test partition exists), we will use all the test samples available in the appropriate data file.</p>
<p>You can review the section on samplers for more information.</p>
</li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">cpu_n_threads</span></code>: Default is 1. The number of OpenMP threads used for parallelizing CPU operations in PyTorch.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_cuda</span></code>: Default is False. Specify whether CUDA-enabled GPUs are available for torch to use during training.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">data_parallel</span></code>: Default is False. Specify whether multiple GPUs are available for torch to use during training.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">logging_verbosity</span></code>: Default is 2. Possible values are <code class="docutils literal notranslate"><span class="pre">{0,</span> <span class="pre">1,</span> <span class="pre">2}</span></code>. Sets the logging verbosity level:</p>
<ul class="simple">
<li><p>0: only warnings are logged</p></li>
<li><p>1: information and warnings are logged</p></li>
<li><p>2: debug messages, information, and warnings are all logged</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">metrics</span></code>: Default is a dictionary with <cite>“roc_auc”</cite> mapped to <code class="docutils literal notranslate"><span class="pre">sklearn.metrics.roc_auc_score</span></code> and <cite>“average_precision”</cite> mapped to <code class="docutils literal notranslate"><span class="pre">sklearn.metrics.average_precision_score</span></code>. <code class="docutils literal notranslate"><span class="pre">metrics</span></code> is a dictionary that maps metric names (<cite>str</cite>) to metric functions. In addition to the loss function you specified with your <a class="reference internal" href="#model-architecture"><span class="std std-ref">Model architecture</span></a>, these are the metrics that you would like to monitor during the training/evaluation process (they all get reported every <code class="docutils literal notranslate"><span class="pre">report_stats_every_n_steps</span></code>). See the <a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/tutorials/regression_mpra_example/regression_mpra_example.ipynb">Regression Models in Selene</a> tutorial for a different input to the <code class="docutils literal notranslate"><span class="pre">metrics</span></code> parameter. You can <code class="docutils literal notranslate"><span class="pre">!import</span></code> metrics from <code class="docutils literal notranslate"><span class="pre">scipy</span></code>, <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code>, <code class="docutils literal notranslate"><span class="pre">statsmodels</span></code>. Each metric function should require, in order, the true values and predicted values as input arguments. For example,
<a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html">sklearn.metrics.average_precision_score</a> takes <code class="docutils literal notranslate"><span class="pre">y_true</span></code> and <code class="docutils literal notranslate"><span class="pre">y_score</span></code> as input.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">checkpoint_resume</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. If not <code class="docutils literal notranslate"><span class="pre">None</span></code>, you should pass in the path to a model weights file generated by <code class="docutils literal notranslate"><span class="pre">torch.save</span></code> (and can now be read by <code class="docutils literal notranslate"><span class="pre">torch.load</span></code>) to resume training.</p></li>
</ul>
</div>
<div class="section" id="additional-notes">
<h3>Additional notes<a class="headerlink" href="#additional-notes" title="Permalink to this headline">¶</a></h3>
<p>Attentive readers might have noticed that in the documentation for the <a class="reference external" href="https://selene.flatironinstitute.org/selene.html#trainmodel">TrainModel class</a> there are more input arguments than are required to instantiate the class through the CLI configuration file. This is because they are assumed to be carried through/retrieved from other configuration keys for consistency. Specifically:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">output_dir</span></code> can be specified as a top-level key in the configuration. You can specify it within each function-type constructor (e.g.  <code class="docutils literal notranslate"><span class="pre">!obj:selene_sdk.TrainModel</span></code>) if you prefer. If <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> exists as a top-level key, Selene does use the top-level <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> and ignores all other <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> keys. <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> is omitted in many of the configurations for this reason.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">model</span></code>, <code class="docutils literal notranslate"><span class="pre">loss_criterion</span></code>, <code class="docutils literal notranslate"><span class="pre">optimizer_class</span></code>, <code class="docutils literal notranslate"><span class="pre">optimizer_kwargs</span></code> are all retrieved from the path in the <a class="reference internal" href="#model-architecture"><span class="std std-ref">Model architecture</span></a> configuration.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">data_sampler</span></code>has its own separate configuration that you will need to specify in the same YAML file. Please see <a class="reference internal" href="#sampler-configurations"><span class="std std-ref">Sampler configurations</span></a> for more information.</p></li>
</ul>
</div>
<div class="section" id="expected-outputs-for-training">
<h3>Expected outputs for training<a class="headerlink" href="#expected-outputs-for-training" title="Permalink to this headline">¶</a></h3>
<p>These outputs will be written to <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> (a top-level parameter, can also  be specified within the function-type constructor, see above).</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">best_model.pth.tar</span></code>: the best performing model so far. IMPORTANT: for all <code class="docutils literal notranslate"><span class="pre">*.pth.tar</span></code> files output by Selene right now, we save additional information beyond the model’s state dictionary so that users may continue training these models through Selene if they wish. If you would like to save only the state dictionary, you can run <code class="docutils literal notranslate"><span class="pre">out</span> <span class="pre">=</span> <span class="pre">torch.load(&lt;*.pth.tar&gt;)</span></code> and then save only the <code class="docutils literal notranslate"><span class="pre">state_dict</span></code> key with <code class="docutils literal notranslate"><span class="pre">torch.save(out[&quot;state_dict&quot;],</span> <span class="pre">&lt;state_dict_only.pth.tar&gt;)</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">checkpoint.pth.tar</span></code>: model saved every <code class="docutils literal notranslate"><span class="pre">save_checkpoint_every_n_steps</span></code> steps</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">selene_sdk.train_model.log</span></code>: a detailed log file containing information about how much time it takes for batches to sampled and propagated through the model, how the model is performing, etc.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">selene_sdk.train_model.train.txt</span></code>: model training loss is printed to this file every <code class="docutils literal notranslate"><span class="pre">report_stats_every_n_steps</span></code>.</p>
<ul>
<li><p>Visualize using <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> (<code class="docutils literal notranslate"><span class="pre">plt.plot</span></code>)</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">selene_sdk.train_model.validation.txt</span></code>: model validation loss and other metrics you have specified (defaults would be ROC AUC and AUPRC) are printed to this file (tab-separated) every <code class="docutils literal notranslate"><span class="pre">report_stats_every_n_steps</span></code>.</p>
<ul>
<li><p>Visualize one of these columns using <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> (<code class="docutils literal notranslate"><span class="pre">plt.plot</span></code>)</p></li>
</ul>
</li>
<li><p>saved sampled datasets (if applicable), e.g. <code class="docutils literal notranslate"><span class="pre">test_data.bed</span></code>: if the <code class="docutils literal notranslate"><span class="pre">save_datasets</span></code> value is not an empty list, Selene periodically saves all the data sampled so far in these .bed files. The columns of these files are <code class="docutils literal notranslate"><span class="pre">[chr,</span> <span class="pre">start,</span> <span class="pre">end,</span> <span class="pre">strand,</span> <span class="pre">semicolon_separated_class_indices]</span></code>. In the future, we will adjust this file to support non-binary labels (i.e. since we are only storing class indices in these output .bed files, we can only label sequences with 1/0, presence/absence, of a given class).</p></li>
</ul>
</div>
</div>
<div class="section" id="evaluate">
<h2>Evaluate<a class="headerlink" href="#evaluate" title="Permalink to this headline">¶</a></h2>
<p>An example configuration for evaluation:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">evaluate_model</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.EvaluateModel</span> <span class="p p-Indicator">{</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/features_list.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    use_features_ord</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/features_subset_ordered.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    trained_model_path</span><span class="p">:</span> <span class="nv">/path/to/trained/model.pth.tar</span><span class="p p-Indicator">,</span>
<span class="nt">    batch_size</span><span class="p">:</span> <span class="nv">64</span><span class="p p-Indicator">,</span>
<span class="nt">    n_test_samples</span><span class="p">:</span> <span class="nv">640000</span><span class="p p-Indicator">,</span>
<span class="nt">    report_gt_feature_n_positives</span><span class="p">:</span> <span class="nv">50</span><span class="p p-Indicator">,</span>
<span class="nt">    use_cuda</span><span class="p">:</span> <span class="nv">True</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="id1">
<h3>Required parameters<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">features</span></code>: The list of distinct features the model predicts. (<code class="docutils literal notranslate"><span class="pre">input_path</span></code> to the function-type value that loads the features as a list.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">trained_model_path</span></code>: Path to the trained model weights file, which should have been generated/saved using <code class="docutils literal notranslate"><span class="pre">torch.save</span></code>. (i.e. you can pass in the saved model file generated by Selene’s <code class="docutils literal notranslate"><span class="pre">TrainModel</span></code> class.)</p></li>
</ul>
</div>
<div class="section" id="id2">
<h3>Optional parameters<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">batch_size</span></code>: Default is 64. Specify the batch size to process examples. Should be a power of 2.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_test_samples</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. Use <code class="docutils literal notranslate"><span class="pre">n_test_samples</span></code> if you want to limit the number of samples on which you evaluate your model. If you are using a sampler of type <code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.OnlineSampler</span></code>—you must specify a test partition in this case—it will default to 640000 test samples if <code class="docutils literal notranslate"><span class="pre">n_test_samples</span> <span class="pre">=</span> <span class="pre">None</span></code>. If you are using a file sampler (<a class="reference internal" href="#multiple-file-sampler"><span class="std std-ref">Multiple-file sampler</span></a>, <a class="reference internal" href="#bed-file-sampler"><span class="std std-ref">BED file sampler</span></a>, or <a class="reference internal" href="#matrix-file-sampler"><span class="std std-ref">Matrix file sampler</span></a>), it will use all samples available in the file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">report_gt_feature_n_positives</span></code>: Default is 10. In total, each class/feature must have more than <code class="docutils literal notranslate"><span class="pre">report_gt_feature_n_positives</span></code> positive examples in the test set to be considered in the performance computation. The output file that reports each class’s performance will report ‘NA’ for classes that do not have enough positive samples.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_cuda</span></code>: Default is False. Specify whether CUDA-enabled GPUs are available for torch to use.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">data_parallel</span></code>: Default is False. Specify whether multiple GPUs are available for torch to use.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_features_ord</span></code>: Default is None. Specify an ordered list of features for which to run the evaluation. The features in this list must be identical to or a subset of <code class="docutils literal notranslate"><span class="pre">features</span></code>, and in the order you want the resulting <code class="docutils literal notranslate"><span class="pre">test_targets.npz</span></code> and <code class="docutils literal notranslate"><span class="pre">test_predictions.npz</span></code> to be saved.</p></li>
</ul>
</div>
<div class="section" id="id3">
<h3>Additional notes<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>Similar to the <code class="docutils literal notranslate"><span class="pre">train_model</span></code> configuration, any arguments that you find in the <a class="reference external" href="https://selene.flatironinstitute.org/selene.html#evaluatemodel">EvaluateModel</a> documentation that are not present in the function-type value’s arguments are automatically instantiated and passed in by Selene.</p>
<p>If you use a sampler with multiple data partitions with the <code class="docutils literal notranslate"><span class="pre">evaluate_model</span></code> configuration, please make sure that your sampler configuration’s <code class="docutils literal notranslate"><span class="pre">mode</span></code> parameter is set to <code class="docutils literal notranslate"><span class="pre">test</span></code>.</p>
</div>
<div class="section" id="expected-outputs-for-evaluation">
<h3>Expected outputs for evaluation<a class="headerlink" href="#expected-outputs-for-evaluation" title="Permalink to this headline">¶</a></h3>
<p>These outputs will be written to <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> (a top-level parameter, can also  be specified within the function-type constructor).</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">test_performance.txt</span></code>: columns are <code class="docutils literal notranslate"><span class="pre">class</span></code> and whatever other metrics you specified (defaults: <code class="docutils literal notranslate"><span class="pre">roc_auc</span></code> and <code class="docutils literal notranslate"><span class="pre">average_precision</span></code>). The breakdown of performance metrics by each class that the model predicts.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">test_predictions.npz</span></code>: The model predictions for each sample in the test set. Useful if you want to make your own visualizations/figures.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">test_targets.npz</span></code>: The actual classes for each sample in the test set. Useful if you want to make your own visualizations/figures.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">precision_recall_curves.svg</span></code>: If using AUPRC as a metric, this is an AUPRC figure that we generate for you. Each curve corresponds to one of the classes the model predicts.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">roc_curves.svg</span></code>: If using ROC AUC as a metric, this is an ROC AUC figure that we generate for you. Each curve corresponds to one of the classes the model predicts.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">selene_sdk.evaluate_model.log</span></code>: Note that if <code class="docutils literal notranslate"><span class="pre">evaluate</span></code> is run through <code class="docutils literal notranslate"><span class="pre">TrainModel</span></code> (that is, no <code class="docutils literal notranslate"><span class="pre">EvaluateModel</span></code> configuration was specified, but you used <code class="docutils literal notranslate"><span class="pre">ops:</span> <span class="pre">[train,</span> <span class="pre">evaluate]</span></code>) you will only see <code class="docutils literal notranslate"><span class="pre">selene_sdk.train_model.log</span></code>. <code class="docutils literal notranslate"><span class="pre">selene_sdk.evaluate_model.log</span></code> is only created when <code class="docutils literal notranslate"><span class="pre">EvaluateModel</span></code> is initialized.</p></li>
</ul>
</div>
</div>
<div class="section" id="analyze-sequences">
<h2>Analyze sequences<a class="headerlink" href="#analyze-sequences" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">analyze</span></code> operation allows you to apply a trained model to new sequences of interest. Currently, we support 3 “sub-operations” for <code class="docutils literal notranslate"><span class="pre">analyze</span></code>:</p>
<ol class="arabic simple">
<li><p>Prediction on sequences: Output the model predictions for a list of sequences.</p></li>
<li><p>Variant effect prediction: Output the model predictions for sequences centered on specific variants (will output reference and alternate predictions as separate files).</p></li>
<li><p><em>In silico</em> mutagenesis: <em>In silico</em> mutagenesis (ISM) involves computationally “mutating” every position in the sequence to every other possible base (DNA and RNA) or amino acid (protein sequences) and examining the consequences of these “mutations”. For ISM, Selene outputs the model predictions for the reference (original) sequence along with each of the mutated sequences.</p></li>
</ol>
<p>For variant effect prediction and <em>in silico</em> mutagenesis, a number of scores can be computed using the predictions from the reference and alternate alleles. You may select 1 or more of the following as outputs:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">predictions</span></code> (output the predictions for each variant, as described above)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">diffs</span></code> (difference scores): The difference between alt and ref predictions.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">abs_diffs</span></code> (absolute difference scores): The absolute difference between alt and ref predictions.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">logits</span></code> (log-fold change scores): The difference between <code class="docutils literal notranslate"><span class="pre">logit(alt)</span></code> and <code class="docutils literal notranslate"><span class="pre">logit(ref)</span></code> predictions.</p></li>
</ul>
<p>You’ll find examples of how these scores are specified in the <a class="reference internal" href="#id6"><span class="std std-ref">Variant effect prediction</span></a> and <span class="xref std std-ref">In silico mutagenesis</span> sections.</p>
<p>In all <code class="docutils literal notranslate"><span class="pre">analyze</span></code>-related operations, we ask that you specify 2 configuration keys. One will always be the <code class="docutils literal notranslate"><span class="pre">analyze_sequences</span></code> key and the other one is dependent on which of the 3 sub-operations you use: <code class="docutils literal notranslate"><span class="pre">prediction</span></code>, <code class="docutils literal notranslate"><span class="pre">variant_effect_prediction</span></code> or <code class="docutils literal notranslate"><span class="pre">in_silico_mutagenesis</span></code>.</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">analyze_sequences</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.predict.AnalyzeSequences</span> <span class="p p-Indicator">{</span>
<span class="nt">    trained_model_path</span><span class="p">:</span> <span class="nv">/path/to/trained/model.pth.tar</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_length</span><span class="p">:</span> <span class="nv">1000</span><span class="p p-Indicator">,</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/features_list.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    batch_size</span><span class="p">:</span> <span class="nv">64</span><span class="p p-Indicator">,</span>
<span class="nt">    use_cuda</span><span class="p">:</span> <span class="nv">False</span><span class="p p-Indicator">,</span>
<span class="nt">    reference_sequence</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.sequences.Genome</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/reference_sequence.fa</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    write_mem_limit</span><span class="p">:</span> <span class="nv">5000</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="id4">
<h3>Required parameters<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">trained_model_path</span></code>: Path to the trained model weights file, which should have been generated/saved using <code class="docutils literal notranslate"><span class="pre">torch.save</span></code>. (i.e. You can pass in the saved model file generated by Selene’s <code class="docutils literal notranslate"><span class="pre">TrainModel</span></code> class.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>: The sequence length the model is expecting for each input.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">features</span></code>: The list of distinct features the model predicts. (<code class="docutils literal notranslate"><span class="pre">input_path</span></code> to the function-type value that loads the features as a list.)</p></li>
</ul>
</div>
<div class="section" id="id5">
<h3>Optional parameters<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">batch_size</span></code>: Default is 64. The size of the mini-batches to use.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_cuda</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>. Specify whether CUDA-enabled GPUs are available for torch to use.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code>: Default is the class <code class="docutils literal notranslate"><span class="pre">selene_sdk.sequences.Genome</span></code>. The type of sequence on which this analysis will be performed (must be type <code class="docutils literal notranslate"><span class="pre">selene.sequences.Sequence</span></code>).</p>
<ul>
<li><p>IMPORTANT: For variant effect prediction and prediction on sequences in a BED file, the reference sequence version should correspond to the version used to specify the chromosome and position of each variant, NOT necessarily the one on which your model was trained.</p></li>
<li><p>For prediction on sequences in a FASTA file and <em>in silico</em> mutagenesis, the only thing that matters is the sequence type—that is, Selene uses the static variables in the class for information about the sequence alphabet and encoding. One problem with our current configuration file parsing is that it asks you to pass in a valid input FASTA file even though you do not need the reference sequence for these 2 sub-operations. We aim to resolve this issue in the future.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">write_mem_limit</span></code>: Default is 5000. Specify, in MB, the amount of memory you want to allocate to storing model predictions/scores. When running one of the sub-operations in <code class="docutils literal notranslate"><span class="pre">analyze</span></code>, prediction/score handlers will accumulate data in memory and write this data to files periodically. By default, Selene will write to files when the <strong>total amount</strong> of data (that is, across all handlers) takes up 5000MB of space. Please keep in mind that Selene will not monitor the amount of memory needed to actually carry out a sub-operation (or load the model beforehand), so <code class="docutils literal notranslate"><span class="pre">write_mem_limit</span></code> must always be less than the total amount of CPU memory you have available on your machine. It is hard to recommend a specific proportion of memory you would allocate for <code class="docutils literal notranslate"><span class="pre">write_mem_limit</span></code> because it is dependent on your input file size (we may change this soon, but Selene currently loads all variants/sequences in a file into memory before running the sub-operation), the model size, and whether the model will run on CPU or GPU.</p></li>
</ul>
</div>
<div class="section" id="prediction-on-sequences">
<h3>Prediction on sequences<a class="headerlink" href="#prediction-on-sequences" title="Permalink to this headline">¶</a></h3>
<p>For prediction on sequences, we require that a user specifies the path to a FASTA file or BED file.</p>
<p>An example configuration for prediction on sequences:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">prediction</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    input_path</span><span class="p">:</span> <span class="nv">/path/to/sequences.bed</span><span class="p p-Indicator">,</span>
<span class="nt">    output_dir</span><span class="p">:</span> <span class="nv">/path/to/output/dir</span><span class="p p-Indicator">,</span>
<span class="nt">    output_format</span><span class="p">:</span> <span class="nv">tsv</span><span class="p p-Indicator">,</span>
<span class="nt">    strand_index</span><span class="p">:</span> <span class="nv">5</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="parameters">
<h4>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">input_path</span></code>: Input path to the FASTA or BED file. For BED file input, we only use the genome regions specified in each row for finding the center position of the input sequence to the model. That is, the start and end of each coordinate does not need to be the same length as the expected model input sequence length–Selene will handle creating the correct sequence input for you.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_dir</span></code>: Output directory to write the model predictions. The resulting file will have the same filename prefix (e.g. <code class="docutils literal notranslate"><span class="pre">example.fasta</span></code> will output <code class="docutils literal notranslate"><span class="pre">example_predictions.tsv</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_format</span></code>: Default is ‘tsv’. You may specify either ‘tsv’ or ‘hdf5’. ‘tsv’ is suitable if you do not have many sequences (&lt;1000) or your model does not predict very many classes (&lt;1000) and you want to be able to view the full set of predictions quickly and easily (via a text editor or Excel). ‘hdf5’ is suitable for downstream analysis. You can access the data in the HDF5 file using the Python package <code class="docutils literal notranslate"><span class="pre">h5py</span></code>. Once the file is loaded, the full matrix is accessible under the key/name <code class="docutils literal notranslate"><span class="pre">&quot;data&quot;</span></code>. Saving to TSV is much slower (more than 2x slower) than saving to HDF5. An additional .txt file with the row labels (descriptions for each sequence in the FASTA) will be output for the HDF5 format as well. It should be ordered in the same way as your input file. The matrix rows will correspond to each sequence and the columns the classes the model predicts.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">strand_index</span></code>: Default is None. If input is BED file, you may specify the column index (0-based) that contains strand information. Otherwise we assume all sequences passed into the model will be fetched from the forward strand. The reference and alternate alleles specified in the VCF should still be for the forward strand–Selene will apply reverse complement to those alleles when strand is ‘-‘.</p></li>
</ul>
</div>
</div>
<div class="section" id="id6">
<h3>Variant effect prediction<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<p>Currently, we expect that all sequences passed as input to a model must be the same length <code class="docutils literal notranslate"><span class="pre">N</span></code>.</p>
<ul class="simple">
<li><p>For SNPs, Selene outputs the model predictions for the ref and alt sequences centered at the <code class="docutils literal notranslate"><span class="pre">(chr,</span> <span class="pre">pos)</span></code> specified.</p></li>
<li><p>For indels, sequences are centered at <code class="docutils literal notranslate"><span class="pre">pos</span> <span class="pre">+</span> <span class="pre">(N_bases</span> <span class="pre">/</span> <span class="pre">2)</span></code>, for the reference sequence of length <code class="docutils literal notranslate"><span class="pre">N_bases</span></code>.</p></li>
</ul>
<p>An example configuration for variant effect prediction:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">variant_effect_prediction</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    vcf_files</span><span class="p">:</span> <span class="p p-Indicator">[</span>
        <span class="nv">/path/to/file1.vcf</span><span class="p p-Indicator">,</span>
        <span class="nv">/path/to/file2.vcf</span><span class="p p-Indicator">,</span>
        <span class="nv">...</span>
    <span class="p p-Indicator">],</span>
<span class="nt">    save_data</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">abs_diffs</span><span class="p p-Indicator">],</span>
<span class="nt">    output_dir</span><span class="p">:</span> <span class="nv">/path/to/output/predictions/dir</span><span class="p p-Indicator">,</span>
<span class="nt">    output_format</span><span class="p">:</span> <span class="nv">hdf5</span><span class="p p-Indicator">,</span>
<span class="nt">    strand_index</span><span class="p">:</span> <span class="nv">7</span><span class="p p-Indicator">,</span>
<span class="nt">    require_strand</span><span class="p">:</span> <span class="nv">True</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="id7">
<h4>Parameters<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">vcf_files</span></code>: Path to a VCF file. Must contain the columns <code class="docutils literal notranslate"><span class="pre">[#CHROM,</span> <span class="pre">POS,</span> <span class="pre">ID,</span> <span class="pre">REF,</span> <span class="pre">ALT]</span></code>, in order. Column header does not need to be present. (All other columns in the file will be ignored.)</p></li>
<li><p><cite>save_data</cite>: A list of the data files to output. Must input 1 or more of the following options: <code class="docutils literal notranslate"><span class="pre">[abs_diffs,</span> <span class="pre">diffs,</span> <span class="pre">logits,</span> <span class="pre">predictions]</span></code>. (Note that the raw prediction values will not be outputted by default—you must specify <code class="docutils literal notranslate"><span class="pre">predictions</span></code> in the list if you want them.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_dir</span></code>: Output directory to write the model predictions. The resulting file will have the same filename prefix.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_format</span></code>: Default is ‘tsv’. You may specify either ‘tsv’ or ‘hdf5’. ‘tsv’ is suitable if you do not have many variants (on the order of 10^4 or less) or your model does not predict very many classes (&lt;1000) and you want to be able to view the full set of predictions quickly and easily (via a text editor or Excel). ‘hdf5’ is suitable for downstream analysis. You can access the data in the HDF5 file using the Python package <code class="docutils literal notranslate"><span class="pre">h5py</span></code>. Once the file is loaded, the full matrix is accessible under the key/name <code class="docutils literal notranslate"><span class="pre">&quot;data&quot;</span></code>. Saving to TSV is much slower (more than 2x slower) than saving to HDF5. When the output is in HDF5 format, an additional .txt file of row labels (corresponding to the columns (chrom, pos, id, ref, alt)) will be output so that you can match up the data matrix rows with the particular variant. Columns of the matrix correspond to the classes the model predicts.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">strand_index</span></code>: Default is None. If applicable, specify the column index (0-based) in the VCF file that contains strand information for each variant. Note that currently Selene assumes that, for multiple input VCF files, the strand column is the same for all the files. Importantly, the VCF file ref and alt alleles should still be specified for the forward strand–Selene will take the reverse complement for both if strand = ‘-‘.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">require_strand</span></code>: Default is False. If <code class="docutils literal notranslate"><span class="pre">strand_index</span></code> is not None, <code class="docutils literal notranslate"><span class="pre">require_strand</span> <span class="pre">=</span> <span class="pre">True</span></code> means that Selene will skip all variants with strand specified as ‘.’ (that is, only keep variants with strand column value being ‘+’ or ‘-‘). If <code class="docutils literal notranslate"><span class="pre">require_strand</span> <span class="pre">=</span> <span class="pre">False</span></code>, variants with strand specified as ‘.’ will be treated as being on the ‘+’ strand.</p></li>
</ul>
</div>
<div class="section" id="additional-note">
<h4>Additional note<a class="headerlink" href="#additional-note" title="Permalink to this headline">¶</a></h4>
<p>You may find that there are more output files than you expect in <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> at the end of variant effect prediction. The following cases may occur:</p>
<ul class="simple">
<li><p><strong>NAs:</strong> for some variants, Selene may not be able to construct a reference sequence centered at <code class="docutils literal notranslate"><span class="pre">pos</span></code> of the specified sequence length. This is likely because <code class="docutils literal notranslate"><span class="pre">pos</span></code> is near the end or the beginning of the chromosome and the sequence length the model accepts as input is large. You will find a list of NA variants in a file that ends with the extension <code class="docutils literal notranslate"><span class="pre">.NA</span></code>.</p></li>
<li><p><strong>Warnings:</strong> Selene may detect that the <code class="docutils literal notranslate"><span class="pre">ref</span></code> base(s) in a variant do not match with the bases specified in the reference sequence FASTA at the <code class="docutils literal notranslate"><span class="pre">(chrom,</span> <span class="pre">pos)</span></code>. In this case, Selene will use the <code class="docutils literal notranslate"><span class="pre">ref</span></code> base(s) specified in the VCF file in place of those in the reference genome and output predictions accordingly. These predictions will be distinguished by the row label column <code class="docutils literal notranslate"><span class="pre">ref_match</span></code> value <code class="docutils literal notranslate"><span class="pre">False</span></code>. You may review these variants and determine whether you still want to use those predictions/scores. If you find that most of the variants have <code class="docutils literal notranslate"><span class="pre">ref_match</span> <span class="pre">=</span> <span class="pre">False</span></code>, it may be that you have specified the wrong reference genome version—please check this before proceeding.</p></li>
</ul>
</div>
</div>
<div class="section" id="id8">
<h3><em>In silico</em> mutagenesis<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<p>An example configuration for <em>in silico</em> mutagenesis of the whole sequence (i.e. rather than a subsequence), when using a single sequence as input:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">in_silico_mutagenesis</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    input_sequence</span><span class="p">:</span> <span class="nv">ATCGATAAAATTCTGGAG...</span><span class="p p-Indicator">,</span>
<span class="nt">    save_data</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">predictions</span><span class="p p-Indicator">,</span> <span class="nv">diffs</span><span class="p p-Indicator">],</span>
<span class="nt">    output_path_prefix</span><span class="p">:</span> <span class="nv">/path/to/output/dir/filename_prefix</span><span class="p p-Indicator">,</span>
<span class="nt">    mutate_n_bases</span><span class="p">:</span> <span class="nv">1</span><span class="p p-Indicator">,</span>
<span class="nt">    start_position</span><span class="p">:</span> <span class="nv">0</span><span class="p p-Indicator">,</span>
<span class="nt">    end_position</span><span class="p">:</span> <span class="nv">None</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="parameters-for-a-single-sequence-input">
<h4>Parameters for a single sequence input<a class="headerlink" href="#parameters-for-a-single-sequence-input" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sequence</span></code>: A sequence you are interested in. If the sequence length is less than or greater than the expected model’s input sequence length, Selene truncates or pads (with unknown base, e.g. <code class="docutils literal notranslate"><span class="pre">N</span></code>) the sequence for you.</p></li>
<li><p><cite>save_data</cite>: A list of the data files to output. Must input 1 or more of the following options: <code class="docutils literal notranslate"><span class="pre">[abs_diffs,</span> <span class="pre">diffs,</span> <span class="pre">logits,</span> <span class="pre">predictions]</span></code>. (Note that the raw prediction values will not be outputted by default—you must specify <code class="docutils literal notranslate"><span class="pre">predictions</span></code> in the list if you want them.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_path_prefix</span></code>: Optional, default is “ism”. The path to which the data files are written. We have specified that it should be a filename <em>prefix</em> because we will append additional information depending on what files you would like to output (e.g. <code class="docutils literal notranslate"><span class="pre">fileprefix_logits.tsv</span></code>) If directories in the path do not yet exist, they will automatically be created.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutate_n_bases</span></code>: Optional, default is 1. The number of bases to mutate at any time. Standard <em>in silico</em> mutagenesis only mutates a single base at a time, so we encourage users to start by leaving this value at 1. Double/triple mutations will be more difficult to interpret and are something we may work on in the future.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">start_position</span></code>: Optional, default is 0. The starting position of the subsequence that should be mutated. This value should be nonnegative, and less than <code class="docutils literal notranslate"><span class="pre">end_position</span></code>. Also, the value of <code class="docutils literal notranslate"><span class="pre">end_position</span> <span class="pre">-</span> <span class="pre">start_position</span></code> should be at least <code class="docutils literal notranslate"><span class="pre">mutate_n_bases</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">end_position</span></code>: Optional, default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. If left as <code class="docutils literal notranslate"><span class="pre">None</span></code>, Selene will use the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> parameter from <code class="docutils literal notranslate"><span class="pre">analyze_sequences</span></code>. This is the ending position of the subsequence that should be mutated. This value should be nonnegative, and greater than <code class="docutils literal notranslate"><span class="pre">start_position</span></code>. The value of <code class="docutils literal notranslate"><span class="pre">end_position</span> <span class="pre">-</span>&#160; <span class="pre">start_position</span></code> should be at least <code class="docutils literal notranslate"><span class="pre">mutate_n_bases</span></code>.</p></li>
</ul>
<p>An example configuration for <em>in silico</em> mutagenesis of the center 100 bases of a 1000 base sequence read from a FASTA file input:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">in_silico_mutagenesis</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    input_path</span><span class="p">:</span> <span class="nv">/path/to/sequences1.fa</span><span class="p p-Indicator">,</span>
<span class="nt">    save_data</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">logits</span><span class="p p-Indicator">],</span>
<span class="nt">    output_dir</span><span class="p">:</span> <span class="nv">/path/to/output/predictions/dir</span><span class="p p-Indicator">,</span>
<span class="nt">    mutate_n_bases</span><span class="p">:</span> <span class="nv">1</span><span class="p p-Indicator">,</span>
<span class="nt">    use_sequence_name</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    start_position</span><span class="p">:</span> <span class="nv">450</span><span class="p p-Indicator">,</span>
<span class="nt">    end_position</span><span class="p">:</span> <span class="nv">550</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
</div>
<div class="section" id="parameters-for-fasta-file-input">
<h4>Parameters for FASTA file input:<a class="headerlink" href="#parameters-for-fasta-file-input" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">input_path</span></code>: Input path to the FASTA file. If you have multiple FASTA files, you can replace this key with <code class="docutils literal notranslate"><span class="pre">fa_files</span></code> and submit an input list, the same way it is done in variant effect prediction.</p></li>
<li><p><cite>save_data</cite>: A list of the data files to output. Must input 1 or more of the following options: <code class="docutils literal notranslate"><span class="pre">[abs_diffs,</span> <span class="pre">diffs,</span> <span class="pre">logits,</span> <span class="pre">predictions]</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_dir</span></code>: Output directory to write the model predictions.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutate_n_bases</span></code>: Optional, default is 1. The number of bases to mutate at any time. Standard <em>in silico</em> mutagenesis only mutates a single base at a time, so we encourage users to start by leaving this value at 1.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_sequence_name</span></code>: Optional, default is <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<ul>
<li><p>If <code class="docutils literal notranslate"><span class="pre">use_sequence_name</span></code>, output files are prefixed by the sequence name/description corresponding to each sequence in the FASTA file. Spaces in the description are replaced with underscores ‘_’.</p></li>
<li><p>If not <code class="docutils literal notranslate"><span class="pre">use_sequence_name</span></code>, output files are prefixed with the index <code class="docutils literal notranslate"><span class="pre">i</span></code> corresponding to the <code class="docutils literal notranslate"><span class="pre">i</span></code>th sequence in the FASTA file.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">start_position</span></code>: Optional, default is 0. The starting position of the subsequence that should be mutated. This value should be nonnegative, and less than <code class="docutils literal notranslate"><span class="pre">end_position</span></code>. The value of <code class="docutils literal notranslate"><span class="pre">end_position</span> <span class="pre">-</span> <span class="pre">start_position</span></code> should be at least <code class="docutils literal notranslate"><span class="pre">mutate_n_bases</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">end_position</span></code>: Optional, default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. If left as <code class="docutils literal notranslate"><span class="pre">None</span></code>, Selene will use the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> parameter passed to <code class="docutils literal notranslate"><span class="pre">analyze_sequences</span></code>. This is the ending position of the subsequence that should be mutated. This value should be nonnegative, and greater than <code class="docutils literal notranslate"><span class="pre">start_position</span></code>. The value of <code class="docutils literal notranslate"><span class="pre">end_position</span> <span class="pre">-</span>&#160; <span class="pre">start_position</span></code> should be at least <code class="docutils literal notranslate"><span class="pre">mutate_n_bases</span></code>.</p></li>
</ul>
</div>
</div>
</div>
<div class="section" id="sampler-configurations">
<h2>Sampler configurations<a class="headerlink" href="#sampler-configurations" title="Permalink to this headline">¶</a></h2>
<p>Data sampling is used during model training and evaluation. You must specify the sampler in the configuration YAML file alongside the other operation-specific configurations (i.e. <code class="docutils literal notranslate"><span class="pre">train_model</span></code> or <code class="docutils literal notranslate"><span class="pre">evaluate_model</span></code>).</p>
<div class="section" id="samplers-used-for-training-and-evaluation-optionally">
<h3>Samplers used for training (and evaluation, optionally)<a class="headerlink" href="#samplers-used-for-training-and-evaluation-optionally" title="Permalink to this headline">¶</a></h3>
<p>Training requires a sampler that specifies the data for training, validation, and (optionally) testing. While Selene can directly evaluate a trained model on a test dataset when training is finished, it is not a required step and so the test dataset specification is also optional. Here, we provide examples for the samplers we have implemented that can be used for training.</p>
<p>There are 2 kinds of samplers implemented in Selene right now: “online” samplers and file samplers. Online samplers generate data samples on-the-fly and require you to pass in a reference sequence FASTA file and a tabix-indexed BED file so that Selene can query for an input sequence and its associated biological classes using genomic coordinates. The file sampler we use supports loading different <code class="docutils literal notranslate"><span class="pre">.mat</span></code> or <code class="docutils literal notranslate"><span class="pre">.bed</span></code> files (can support more formats upon request) for the training, validation, and test sets.</p>
<p>For increased efficiency during the training of large models, we would recommend using the online sampler to create datasets (.bed or .mat) and then loading the generated data with a file sampler. We are actively working to incorporate PyTorch dataloaders and other improvements to data sampling into Selene to reduce the time and memory requirements of training. Feel free to contact us through our <a class="reference external" href="https://github.com/FunctionLab/selene/issues">Github issues</a> if you have comments or want to contribute to this effort!</p>
<div class="section" id="random-positions-sampler">
<h4>Random positions sampler<a class="headerlink" href="#random-positions-sampler" title="Permalink to this headline">¶</a></h4>
<p>The random positions sampler will construct data samples by randomly selecting a position in the genome and then using the sequence and classes centered at that position as the input and targets for the model to predict.</p>
<p>An example configuration for the random positions sampler:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.RandomPositionsSampler</span> <span class="p p-Indicator">{</span>
<span class="nt">    reference_sequence</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.sequences.Genome</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/reference_sequence.fa</span><span class="p p-Indicator">,</span>
<span class="nt">        blacklist_regions</span><span class="p">:</span> <span class="nv">hg19</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/features_list.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    target_path</span><span class="p">:</span> <span class="nv">/path/to/targets_bed.gz</span><span class="p p-Indicator">,</span>
<span class="nt">    seed</span><span class="p">:</span> <span class="nv">123</span><span class="p p-Indicator">,</span>
<span class="nt">    validation_holdout</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">chr6</span><span class="p p-Indicator">,</span> <span class="nv">chr7</span><span class="p p-Indicator">],</span>
<span class="nt">    test_holdout</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">chr8</span><span class="p p-Indicator">,</span> <span class="nv">chr9</span><span class="p p-Indicator">],</span>
<span class="nt">    sequence_length</span><span class="p">:</span> <span class="nv">1000</span><span class="p p-Indicator">,</span>
<span class="nt">    center_bin_to_predict</span><span class="p">:</span> <span class="nv">200</span><span class="p p-Indicator">,</span>
<span class="nt">    feature_thresholds</span><span class="p">:</span> <span class="nv">0.5</span><span class="p p-Indicator">,</span>
<span class="nt">    mode</span><span class="p">:</span> <span class="nv">train</span><span class="p p-Indicator">,</span>
<span class="nt">    save_datasets</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">train</span><span class="p p-Indicator">,</span> <span class="nv">validate</span><span class="p p-Indicator">,</span> <span class="nv">test</span><span class="p p-Indicator">]</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="id9">
<h5>Required parameters<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code>: Path to a reference sequence FASTA file we can query to create our data samples.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">blacklist_regions</span></code> is an optional argument for <code class="docutils literal notranslate"><span class="pre">selene_sdk.sequences.Genome</span></code> that allows you to specify the blacklist regions for the hg19 or hg38 reference sequence. The lists of blacklisted intervals are provided by <a class="reference external" href="https://sites.google.com/site/anshulkundaje/projects/blacklists">Anshul Kundaje for ENCODE</a> and support for more organisms can be included upon request.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">target_path</span></code>: Path to a tabix-indexed, compressed BED file (<code class="docutils literal notranslate"><span class="pre">.bed.gz</span></code>) of genomic coordinates corresponding to the measurements for genomic features/classes the model should predict.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">features</span></code>: The list of distinct features the model predicts. (<code class="docutils literal notranslate"><span class="pre">input_path</span></code> to the function-type value that loads the file of features as a list.)</p></li>
</ul>
</div>
<div class="section" id="id10">
<h5>Optional parameters<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">seed</span></code>: Default is 436.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">validation_holdout</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">[chr6,</span> <span class="pre">chr7]</span></code>. Holdout can be regional (i.e. chromosomal) or proportional.</p>
<ul>
<li><p>If regional, expects a list where the regions must match those specified in the first column of the tabix-indexed BED file <code class="docutils literal notranslate"><span class="pre">target_path</span></code> (which must also match the FASTA descriptions for every record in <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code>).</p></li>
<li><p>If proportional, specify a percentage between (0.0, 1.0). Typically 0.10 or 0.20.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">test_holdout</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">[chr8,</span> <span class="pre">chr9]</span></code>. Holdout can be regional (i.e. chromosomal) or proportional. See description of <code class="docutils literal notranslate"><span class="pre">validation_holdout</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>: Default is 1000. Model is trained on sequences of <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">center_bin_to_predict</span></code>: Default is 200. Query the tabix-indexed file for a region of length <code class="docutils literal notranslate"><span class="pre">center_bin_to_predict</span></code>, centered in the input sequence of <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">feature_thresholds</span></code>: Default is 0.5. The threshold to pass to the <code class="docutils literal notranslate"><span class="pre">selene_sdk.targets.Targets</span></code> object. Because we have only implemented support for genomic features right now, we reproduce the threshold inputs for that here:</p>
<ul>
<li><p>A genomic region is determined to be a positive sample if at least one genomic feature interval takes up some proportion of the region greater than or equal to the corresponding threshold.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">float</span></code>: A single threshold applied to all the features in your dataset.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dict</span></code>: A dictionary mapping feature names (<code class="docutils literal notranslate"><span class="pre">str</span></code>) to thresholds (<code class="docutils literal notranslate"><span class="pre">float</span></code>). This is used if you want to assign different thresholds for different features. If a feature’s threshold is not specified in the dictionary, you must have the key <code class="docutils literal notranslate"><span class="pre">default</span></code> with a default threshold value we can use for that feature.</p></li>
</ul>
</li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">mode</span></code>: Default is ‘train’. Must be one of <code class="docutils literal notranslate"><span class="pre">{train,</span> <span class="pre">validate,</span> <span class="pre">test}</span></code>. The starting mode in which to run this sampler.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">save_datasets</span></code>: Default is <code class="docutils literal notranslate"><span class="pre">[test]</span></code>. The list of modes for which we should save the sampled data to file. Should be one or more of <code class="docutils literal notranslate"><span class="pre">{train,</span> <span class="pre">validate,</span> <span class="pre">test}</span></code>.</p></li>
</ul>
</div>
</div>
<div class="section" id="intervals-sampler">
<h4>Intervals sampler<a class="headerlink" href="#intervals-sampler" title="Permalink to this headline">¶</a></h4>
<p>The intervals sampler will construct data samples by randomly selecting positions only in the regions specified by an intervals <code class="docutils literal notranslate"><span class="pre">.bed</span></code> file and then using the sequence and classes centered at that position as the input and targets for the model to predict.</p>
<p>An example configuration for the intervals sampler:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.IntervalsSampler</span> <span class="p p-Indicator">{</span>
<span class="nt">    reference_sequence</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.sequences.Genome</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/reference_sequence.fa</span><span class="p p-Indicator">,</span>
<span class="nt">        blacklist_regions</span><span class="p">:</span> <span class="nv">hg38</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    target_path</span><span class="p">:</span> <span class="nv">/path/to/targets.bed.gz</span><span class="p p-Indicator">,</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/features_list.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    intervals_path</span><span class="p">:</span> <span class="nv">/path/to/intervals.bed</span><span class="p p-Indicator">,</span>
<span class="nt">    sample_negative</span><span class="p">:</span> <span class="nv">False</span><span class="p p-Indicator">,</span>
<span class="nt">    seed</span><span class="p">:</span> <span class="nv">436</span><span class="p p-Indicator">,</span>
<span class="nt">    validation_holdout</span><span class="p">:</span> <span class="nv">0.10</span><span class="p p-Indicator">,</span>
<span class="nt">    test_holdout</span><span class="p">:</span> <span class="nv">0.10</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_length</span><span class="p">:</span> <span class="nv">1000</span><span class="p p-Indicator">,</span>
<span class="nt">    center_bin_to_predict</span><span class="p">:</span> <span class="nv">100</span><span class="p p-Indicator">,</span>
<span class="nt">    feature_thresholds</span><span class="p">:</span> <span class="p p-Indicator">{</span><span class="s">&quot;feature1&quot;</span><span class="p p-Indicator">:</span> <span class="nv">0.5</span><span class="p p-Indicator">,</span> <span class="s">&quot;default&quot;</span><span class="p p-Indicator">:</span> <span class="nv">0.1</span><span class="p p-Indicator">},</span>
<span class="nt">    mode</span><span class="p">:</span> <span class="nv">test</span><span class="p p-Indicator">,</span>
<span class="nt">    save_datasets</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">test</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
<div class="section" id="id11">
<h5>Parameters<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h5>
<p>With the exception of <code class="docutils literal notranslate"><span class="pre">intervals_path</span></code> and <code class="docutils literal notranslate"><span class="pre">sample_negative</span></code>, all other parameters match those for the random positions sampler. Please see the previous section for more details on the other parameters.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">intervals_path</span></code>: The path to the intervals file. Must have the columns <code class="docutils literal notranslate"><span class="pre">[chr,</span> <span class="pre">start,</span> <span class="pre">end]</span></code>, where values in <code class="docutils literal notranslate"><span class="pre">chr</span></code> should match the descriptions in the FASTA file. We constrain the regions from which we sample to the regions in this file instead of the using the whole genome.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sample_negative</span></code>: Optional, default is False. Specify whether negative examples (i.e. samples with no positive labels) should be drawn. When False, the sampler will check if the <code class="docutils literal notranslate"><span class="pre">center_bin_to_predict</span></code> in the input sequence contains at least 1 of the features/classes the model wants to predict. When True, no such check is made.</p></li>
</ul>
</div>
</div>
<div class="section" id="multiple-file-sampler">
<h4>Multiple-file sampler<a class="headerlink" href="#multiple-file-sampler" title="Permalink to this headline">¶</a></h4>
<p>The multi-file sampler loads in the training, validation, and optionally, the testing dataset.  The configuration for this therefore asks that you fill in some keys with the function-type constructors of type <code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.file_samplers.FileSampler</span></code>. Please consult the following sections for information about these file samplers.</p>
<p>An example configuration for the multiple-file sampler:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.MultiFileSampler</span> <span class="p p-Indicator">{</span>
<span class="nt">    train_sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.file_samplers.MatFileSampler</span> <span class="p p-Indicator">{</span>
        <span class="nv">...</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    validate_sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.file_samplers.MatFileSampler</span> <span class="p p-Indicator">{</span>
        <span class="nv">...</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/features_list.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    test_sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.file_samplers.BedFileSampler</span> <span class="p p-Indicator">{</span>
        <span class="nv">...</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    mode</span><span class="p">:</span> <span class="nv">train</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="id12">
<h5>Parameters<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">train_sampler</span></code>: Load your training data from either a <code class="docutils literal notranslate"><span class="pre">.bed</span></code> file (<code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.file_sampler.BedFileSampler</span></code>) or <code class="docutils literal notranslate"><span class="pre">.mat</span></code> file (<code class="docutils literal notranslate"><span class="pre">selene_sdk.samplers.file_sampler.MatFileSampler</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">validate_sampler</span></code>: Sample as <code class="docutils literal notranslate"><span class="pre">train_sampler</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">test_sampler</span></code>: Optional, default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. Same as <code class="docutils literal notranslate"><span class="pre">train_sampler</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">features</span></code>: The list of distinct features the model predicts. (<code class="docutils literal notranslate"><span class="pre">input_path</span></code> to the function-type value that loads the file of features as a list.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mode</span></code>: Default is ‘train’. Must be one of <code class="docutils literal notranslate"><span class="pre">{train,</span> <span class="pre">validate,</span> <span class="pre">test}</span></code>. The starting mode in which to run this sampler.</p></li>
</ul>
</div>
</div>
</div>
<div class="section" id="important-note">
<h3>Important note<a class="headerlink" href="#important-note" title="Permalink to this headline">¶</a></h3>
<p>If you use any of these samplers (that is, samplers with multiple data partitions) with the <a class="reference internal" href="#evaluate"><span class="std std-ref">Evaluate</span></a> configuration, please make sure that your <code class="docutils literal notranslate"><span class="pre">mode</span></code> is set to <code class="docutils literal notranslate"><span class="pre">test</span></code>.</p>
</div>
<div class="section" id="samplers-used-for-evaluation">
<h3>Samplers used for evaluation<a class="headerlink" href="#samplers-used-for-evaluation" title="Permalink to this headline">¶</a></h3>
<p>You can use all the samplers specified for training for evaluation as well (see note above). Additionally, you can use single-file samplers, which we describe below.</p>
<div class="section" id="bed-file-sampler">
<h4>BED file sampler<a class="headerlink" href="#bed-file-sampler" title="Permalink to this headline">¶</a></h4>
<p>The BED file sampler loads a dataset from a <code class="docutils literal notranslate"><span class="pre">.bed</span></code> file. This can be generated by one of the online samplers in Selene with the <code class="docutils literal notranslate"><span class="pre">save_dataset</span></code> parameter.</p>
<p>An example configuration for a BED file sampler:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.file_samplers.BedFileSampler</span> <span class="p p-Indicator">{</span>
<span class="nt">    filepath</span><span class="p">:</span> <span class="nv">/path/to/data.bed</span><span class="p p-Indicator">,</span>
<span class="nt">    reference_sequence</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.sequences.Genome</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/reference_sequence.fa</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    n_samples</span><span class="p">:</span> <span class="nv">640000</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_length</span><span class="p">:</span> <span class="nv">1000</span><span class="p p-Indicator">,</span>
<span class="nt">    targets_avail</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    n_features</span><span class="p">:</span> <span class="nv">919</span><span class="p p-Indicator">,</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="id13">
<h5>Parameters<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">filepath</span></code>: Path to the BED file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code>: Path to a reference sequence FASTA file we can query to create our data samples.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_samples</span></code>: Number of lines in the file. (<code class="docutils literal notranslate"><span class="pre">wc</span> <span class="pre">-l</span> <span class="pre">&lt;filepath&gt;</span></code>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>: Optional, default is None. If the coordinates of each sample in the BED file, already account for the full sequence (that is, the columns <code class="docutils literal notranslate"><span class="pre">end</span> <span class="pre">-</span> <span class="pre">start</span> <span class="pre">=</span> <span class="pre">sequence_length</span></code>, there is no need to specify this parameter. If <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> is not None, the length of each sample will be checked to determine whether the sample coordinates need to be adjusted to match the sequence length expected by the model architecture.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">targets_avail</span></code>: Optional, default is False. If <code class="docutils literal notranslate"><span class="pre">targets_avail</span></code>, assumes that it is the last column of the <code class="docutils literal notranslate"><span class="pre">.bed</span></code> file. The last column should contain the indices, separated by semicolons, of features (classes) found within a given sample’s coordinates (e.g. 0;1;45;60). This format assumes that we are only looking for the absence/presence of each feature within the interval.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_features</span></code>: Optional, default is None. If <code class="docutils literal notranslate"><span class="pre">targets_avail</span></code> is True, must specify <code class="docutils literal notranslate"><span class="pre">n_features</span></code>, the total number of features (classes).</p></li>
</ul>
</div>
</div>
<div class="section" id="matrix-file-sampler">
<h4>Matrix file sampler<a class="headerlink" href="#matrix-file-sampler" title="Permalink to this headline">¶</a></h4>
<p>The matrix file sampler loads a dataset from a matrix file.</p>
<p>An example configuration for a matrix file sampler:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.file_samplers.MatFileSampler</span> <span class="p p-Indicator">{</span>
<span class="nt">    filepath</span><span class="p">:</span> <span class="nv">/path/to/data.mat</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_key</span><span class="p">:</span> <span class="nv">sequences</span><span class="p p-Indicator">,</span>
<span class="nt">    targets_key</span><span class="p">:</span> <span class="nv">targets</span><span class="p p-Indicator">,</span>
<span class="nt">    random_seed</span><span class="p">:</span> <span class="nv">123</span><span class="p p-Indicator">,</span>
<span class="nt">    shuffle</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_batch_axis</span><span class="p">:</span> <span class="nv">0</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_alphabet_axis</span><span class="p">:</span> <span class="nv">1</span><span class="p p-Indicator">,</span>
<span class="nt">    targets_batch_axis</span><span class="p">:</span> <span class="nv">0</span>
<span class="p p-Indicator">}</span>
</pre></div>
</div>
<div class="section" id="id14">
<h5>Parameters<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">filepath</span></code>: The path to the file from which to load the data.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_key</span></code>: The key for the sequences data matrix.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">targets_key</span></code>: Optional, default is <code class="docutils literal notranslate"><span class="pre">None</span></code>. The key to the targets data matrix.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">random_seed</span></code>: Optional, default is 436. Sets the random seed for sampling.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">shuffle</span></code>: Optional, default is <code class="docutils literal notranslate"><span class="pre">True</span></code>. Shuffle the order of the samples in the matrix before sampling from it.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_batch_axis</span></code>: Optional, default is 0. Specify the batch axis for the sequences matrix.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_alphabet_axis</span></code>: Optional, default is 1. Specify the alphabet axis.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">targets_batch_axis</span></code>: Optional, default is 0. Specify the batch axis for the targets matrix.</p></li>
</ul>
</div>
</div>
</div>
</div>
<div class="section" id="examples-of-full-configuration-files">
<h2>Examples of full configuration files<a class="headerlink" href="#examples-of-full-configuration-files" title="Permalink to this headline">¶</a></h2>
<p>We do have a more comprehensive set of <a class="reference external" href="https://github.com/FunctionLab/selene/blob/master/config_examples">examples on our Github</a> that you can review. We reproduce a few of these in this document to show how you can put all of the different configuration components together to create a YAML file that can be run by Selene’s CLI:</p>
<div class="section" id="training-using-intervals-sampler">
<h3>Training (using intervals sampler)<a class="headerlink" href="#training-using-intervals-sampler" title="Permalink to this headline">¶</a></h3>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nn">---</span>
<span class="nt">ops</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">train</span><span class="p p-Indicator">,</span> <span class="nv">evaluate</span><span class="p p-Indicator">]</span>
<span class="nt">model</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    path</span><span class="p">:</span> <span class="nv">/absolute/path/to/model/architecture.py</span><span class="p p-Indicator">,</span>
<span class="nt">    class</span><span class="p">:</span> <span class="nv">ModelArchitectureClassName</span><span class="p p-Indicator">,</span>
<span class="nt">    class_args</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">        arg1</span><span class="p">:</span> <span class="nv">val1</span><span class="p p-Indicator">,</span>
<span class="nt">        arg2</span><span class="p">:</span> <span class="nv">val2</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    non_strand_specific</span><span class="p">:</span> <span class="nv">mean</span>
<span class="p p-Indicator">}</span>
<span class="nt">sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.IntervalsSampler</span> <span class="p p-Indicator">{</span>
<span class="nt">    reference_sequence</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.sequences.Genome</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/reference_sequence.fa</span><span class="p p-Indicator">,</span>
<span class="nt">        blacklist_regions</span><span class="p">:</span> <span class="nv">hg19</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    target_path</span><span class="p">:</span> <span class="nv">/path/to/tabix/indexed/targets.bed.gz</span><span class="p p-Indicator">,</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/distinct_features.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    intervals_path</span><span class="p">:</span> <span class="nv">/path/to/intervals.bed</span><span class="p p-Indicator">,</span>
<span class="nt">    sample_negative</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    seed</span><span class="p">:</span> <span class="nv">127</span><span class="p p-Indicator">,</span>
<span class="nt">    validation_holdout</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">chr6</span><span class="p p-Indicator">,</span> <span class="nv">chr7</span><span class="p p-Indicator">],</span>
<span class="nt">    test_holdout</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">chr8</span><span class="p p-Indicator">,</span> <span class="nv">chr9</span><span class="p p-Indicator">],</span>  <span class="c1"># specifying a test partition</span>
<span class="nt">    sequence_length</span><span class="p">:</span> <span class="nv">1000</span><span class="p p-Indicator">,</span>
<span class="nt">    center_bin_to_predict</span><span class="p">:</span> <span class="nv">200</span><span class="p p-Indicator">,</span>
<span class="nt">    feature_thresholds</span><span class="p">:</span> <span class="nv">0.5</span><span class="p p-Indicator">,</span>
<span class="nt">    mode</span><span class="p">:</span> <span class="nv">train</span><span class="p p-Indicator">,</span>  <span class="c1"># starting mode for sampler</span>
<span class="nt">    save_datasets</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">test</span><span class="p p-Indicator">]</span>
<span class="p p-Indicator">}</span>
<span class="nt">train_model</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.TrainModel</span> <span class="p p-Indicator">{</span>
<span class="nt">    batch_size</span><span class="p">:</span> <span class="nv">64</span><span class="p p-Indicator">,</span>
<span class="nt">    max_steps</span><span class="p">:</span> <span class="nv">80000</span><span class="p p-Indicator">,</span>
<span class="nt">    report_stats_every_n_steps</span><span class="p">:</span> <span class="nv">16000</span><span class="p p-Indicator">,</span>
<span class="nt">    n_validation_samples</span><span class="p">:</span> <span class="nv">32000</span><span class="p p-Indicator">,</span>
<span class="nt">    n_test_samples</span><span class="p">:</span> <span class="nv">640000</span><span class="p p-Indicator">,</span>
<span class="nt">    cpu_n_threads</span><span class="p">:</span> <span class="nv">32</span><span class="p p-Indicator">,</span>
<span class="nt">    use_cuda</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    data_parallel</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    logging_verbosity</span><span class="p">:</span> <span class="nv">2</span><span class="p p-Indicator">,</span>
<span class="nt">    checkpoint_resume</span><span class="p">:</span> <span class="nv">False</span>
<span class="p p-Indicator">}</span>
<span class="nt">random_seed</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">133</span>
<span class="nt">output_dir</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">/path/to/output_dir</span>
<span class="nn">...</span>
</pre></div>
</div>
<div class="section" id="some-notes">
<h4>Some notes<a class="headerlink" href="#some-notes" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>Ordering of the keys does not matter.</p></li>
<li><p>We included many of the optional keys in this configuration. You do not need to specify these if you want to use their default values.</p></li>
<li><p>In this example, we specified a test partition in our intervals sampler by assigning a list of chromosomes to <code class="docutils literal notranslate"><span class="pre">test_holdout</span></code>. If no such holdout was specified (e.g. None or empty list), you would not be able to specify <code class="docutils literal notranslate"><span class="pre">n_test_samples</span></code> in <code class="docutils literal notranslate"><span class="pre">TrainModel</span></code> and would need to omit <code class="docutils literal notranslate"><span class="pre">evaluate</span></code> from the <code class="docutils literal notranslate"><span class="pre">ops</span></code> list.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">output_dir</span></code> is specified at the top-level and used by both the sampler and the <code class="docutils literal notranslate"><span class="pre">TrainModel</span></code> class.</p></li>
</ul>
</div>
</div>
<div class="section" id="evaluate-using-matrix-file-sampler">
<h3>Evaluate (using matrix file sampler)<a class="headerlink" href="#evaluate-using-matrix-file-sampler" title="Permalink to this headline">¶</a></h3>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nn">---</span>
<span class="nt">ops</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">evaluate</span><span class="p p-Indicator">]</span>
<span class="nt">model</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    path</span><span class="p">:</span> <span class="nv">/absolute/path/to/model/architecture.py</span><span class="p p-Indicator">,</span>
<span class="nt">    class</span><span class="p">:</span> <span class="nv">ModelArchitectureClassName</span><span class="p p-Indicator">,</span>
<span class="nt">    class_args</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">        arg1</span><span class="p">:</span> <span class="nv">val1</span><span class="p p-Indicator">,</span>
<span class="nt">        arg2</span><span class="p">:</span> <span class="nv">val2</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    non_strand_specific</span><span class="p">:</span> <span class="nv">mean</span>
<span class="p p-Indicator">}</span>
<span class="nt">sampler</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.samplers.file_samplers.MatFileSampler</span> <span class="p p-Indicator">{</span>
<span class="nt">    filepath</span><span class="p">:</span> <span class="nv">/path/to/test.mat</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_key</span><span class="p">:</span> <span class="nv">testxdata</span><span class="p p-Indicator">,</span>
<span class="nt">    targets_key</span><span class="p">:</span> <span class="nv">testdata</span><span class="p p-Indicator">,</span>
<span class="nt">    random_seed</span><span class="p">:</span> <span class="nv">456</span><span class="p p-Indicator">,</span>
<span class="nt">    shufle</span><span class="p">:</span> <span class="nv">False</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_batch_axis</span><span class="p">:</span> <span class="nv">0</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_alphabet_axis</span><span class="p">:</span> <span class="nv">1</span><span class="p p-Indicator">,</span>
<span class="nt">    targets_batch_axis</span><span class="p">:</span> <span class="nv">0</span>
<span class="p p-Indicator">}</span>
<span class="nt">evaluate_model</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.EvaluateModel</span> <span class="p p-Indicator">{</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/features_list.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    trained_model_path</span><span class="p">:</span> <span class="nv">/path/to/trained/model.pth.tar</span><span class="p p-Indicator">,</span>
<span class="nt">    batch_size</span><span class="p">:</span> <span class="nv">64</span><span class="p p-Indicator">,</span>
<span class="nt">    report_gt_feature_n_positives</span><span class="p">:</span> <span class="nv">50</span><span class="p p-Indicator">,</span>
<span class="nt">    use_cuda</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    data_parallel</span><span class="p">:</span> <span class="nv">False</span>
<span class="p p-Indicator">}</span>
<span class="nt">random_seed</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">123</span>
<span class="nt">output_dir</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">/path/to/output_dir</span>
<span class="nt">create_subdirectory</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="nn">...</span>
</pre></div>
</div>
<div class="section" id="id15">
<h4>Some notes<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>For the matrix file sampler, we assume that you know ahead of time the shape of the data matrix. That is, which dimension is the batch dimension? Sequence? Alphabet (should be size 4 for DNA/RNA)? You must specify the keys that end in <code class="docutils literal notranslate"><span class="pre">axis</span></code> unless the shape of the sequences matrix is <code class="docutils literal notranslate"><span class="pre">(n_samples,</span> <span class="pre">n_alphabet,</span> <span class="pre">n_sequence_length)</span></code> and the shape of the targets matrix is <code class="docutils literal notranslate"><span class="pre">(n_samples,</span> <span class="pre">n_targets)</span></code>.</p></li>
<li><p>In this case, since <code class="docutils literal notranslate"><span class="pre">create_subdirectory</span></code> is False, all outputs from evaluate are written to <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> directly (as opposed to being written in a timestamped subdirectory). Be careful of overwriting files.</p></li>
</ul>
</div>
</div>
<div class="section" id="analyze-sequences-variant-effect-prediction">
<h3>Analyze sequences (variant effect prediction)<a class="headerlink" href="#analyze-sequences-variant-effect-prediction" title="Permalink to this headline">¶</a></h3>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">ops</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">analyze</span><span class="p p-Indicator">]</span>
<span class="nt">model</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    path</span><span class="p">:</span> <span class="nv">/absolute/path/to/model/architecture.py</span><span class="p p-Indicator">,</span>
<span class="nt">    class</span><span class="p">:</span> <span class="nv">ModelArchitectureClassName</span><span class="p p-Indicator">,</span>
<span class="nt">    class_args</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">        arg1</span><span class="p">:</span> <span class="nv">val1</span><span class="p p-Indicator">,</span>
<span class="nt">        arg2</span><span class="p">:</span> <span class="nv">val2</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    non_strand_specific</span><span class="p">:</span> <span class="nv">mean</span>
<span class="p p-Indicator">}</span>
<span class="nt">analyze_sequences</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.predict.AnalyzeSequences</span> <span class="p p-Indicator">{</span>
<span class="nt">    trained_model_path</span><span class="p">:</span> <span class="nv">/path/to/trained/model.pth.tar</span><span class="p p-Indicator">,</span>
<span class="nt">    sequence_length</span><span class="p">:</span> <span class="nv">1000</span><span class="p p-Indicator">,</span>
<span class="nt">    features</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.utils.load_features_list</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/distinct_features.txt</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    batch_size</span><span class="p">:</span> <span class="nv">64</span><span class="p p-Indicator">,</span>
<span class="nt">    use_cuda</span><span class="p">:</span> <span class="nv">True</span><span class="p p-Indicator">,</span>
<span class="nt">    reference_sequence</span><span class="p">:</span> <span class="kt">!obj:selene_sdk.sequences.Genome</span> <span class="p p-Indicator">{</span>
<span class="nt">        input_path</span><span class="p">:</span> <span class="nv">/path/to/reference_sequence.fa</span>
    <span class="p p-Indicator">},</span>
<span class="nt">    write_mem_limit</span><span class="p">:</span> <span class="nv">75000</span>
<span class="p p-Indicator">}</span>
<span class="nt">variant_effect_prediction</span><span class="p">:</span> <span class="p p-Indicator">{</span>
<span class="nt">    vcf_files</span><span class="p">:</span> <span class="p p-Indicator">[</span>
        <span class="nv">/path/to/file1.vcf</span><span class="p p-Indicator">,</span>
        <span class="nv">/path/to/file2.vcf</span>
    <span class="p p-Indicator">],</span>
<span class="nt">    save_data</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">predictions</span><span class="p p-Indicator">,</span> <span class="nv">abs_diffs</span><span class="p p-Indicator">],</span>
<span class="nt">    output_dir</span><span class="p">:</span> <span class="nv">/path/to/output/predicts/dir</span><span class="p p-Indicator">,</span>
<span class="nt">    output_format</span><span class="p">:</span> <span class="nv">tsv</span><span class="p p-Indicator">,</span>
<span class="nt">    strand_index</span><span class="p">:</span> <span class="nv">9</span>
<span class="p p-Indicator">}</span>
<span class="nt">random_seed</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">123</span>
</pre></div>
</div>
<div class="section" id="id16">
<h4>Some notes<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>We ask that in all <code class="docutils literal notranslate"><span class="pre">analyze</span></code> cases, you specify the <code class="docutils literal notranslate"><span class="pre">output_dir</span></code> (when applicable) within the sub-operation dictionary. This is because only the sub-operation generates output, so there is no need to share this parameter across multiple configurations.</p></li>
<li><p>In this variant effect prediction example, Selene will go through each VCF file and get the model predictions for each variant (ref and alt). <code class="docutils literal notranslate"><span class="pre">analyze_sequences</span></code> must have the parameter <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code> so that Selene can create sequences centered at each variant position by querying the reference sequence file.</p></li>
<li><p>The output from this operation will be 6 files: 3 for each input VCF file. This is because of what is specified in <code class="docutils literal notranslate"><span class="pre">save_data</span></code>:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">predictions</span></code> will output 2 files per input VCF: the model predictions for all <code class="docutils literal notranslate"><span class="pre">ref</span></code>s and the model predictions for all <code class="docutils literal notranslate"><span class="pre">alts</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">abs_diffs</span></code> will output 1 file per input VCF: the absolute difference between the <code class="docutils literal notranslate"><span class="pre">ref</span></code> and <code class="docutils literal notranslate"><span class="pre">alt</span></code> model predictions. (Certainly, outputting the files from <code class="docutils literal notranslate"><span class="pre">predictions</span></code> is sufficient to compute <code class="docutils literal notranslate"><span class="pre">abs_diffs</span></code> yourself.)</p></li>
</ul>
</li>
</ul>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="tutorials.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2018, Selene Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>