

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene_sdk.interpret.ism_result &mdash; Selene master documentation</title>
  

  
  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.4.7)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html#help">Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-anaconda">Installing with Anaconda</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/installation.html#installing-selene-with-pip">Installing selene with pip:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-from-source">Installing from source</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#additional-dependency-for-running-the-cli">Additional dependency for running the CLI</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id1">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id2">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id3">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id4">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id5">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#variant-effect-prediction">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id6">Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#additional-note">Additional note</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#in-silico-mutagenesis"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#id13">Some notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#id14">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#exporting-a-selene-trained-model-to-kipoi">Exporting a Selene-trained model to Kipoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene_sdk.interpret.ism_result</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene_sdk.interpret.ism_result</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module provides the `ISMResult` class, which stores results for</span>
<span class="sd">an *in silico* mutagenesis experiment.</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">..sequences</span> <span class="kn">import</span> <span class="n">Genome</span>


<div class="viewcode-block" id="ISMResult"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.ISMResult">[docs]</a><span class="k">class</span> <span class="nc">ISMResult</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    An object storing the results of an *in silico* mutagenesis</span>
<span class="sd">    experiment.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_frame : pandas.DataFrame</span>
<span class="sd">        The data frame with the results from the *in silico*</span>
<span class="sd">        mutagenesis experiments.</span>

<span class="sd">    sequence_type : class, optional</span>
<span class="sd">        Default is `selene_sdk.sequences.Genome`. The type of sequence</span>
<span class="sd">        that the *in silico* mutagenesis results are associated</span>
<span class="sd">        with. This should generally be a subclass of</span>
<span class="sd">        `selene_sdk.sequences.Sequence`</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the input data frame contains a base not included in the</span>
<span class="sd">        alphabet of `sequence_type`.</span>

<span class="sd">    Exception</span>
<span class="sd">        If multiple reference positions are specified in the input</span>
<span class="sd">        data frame.</span>

<span class="sd">    Exception</span>
<span class="sd">        If the input data does not contain scores for every mutation</span>
<span class="sd">        at every position.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_frame</span><span class="p">,</span> <span class="n">sequence_type</span><span class="o">=</span><span class="n">Genome</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructs a new `selene_sdk.interpret.ISMResult` object.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Construct the reference sequence.</span>
        <span class="n">alpha</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)</span>
        <span class="n">ref_seq</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">data_frame</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">row_idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">data_frame</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="c1"># Skip the reference value</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">row_idx</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span>
                    <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">):</span>
                <span class="n">cur_ref</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">cur_ref</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">alpha</span> <span class="ow">and</span> <span class="n">cur_ref</span> <span class="o">!=</span> <span class="n">sequence_type</span><span class="o">.</span><span class="n">UNK_BASE</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Found character </span><span class="se">\&#39;</span><span class="si">{0}</span><span class="se">\&#39;</span><span class="s2"> from outside current alphabet&quot;</span>
                        <span class="s2">&quot; on row </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cur_ref</span><span class="p">,</span> <span class="n">row_idx</span><span class="p">))</span>
                <span class="n">i</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
                <span class="n">seen</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">ref_seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ref_seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">cur_ref</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                            <span class="s2">&quot;Found 2 different letters for reference </span><span class="se">\&#39;</span><span class="si">{0}</span><span class="se">\&#39;</span><span class="s2">&quot;</span>
                            <span class="s2">&quot; and </span><span class="se">\&#39;</span><span class="si">{1}</span><span class="se">\&#39;</span><span class="s2"> on row </span><span class="si">{2}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                                              <span class="n">cur_ref</span><span class="p">,</span>
                                                              <span class="n">row_idx</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">ref_seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">cur_ref</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seen</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                <span class="s2">&quot;Expected characters for </span><span class="si">{0}</span><span class="s2"> positions, but only found </span><span class="si">{1}</span><span class="s2"> of &quot;</span>
                <span class="s2">&quot;them.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">seen</span><span class="p">)))</span>
        <span class="n">ref_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_reference_sequence</span> <span class="o">=</span> <span class="n">ref_seq</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_data_frame</span> <span class="o">=</span> <span class="n">data_frame</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sequence_type</span> <span class="o">=</span> <span class="n">sequence_type</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">reference_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The reference sequence that the *in silico* mutagenesis</span>
<span class="sd">        experiment was performed on.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            The reference sequence (i.e. non-mutated input) as a</span>
<span class="sd">            string of characters.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference_sequence</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">sequence_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The type of underlying sequence. This should generally be a</span>
<span class="sd">        subclass of `selene_sdk.sequences.Sequence`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        class</span>
<span class="sd">            The type of sequence that the *in silico* mutagenesis was</span>
<span class="sd">            performed on.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequence_type</span>

<div class="viewcode-block" id="ISMResult.get_score_matrix_for"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.ISMResult.get_score_matrix_for">[docs]</a>    <span class="k">def</span> <span class="nf">get_score_matrix_for</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">reference_mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                             <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts a feature from the *in silico* mutagenesis results</span>
<span class="sd">        as a matrix, where the reference base positions hold the value</span>
<span class="sd">        for the reference prediction, and alternative positions hold the</span>
<span class="sd">        results for making a one-base change from the reference base to</span>
<span class="sd">        the specified alternative base.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        feature : str</span>
<span class="sd">            The name of the feature to extract as a matrix.</span>

<span class="sd">        reference_mask : float or None, optional</span>
<span class="sd">            Default is `None`. A value to mask the reference entries</span>
<span class="sd">            with. If left as `None`, then no masking will be performed</span>
<span class="sd">            on the reference positions.</span>

<span class="sd">        dtype : numpy.dtype, optional</span>
<span class="sd">            Default is `numpy.float64`. The data type to use for the</span>
<span class="sd">            returned matrix.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        numpy.ndarray</span>
<span class="sd">            A :math:`L \\times N` shaped array (where :math:`L` is the</span>
<span class="sd">            sequence length, and :math:`N` is the size of the alphabet</span>
<span class="sd">            of `sequence_type`) that holds the results from the</span>
<span class="sd">            *in silico* mutagenesis experiment for the specified</span>
<span class="sd">            feature. The elements will be of type `dtype`.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the input data frame contains a base not included in the</span>
<span class="sd">            alphabet of `sequence_type`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">reference_mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">reference_mask</span> <span class="o">=</span> <span class="n">dtype</span><span class="p">(</span><span class="n">reference_mask</span><span class="p">)</span>
        <span class="n">ret</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequence_type</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_reference_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="n">dtype</span><span class="p">)</span>
        <span class="n">ret</span><span class="p">[</span><span class="n">ret</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.</span>  <span class="c1"># Set N&#39;s to zero to avoid spurious masking.</span>
        <span class="n">alpha</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_sequence_type</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">row_idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data_frame</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="c1"># Extract reference value in first row.</span>
            <span class="k">if</span> <span class="n">row_idx</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">reference_mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">reference_mask</span> <span class="o">=</span> <span class="n">dtype</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">feature</span><span class="p">])</span>
                    <span class="n">ret</span> <span class="o">*=</span> <span class="n">reference_mask</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">reference_mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">reference_mask</span> <span class="o">=</span> <span class="mf">0.</span>
                    <span class="n">ret</span> <span class="o">*=</span> <span class="n">reference_mask</span>
            <span class="n">base</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span>
            <span class="n">i</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">base</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">alpha</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">base</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequence_type</span><span class="o">.</span><span class="n">UNK_BASE</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Found character </span><span class="se">\&#39;</span><span class="si">{0}</span><span class="se">\&#39;</span><span class="s2"> from outside current alphabet&quot;</span>
                        <span class="s2">&quot; on row </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">base</span><span class="p">,</span> <span class="n">row_idx</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ret</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequence_type</span><span class="o">.</span><span class="n">BASE_TO_INDEX</span><span class="p">[</span><span class="n">base</span><span class="p">]]</span> <span class="o">=</span> <span class="n">dtype</span><span class="p">(</span>
                    <span class="n">row</span><span class="p">[</span><span class="n">feature</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">ret</span></div>

<div class="viewcode-block" id="ISMResult.from_file"><a class="viewcode-back" href="../../../interpret.html#selene_sdk.interpret.ISMResult.from_file">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">from_file</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">sequence_type</span><span class="o">=</span><span class="n">Genome</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads a `selene_sdk.interpret.ISMResult` from a `pandas.DataFrame`</span>
<span class="sd">        stored in a file of comma separated values (CSV).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : str</span>
<span class="sd">            A path to the file of comma separated input values.</span>

<span class="sd">        sequence_type : class, optional</span>
<span class="sd">            Default is `selene_sdk.sequences.Genome`. The type of sequence</span>
<span class="sd">            that the *in silico* mutagenesis results are associated</span>
<span class="sd">            with. This should generally be a subclass of</span>
<span class="sd">            `selene_sdk.sequences.Sequence`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        selene_sdk.interpret.ISMResult</span>
<span class="sd">            The *in silico* mutagenesis results that were stored in the</span>
<span class="sd">            specified input file.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">ISMResult</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                     <span class="n">dtype</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">na_values</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                     <span class="n">keep_default_na</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
                         <span class="n">sequence_type</span><span class="o">=</span><span class="n">sequence_type</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 



</body>
</html>